{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e32b56e-e65d-41c3-b527-c7ad1573c961",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for testing GC perturbation to Alu and Alu-flanking sequences "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "fe2aefff-97d0-4e71-8966-6754a57a9105",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7f90f94a-7223-46c8-a43b-b17bdb07afd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import basenji\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import os, psutil, io, gzip, time, sys\n",
    "import pysam\n",
    "import warnings\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from pybedtools import BedTool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ef87c230-ab6b-4310-be00-05500e60329f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ce54f097-cf1d-41a8-bff9-10d1ea4de9e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_gc(chrom, start, end):\n",
    "    seq = hg38_fa.fetch(chrom, start, end).upper()\n",
    "    gc_count = seq.count('G') + seq.count('C')\n",
    "    return (gc_count / len(seq)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7e0d4522-bcbc-4fbc-b7de-defc81e1de15",
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_file = '/home/data/genomes/hg38.fa'\n",
    "hg38_fa = pysam.Fastafile(hg38_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "44519908-4cec-441a-8d76-527cb04f7eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input variables\n",
    "in_file = ''\n",
    "input_sequences = None\n",
    "fasta_path = hg38_file\n",
    "genome = 'hg38'\n",
    "scores_to_use = ['mse', 'corr']\n",
    "shift_by = [-1, 0, 1]\n",
    "out_file = '20240116_test'\n",
    "out_dir = '/home/shu_zhang/SuPreMo/test_out/'\n",
    "seq_len = 1048576\n",
    "\n",
    "revcomp = 'add_revcomp'\n",
    "augment = True\n",
    "get_seq = True\n",
    "get_tracks = False\n",
    "get_maps = False\n",
    "get_Akita_scores = False\n",
    "var_set_size = 0\n",
    "svlen_limit=700e3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ecb5e9a7-970f-48cc-bb64-e509dcf7bec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add option for mutate\n",
    "#first option is \"gc\" or \"shuffle\"\n",
    "#second and third are flanking len (upstream and downstream of variant)\n",
    "\n",
    "#gc_mutate=['up_flank', 'down_flank', 'mut_percent']\n",
    "gc_mutate={'up_flank': 1000, 'down_flank':1000, 'mut_percent':0.2}\n",
    "\n",
    "#shuffle_mutate=['up_flank', 'down_flank']\n",
    "shuffle_mutate={'up_flank': 1000, 'down_flank':1000}\n",
    "\n",
    "#tf_mutate=['up_flank', 'down_flank', 'tf_mutate_type', 'alter_variant', 'track_path']\n",
    "#options for tf_mutate_type: 'shuffle', 'INV'\n",
    "#options for alter_variant: True/False\n",
    "#track path: string with path to the tf binding motif track (from jaspar, or in similar bed format )\n",
    "tf_mutate={'up_flank': 1000, 'down_flank':1000, 'tf_mutate_type': 'INV', 'alter_variant': False, 'track_path':''}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ba038f88-14bc-43ff-bfed-8c8c9407ec51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1000"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc_mutate['up_flank']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6c0ef53d-638c-45ee-978d-71a8c412be5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NOTE: WHEN PARSING ARGUMENTS NEED TO CONVERT STRING REPRESENTATION INTO DICTIONARY \n",
    "def parse_arguments(argument_string):\n",
    "    # Convert the string representation of the dictionary into an actual dictionary\n",
    "    arguments = ast.literal_eval(argument_string)\n",
    "    return arguments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f681e462-5067-4d1d-a1cd-d918fecce4a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi\n"
     ]
    }
   ],
   "source": [
    "if tf_mutate:\n",
    "    print('hi')\n",
    "else:\n",
    "    print('none')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f61cd628-dae1-4dff-b881-c4bbe7c30a5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "419430.4\n"
     ]
    }
   ],
   "source": [
    "# # # # # # # # # # # # # # # # # # # # # # # # # # # # #\n",
    "# Adjust inputs from arguments\n",
    "\n",
    "\n",
    "# Handle paths\n",
    "\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "# This file path and repo path\n",
    "repo_path = '/home/shu_zhang/SuPreMo/'\n",
    "\n",
    "# Output directory and file\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "out_file = os.path.join(out_dir, out_file)\n",
    "\n",
    "\n",
    "# Data path\n",
    "chrom_lengths_path = f'{repo_path}/data/chrom_lengths_{genome}'\n",
    "centromere_coords_path = f'{repo_path}/data/centromere_coords_{genome}'\n",
    "   \n",
    "\n",
    "# Adjust shift input: Remove shifts that are outside of allowed range\n",
    "max_shift = 0.4*seq_len\n",
    "print(max_shift)\n",
    "shift_by = [x for x in shift_by if x > -max_shift and x < max_shift]\n",
    "\n",
    "\n",
    "# Adjust input for taking the reverse complement\n",
    "if revcomp == 'no_revcomp':\n",
    "    revcomp_decision = [False]\n",
    "elif revcomp == 'add_revcomp':\n",
    "    revcomp_decision = [False, True]\n",
    "elif revcomp == 'only_revcomp':\n",
    "    revcomp_decision = [True]\n",
    "\n",
    "if augment and shift_by == [0] and revcomp == 'no_revcomp':\n",
    "    shift_by = [-1,0,1]\n",
    "    revcomp_decision = [False, True]\n",
    "\n",
    "revcomp_decision_i = revcomp_decision\n",
    "\n",
    "if input_sequences is not None:\n",
    "    seq_names = pysam.Fastafile(input_sequences).references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bb88120f-ff91-4726-b7b6-4a3458062da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionaries to save sequences, maps, and disruption score tracks, if specified\n",
    "if get_seq:\n",
    "    sequences = {}\n",
    "    \n",
    "# # # # # # # # # # # # # # # # # # # # # # # # # # # # #\n",
    "# Read in (and adjust) data\n",
    "\n",
    "\n",
    "chrom_lengths = pd.read_table(chrom_lengths_path, header = None, names = ['CHROM', 'chrom_max'])\n",
    "centromere_coords = pd.read_table(centromere_coords_path, sep = '\\t')\n",
    "fasta_open = pysam.Fastafile(fasta_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ae4f0726-f583-42ff-8f7b-1197c2b6eede",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/shu_zhang/SuPreMo\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/shu_zhang/miniconda3/envs/supremo_akita_env/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "%cd /home/shu_zhang/SuPreMo/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7f28382d-5df6-47fe-9ace-0874a737b427",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi\n"
     ]
    }
   ],
   "source": [
    "# Assign necessary values to variables across module\n",
    "\n",
    "# Module 1: reading utilities\n",
    "sys.path.insert(0, './scripts/')\n",
    "import reading_utils\n",
    "reading_utils.var_set_size = var_set_size\n",
    "\n",
    "\n",
    "# Module 2: get_seq utilities\n",
    "import get_seq_utils\n",
    "get_seq_utils.fasta_open = fasta_open\n",
    "get_seq_utils.chrom_lengths = chrom_lengths\n",
    "get_seq_utils.centromere_coords = centromere_coords\n",
    "\n",
    "get_seq_utils.svlen_limit = svlen_limit\n",
    "get_seq_utils.seq_length = seq_len\n",
    "get_seq_utils.half_patch_size = round(seq_len/2)\n",
    "\n",
    "\n",
    "# Module 2: get_Akita_scores utilities\n",
    "if get_Akita_scores:\n",
    "    import get_Akita_scores_utils\n",
    "    get_Akita_scores_utils.chrom_lengths = chrom_lengths\n",
    "    get_Akita_scores_utils.centromere_coords = centromere_coords\n",
    "    \n",
    "#module 3: seq-MUTATE_UTILS\n",
    "mutate=False\n",
    "if gc_mutate or shuffle_mutate or tf_mutate:\n",
    "    mutate=True\n",
    "\n",
    "if mutate:\n",
    "    print('hi')\n",
    "    import seq_mutate_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "df5b9ba1-1aae-4d78-ba40-2a1507da94c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames = ['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']\n",
    "\n",
    "test_dir='/home/shu_zhang/SuPreMo/test_data/test_set_edge_cases/'\n",
    "test_sv=pd.read_csv(f'{test_dir}test_set_edge_SV.bed', sep='\\t', names=colnames)\n",
    "test_sv_no_bnd=test_sv[test_sv.SVTYPE!='BND']\n",
    "test_simple=pd.read_csv(f'{test_dir}test_set_edge_simple.bed', sep='\\t', names=colnames)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "05993bc1-83c5-49fc-89fa-f0e50892b811",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames = ['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']\n",
    "\n",
    "custom_perb='/home/shu_zhang/SuPreMo/test_data/custom_perturbations/input/CTCF_del_symb_alleles.bed'\n",
    "\n",
    "test_sv=pd.read_csv(custom_perb, sep='\\t', names=colnames)\n",
    "test_sv_no_bnd=test_sv[test_sv.SVTYPE!='BND']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "354f3e93-c6b5-4fbc-93ed-5717ace2f28d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>END</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>SVLEN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100276250</td>\n",
       "      <td>100276269</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101106697</td>\n",
       "      <td>101106716</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101159421</td>\n",
       "      <td>101159440</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101442377</td>\n",
       "      <td>101442396</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101526743</td>\n",
       "      <td>101526762</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101595702</td>\n",
       "      <td>101595721</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101693506</td>\n",
       "      <td>101693525</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101744879</td>\n",
       "      <td>101744898</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>102007853</td>\n",
       "      <td>102007872</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10192988</td>\n",
       "      <td>10193007</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>103208826</td>\n",
       "      <td>103208845</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>103344645</td>\n",
       "      <td>103344664</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>103776583</td>\n",
       "      <td>103776602</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>989207</td>\n",
       "      <td>989226</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>104132695</td>\n",
       "      <td>104132714</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>104271555</td>\n",
       "      <td>104271574</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10410881</td>\n",
       "      <td>10410900</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10430990</td>\n",
       "      <td>10431009</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>104992481</td>\n",
       "      <td>104992500</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>106055418</td>\n",
       "      <td>106055437</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>106137878</td>\n",
       "      <td>106137897</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10664817</td>\n",
       "      <td>10664836</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107486411</td>\n",
       "      <td>107486430</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107497402</td>\n",
       "      <td>107497421</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10733986</td>\n",
       "      <td>10734005</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107671377</td>\n",
       "      <td>107671396</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107873874</td>\n",
       "      <td>107873893</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10776655</td>\n",
       "      <td>10776674</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108094709</td>\n",
       "      <td>108094728</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10861279</td>\n",
       "      <td>10861298</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108906070</td>\n",
       "      <td>108906089</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108987868</td>\n",
       "      <td>108987887</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10892736</td>\n",
       "      <td>10892755</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109221228</td>\n",
       "      <td>109221247</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109444187</td>\n",
       "      <td>109444206</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109574280</td>\n",
       "      <td>109574299</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109743693</td>\n",
       "      <td>109743712</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109974651</td>\n",
       "      <td>109974670</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110255473</td>\n",
       "      <td>110255492</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110373895</td>\n",
       "      <td>110373914</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr1</td>\n",
       "      <td>110529500</td>\n",
       "      <td>110529519</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11139402</td>\n",
       "      <td>11139421</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>111744127</td>\n",
       "      <td>111744146</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112055523</td>\n",
       "      <td>112055542</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112419004</td>\n",
       "      <td>112419023</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112532948</td>\n",
       "      <td>112532967</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113062991</td>\n",
       "      <td>113063010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113225265</td>\n",
       "      <td>113225284</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113539673</td>\n",
       "      <td>113539692</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114180016</td>\n",
       "      <td>114180035</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS        END REF ALT SVTYPE  SVLEN\n",
       "0   chr1  100276250  100276269   -   -    DEL     19\n",
       "1   chr1  101106697  101106716   -   -    DEL     19\n",
       "2   chr1  101159421  101159440   -   -    DEL     19\n",
       "3   chr1  101442377  101442396   -   -    DEL     19\n",
       "4   chr1  101526743  101526762   -   -    DEL     19\n",
       "5   chr1  101595702  101595721   -   -    DEL     19\n",
       "6   chr1  101693506  101693525   -   -    DEL     19\n",
       "7   chr1  101744879  101744898   -   -    DEL     19\n",
       "8   chr1  102007853  102007872   -   -    DEL     19\n",
       "9   chr1   10192988   10193007   -   -    DEL     19\n",
       "10  chr1  103208826  103208845   -   -    DEL     19\n",
       "11  chr1  103344645  103344664   -   -    DEL     19\n",
       "12  chr1  103776583  103776602   -   -    DEL     19\n",
       "13  chr1     989207     989226   -   -    DEL     19\n",
       "14  chr1  104132695  104132714   -   -    DEL     19\n",
       "15  chr1  104271555  104271574   -   -    DEL     19\n",
       "16  chr1   10410881   10410900   -   -    DEL     19\n",
       "17  chr1   10430990   10431009   -   -    DEL     19\n",
       "18  chr1  104992481  104992500   -   -    DEL     19\n",
       "19  chr1  106055418  106055437   -   -    DEL     19\n",
       "20  chr1  106137878  106137897   -   -    DEL     19\n",
       "21  chr1   10664817   10664836   -   -    DEL     19\n",
       "22  chr1  107486411  107486430   -   -    DEL     19\n",
       "23  chr1  107497402  107497421   -   -    DEL     19\n",
       "24  chr1   10733986   10734005   -   -    DEL     19\n",
       "25  chr1  107671377  107671396   -   -    DEL     19\n",
       "26  chr1  107873874  107873893   -   -    DEL     19\n",
       "27  chr1   10776655   10776674   -   -    DEL     19\n",
       "28  chr1  108094709  108094728   -   -    DEL     19\n",
       "29  chr1   10861279   10861298   -   -    DEL     19\n",
       "30  chr1  108906070  108906089   -   -    DEL     19\n",
       "31  chr1  108987868  108987887   -   -    DEL     19\n",
       "32  chr1   10892736   10892755   -   -    DEL     19\n",
       "33  chr1  109221228  109221247   -   -    DEL     19\n",
       "34  chr1  109444187  109444206   -   -    DEL     19\n",
       "35  chr1  109574280  109574299   -   -    DEL     19\n",
       "36  chr1  109743693  109743712   -   -    DEL     19\n",
       "37  chr1  109974651  109974670   -   -    DEL     19\n",
       "38  chr1  110255473  110255492   -   -    DEL     19\n",
       "39  chr1  110373895  110373914   -   -    DEL     19\n",
       "40  chr1  110529500  110529519   -   -    DEL     19\n",
       "41  chr1   11139402   11139421   -   -    DEL     19\n",
       "42  chr1  111744127  111744146   -   -    DEL     19\n",
       "43  chr1  112055523  112055542   -   -    DEL     19\n",
       "44  chr1  112419004  112419023   -   -    DEL     19\n",
       "45  chr1  112532948  112532967   -   -    DEL     19\n",
       "46  chr1  113062991  113063010   -   -    DEL     19\n",
       "47  chr1  113225265  113225284   -   -    DEL     19\n",
       "48  chr1  113539673  113539692   -   -    DEL     19\n",
       "49  chr1  114180016  114180035   -   -    DEL     19"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants=test_sv_no_bnd\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "20daae0a-a0c9-40ec-b8b4-ac9d8b07aef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the first 13 for testing \n",
    "mutate_dir='/home/shu_zhang/SuPreMo/mutate_data/test_data/'\n",
    "variants[0:20].to_csv(f'{mutate_dir}CTCF_DEL_20.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "69e70799-7070-44b6-bd3d-9e57bf6bda2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_654536/2845672274.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  variants['var_index'] = list(range(var_set*var_set_size, var_set*var_set_size + len(variants)))\n",
      "/tmp/ipykernel_654536/2845672274.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  variants['var_index'] = variants['var_index'].astype(str)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "i 0\n",
      "variant CHROM                                                     chr1\n",
      "POS                                                  244397848\n",
      "END                                                244398348.0\n",
      "REF          AAACTAAACTAAAATAAACTAAAATAAACTAAAATAAACTAAACTA...\n",
      "ALT                                                          A\n",
      "SVTYPE                                                     DEL\n",
      "SVLEN                                                    500.0\n",
      "var_index                                                    0\n",
      "Name: 0, dtype: object\n",
      "shift -1\n",
      "revcomp False\n",
      "input_sequences None\n",
      "var rel pos [524038_524288]\n",
      "var rel pos from sequences\n",
      "[524038, 524288]\n",
      "AATATTTAAT\n",
      "0 (-1 shift): Error: name 'posflank' is not defined\n",
      "shift 0\n",
      "revcomp False\n",
      "input_sequences None\n",
      "var rel pos [524037_524287]\n",
      "var rel pos from sequences\n",
      "[524037, 524287]\n",
      "ATATTTAATG\n",
      "0 (0 shift): Error: name 'posflank' is not defined\n",
      "revcomp True\n",
      "input_sequences None\n",
      "var rel pos [524037_524287]\n",
      "var rel pos from sequences\n",
      "[524037, 524287]\n",
      "TAAAATTATT\n",
      "0 (0 shift_revcomp): Error: name 'posflank' is not defined\n",
      "shift 1\n",
      "revcomp False\n",
      "input_sequences None\n",
      "var rel pos [524036_524286]\n",
      "var rel pos from sequences\n",
      "[524036, 524286]\n",
      "TATTTAATGG\n",
      "0 (1 shift): Error: name 'posflank' is not defined\n",
      "i 1\n",
      "variant CHROM               chr1\n",
      "POS            244397848\n",
      "END          244398348.0\n",
      "REF                    A\n",
      "ALT                <DEL>\n",
      "SVTYPE               DEL\n",
      "SVLEN              500.0\n",
      "var_index              1\n",
      "Name: 1, dtype: object\n",
      "shift -1\n",
      "revcomp False\n",
      "input_sequences None\n",
      "var rel pos [524038_524288]\n",
      "var rel pos from sequences\n",
      "[524038, 524288]\n",
      "AATATTTAAT\n",
      "1 (-1 shift): Error: name 'posflank' is not defined\n",
      "shift 0\n",
      "revcomp False\n",
      "input_sequences None\n",
      "var rel pos [524037_524287]\n",
      "var rel pos from sequences\n",
      "[524037, 524287]\n",
      "ATATTTAATG\n",
      "1 (0 shift): Error: name 'posflank' is not defined\n",
      "revcomp True\n",
      "input_sequences None\n",
      "var rel pos [524037_524287]\n",
      "var rel pos from sequences\n",
      "[524037, 524287]\n",
      "TAAAATTATT\n",
      "1 (0 shift_revcomp): Error: name 'posflank' is not defined\n",
      "shift 1\n",
      "revcomp False\n",
      "input_sequences None\n",
      "var rel pos [524036_524286]\n",
      "var rel pos from sequences\n",
      "[524036, 524286]\n",
      "TATTTAATGG\n",
      "1 (1 shift): Error: name 'posflank' is not defined\n"
     ]
    }
   ],
   "source": [
    "# # # # # # # # # # # # # # # # # # # # # # # # # # # # #\n",
    "    # Run: Make Akita predictions and calculate disruption scores\n",
    "nt = ['A', 'T', 'C', 'G']\n",
    "    \n",
    "var_set = 0\n",
    "var_set_list = []\n",
    "\n",
    "while True:\n",
    "    \n",
    "    # Read in variants\n",
    "    #variants = reading_utils.read_input(in_file, var_set)\n",
    "    if len(variants) == 0:\n",
    "        break\n",
    "        \n",
    "        \n",
    "    # Index input based on row number and create output with same indexes\n",
    "    variants['var_index'] = list(range(var_set*var_set_size, var_set*var_set_size + len(variants)))\n",
    "    variants['var_index'] = variants['var_index'].astype(str)\n",
    "\n",
    "    \n",
    "    # If there are multiple alternate alleles, split those into new rows and indexes\n",
    "    if any([',' in x for x in variants.ALT]):\n",
    "\n",
    "        variants = (variants\n",
    "                 .set_index(['CHROM', 'POS', 'REF', 'var_index'])\n",
    "                 .apply(lambda x: x.str.split(',').explode())\n",
    "                 .reset_index())\n",
    "\n",
    "        g = variants.groupby(['var_index'])\n",
    "        variants.loc[g['var_index'].transform('size').gt(1),\n",
    "               'var_index'] += '-'+g.cumcount().astype(str)\n",
    "        \n",
    "    variant_scores = pd.DataFrame({'var_index':variants.var_index})\n",
    "    \n",
    "  \n",
    "    #filtering normally occurs here, removed because these tests don't need it\n",
    "\n",
    "\n",
    "    \n",
    "    # Loop through each row (not index) and get disruption scores \n",
    "    for i in range(len(variants))[0:2]:\n",
    "        print('i', i)\n",
    "\n",
    "        variant = variants.iloc[i]\n",
    "\n",
    "        print('variant', variant)\n",
    "\n",
    "        var_index = variant.var_index\n",
    "        CHR = variant.CHROM\n",
    "        POS = variant.POS\n",
    "        REF = variant.REF\n",
    "        ALT = variant.ALT\n",
    "            \n",
    "\n",
    "        if 'SVTYPE' in variants.columns:\n",
    "            END = variant.END\n",
    "            SVTYPE = variant.SVTYPE\n",
    "            SVLEN = variant.SVLEN\n",
    "        else:\n",
    "            END = np.nan\n",
    "            SVTYPE = np.nan\n",
    "            SVLEN = 0\n",
    "\n",
    "        for shift in shift_by:\n",
    "            print('shift', shift)\n",
    "\n",
    "            # Take reverse complement only with 0 shift\n",
    "            if shift != 0 & True in revcomp_decision:\n",
    "                revcomp_decision_i = [False]\n",
    "            else:\n",
    "                revcomp_decision_i = revcomp_decision\n",
    "            #revcomp_decision is list of T/F\n",
    "            for revcomp in revcomp_decision_i:\n",
    "                print('revcomp', revcomp)\n",
    "\n",
    "                try:\n",
    "\n",
    "                    if revcomp:\n",
    "                        revcomp_annot = '_revcomp'\n",
    "                    else:\n",
    "                        revcomp_annot = ''\n",
    "\n",
    "                    print('input_sequences', input_sequences)\n",
    "                    if input_sequences is not None:\n",
    "\n",
    "                        # Generate sequences_i from sequence input\n",
    "                        if revcomp_annot == '':\n",
    "                            sequence_names = [x for x in seq_names if x.startswith(f'{var_index}_{shift}') and \n",
    "                                              'revcomp' not in x]\n",
    "                        elif revcomp_annot == '_revcomp':\n",
    "                            sequence_names = [x for x in seq_names if x.startswith(f'{var_index}_{shift}{revcomp_annot}')]\n",
    "\n",
    "                        sequences_i = []\n",
    "                        print('sequence_names', sequence_names)\n",
    "                        for sequence_name in sequence_names:\n",
    "                            #sequences_i.append(pysam.Fastafile(input_sequences).fetch(sequence_name, 0, seq_len).upper())\n",
    "                            sequences_i.append(pysam.Fastafile(input_sequences).fetch(sequence_name, 0, seq_len).upper())\n",
    "\n",
    "\n",
    "                        sequences_i.append([int(x) for x in sequence_name.split('[')[1].split(']')[0].split('_')])\n",
    "\n",
    "\n",
    "                    else:\n",
    "\n",
    "                        # Create sequences_i from variant input\n",
    "                        #get_sequences_SV (for non-BND) returns:\n",
    "                        #REF_seq, ALT_seq, [var_rel_pos_REF, var_rel_pos_ALT]\n",
    "                        #var_rel_pos_REF is variant position relative to the reference sequence \n",
    "                        #first one is REF, second is ALT\n",
    "                        #524288 is halfway point of 1Mb region \n",
    "                        \n",
    "                        #(relative position of POS, or start of varianat)\n",
    "                        #var_rel_pos_ALT is variant position relative to reference still? \n",
    "                        sequences_i = get_seq_utils.get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)\n",
    "                        #print('sequences')\n",
    "                        #print(len(sequences_i))\n",
    "\n",
    "                    if get_seq:\n",
    "\n",
    "                        # Get relative position of variant in sequence\n",
    "                        var_rel_pos = str(sequences_i[-1]).replace(', ', '_')\n",
    "                        print('var rel pos', var_rel_pos)\n",
    "\n",
    "                        for ii in range(len(sequences_i[:-1][:3])): \n",
    "                            sequences[f'{var_index}_{shift}{revcomp_annot}_{ii}_{var_rel_pos}'] = sequences_i[:-1][ii] \n",
    "                    #shu edit here\n",
    "                    #insert options post sequence whatever \n",
    "                    \n",
    "\n",
    "                    # if gc_mutate>0:\n",
    "\n",
    "                    #     # sequences is tuple of length 3\n",
    "                    #     # 0 index is presumably REF\n",
    "                    #     # 1 index is presubably ALT\n",
    "                    #     # 2 index is the var_index of the two sequences [REF, ALT], in a list format \n",
    "                    #     var_rel_pos=sequences_i[-1][0]\n",
    "                    #     new_seq=seq_mutate_utils.mutate_gc(sequences_i[0], var_rel_pos, var_rel_pos+SVLEN, \n",
    "                    #                                        posflank,endflank, revcomp, gc_mutate)\n",
    "                                            \n",
    "                    if mutate:\n",
    "                        #we want to call a master mutate function\n",
    "                        #that takes in parameters gc_mutate, shuffle_mutate, tf_mutate\n",
    "                        #does mutations based on either of these \n",
    "                        #eventually, we want them to be able to get tacked onto sequences_i\n",
    "\n",
    "                        #makbe instead of \n",
    "                        pass\n",
    "\n",
    "                        #right now, sequences_i:\n",
    "                            # 1. sequences_i[0] is REF\n",
    "                            # 2. sequences_i[1] is REF + mutation, with whatever it is\n",
    "                            # 3. sequences_i[2] is var_indices of the two sequences [REF, ALT], in a list format \n",
    "                        sequences_i=seq_mutate_utils.mutate_sequence(CHR, POS, SVTYPE, SVLEN, sequences_i, shift, revcomp, \n",
    "                                                                    gc_mutate, shuffle_mutate, tf_mutate) \n",
    "\n",
    "                            \n",
    "                    if get_Akita_scores:\n",
    "\n",
    "                        scores = get_Akita_scores_utils.get_scores(POS, SVTYPE, SVLEN, \n",
    "                                                                   sequences_i, scores_to_use, \n",
    "                                                                   shift, revcomp, \n",
    "                                                                   get_tracks, get_maps)\n",
    "                        \n",
    "                        print('akita scores')\n",
    "                        print(scores)\n",
    "\n",
    "\n",
    "                        if get_tracks:\n",
    "                            for track in [x for x in scores.keys() if 'track' in x]:\n",
    "                                variant_tracks[f'{var_index}_{track}_{shift}{revcomp_annot}'] = scores[track]\n",
    "                                del scores[track]\n",
    "\n",
    "                        if get_maps:\n",
    "                            variant_maps[f'{var_index}_{shift}{revcomp_annot}'] = scores['maps']\n",
    "                            del scores['maps']\n",
    "\n",
    "                        for score in scores:\n",
    "                            variant_scores.loc[variant_scores.var_index == var_index, \n",
    "                                               f'{score}_{shift}{revcomp_annot}'] = scores[score]\n",
    "\n",
    "\n",
    "                    print('hi')\n",
    "                    print(str(var_index) + ' (' + str(shift) + f' shift{revcomp_annot})')\n",
    "\n",
    "                except Exception as e: \n",
    "\n",
    "                    print(str(var_index) + ' (' + str(shift) + f' shift{revcomp_annot})' + ': Error:', e)\n",
    "\n",
    "                    pass\n",
    " \n",
    "  \n",
    "    # Write sequences to fasta file\n",
    "    if get_seq:\n",
    "\n",
    "        if var_set == 0:\n",
    "            sequences_all = sequences.copy()\n",
    "        else:\n",
    "            sequences_all.update(sequences)\n",
    "        \n",
    "    break\n",
    "        \n",
    "    var_set_list.append(var_set)\n",
    "    var_set += 1\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20de7e79-e81b-43b4-969a-805781c9d193",
   "metadata": {},
   "outputs": [],
   "source": [
    "# normally this is right above \"if get_seq:\"\n",
    "\n",
    "#     # Write scores to data frame\n",
    "#     if get_Akita_scores:\n",
    "\n",
    "#         # Take average of augmented sequences\n",
    "#         if augment:\n",
    "#             for score in scores:\n",
    "#                 cols = [x for x in variant_scores.columns if score in x]\n",
    "#                 variant_scores[f'{score}_mean'] = variant_scores[cols].mean(axis = 1)\n",
    "#                 variant_scores[f'{score}_median'] = variant_scores[cols].median(axis = 1)\n",
    "#                 variant_scores.drop(cols, axis = 1, inplace = True)\n",
    "\n",
    "#         # Convert scores from float to string so you can merge scores for variants with multiple alleles\n",
    "#         for col in variant_scores.iloc[:,1:].columns:\n",
    "#             variant_scores[col] = [format(x, '.20f') for x in variant_scores[col]]\n",
    "            \n",
    "                \n",
    "#         # Join scores for alternate alleles, separated by a comma\n",
    "#         if any(['-' in x for x in variant_scores.var_index]):\n",
    "#             variant_scores['var_index'] = variant_scores.var_index.str.split('-').str[0]\n",
    "            \n",
    "#             variant_scores = (variant_scores\n",
    "#                               .set_index(['var_index'], drop = False)\n",
    "#                               .rename(columns = {'var_index':'var_index2'})\n",
    "#                               .groupby('var_index2')\n",
    "#                               .transform(','.join)\n",
    "#                               .reset_index()\n",
    "#                               .drop_duplicates())\n",
    "        \n",
    "#         if var_set == 0:\n",
    "#             variant_scores.to_csv(f'{out_file}_scores_{var_set}', sep = '\\t', index = False)\n",
    "#         else:\n",
    "#             variant_scores.to_csv(f'{out_file}_scores_{var_set}', sep = '\\t', index = False, header = False)\n",
    "    \n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
