{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cdfc9efa",
   "metadata": {},
   "source": [
    "# Example application of tool\n",
    "\n",
    "Let's apply the tool to a set of variants from a cancer cell line.\n",
    "\n",
    "Dataset is from [Talsania et. al. 2022](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02816-6). These are structural variants called using the [Manta pipeline](https://github.com/Illumina/manta) on WGS data from Illumina in HCC1395 tumor cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d6995e17",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "import os\n",
    "\n",
    "\n",
    "\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e22cbdbb",
   "metadata": {},
   "source": [
    "# Get input variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b09f6e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get variants\n",
    "\n",
    "ftp_path = 'https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/analysis/SVs/VCFs/tumor.illumina.manta.EA_T_1.vcf.gz'\n",
    "in_file = '../test/tumor.illumina.manta.EA_T_1.vcf'\n",
    "\n",
    "if not Path(in_file).is_file():\n",
    "    os.system(f'wget -P ../test/ {ftp_path}')\n",
    "    os.system(f'gupzip {in_file}.gz')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7193c478",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>END</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>SVLEN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1117831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>[chr22:20272153[C</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1119512</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>ACAGTGC]chr22:20302979]</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3721048</td>\n",
       "      <td>3734333</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>13285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6742482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C[chr12:96100887[</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9357666</td>\n",
       "      <td>9377061</td>\n",
       "      <td>G</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-19395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1543</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14531089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>T[chrX:6219006[</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1544</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14533586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A]chrX:6219008]</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1545</th>\n",
       "      <td>chr16_KI270728v1_random</td>\n",
       "      <td>1769992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A[chr9:129442980[</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1546</th>\n",
       "      <td>chr17_KI270729v1_random</td>\n",
       "      <td>162446</td>\n",
       "      <td>162616</td>\n",
       "      <td>GAGTCCATTCGATGATTTCATTAGATTCCATTGGAAGATGATTCCA...</td>\n",
       "      <td>G</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1547</th>\n",
       "      <td>chrUn_KI270591v1</td>\n",
       "      <td>537</td>\n",
       "      <td>878</td>\n",
       "      <td>CTGTTTCCATACTGCTCAAGAAAATAAAGGTTCAACTTTGTGGGAT...</td>\n",
       "      <td>C</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-341</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1548 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        CHROM       POS      END  \\\n",
       "0                        chr1   1117831      NaN   \n",
       "1                        chr1   1119512      NaN   \n",
       "2                        chr1   3721048  3734333   \n",
       "3                        chr1   6742482      NaN   \n",
       "4                        chr1   9357666  9377061   \n",
       "...                       ...       ...      ...   \n",
       "1543                     chrY  14531089      NaN   \n",
       "1544                     chrY  14533586      NaN   \n",
       "1545  chr16_KI270728v1_random   1769992      NaN   \n",
       "1546  chr17_KI270729v1_random    162446   162616   \n",
       "1547         chrUn_KI270591v1       537      878   \n",
       "\n",
       "                                                    REF  \\\n",
       "0                                                     C   \n",
       "1                                                     A   \n",
       "2                                                     T   \n",
       "3                                                     C   \n",
       "4                                                     G   \n",
       "...                                                 ...   \n",
       "1543                                                  T   \n",
       "1544                                                  A   \n",
       "1545                                                  A   \n",
       "1546  GAGTCCATTCGATGATTTCATTAGATTCCATTGGAAGATGATTCCA...   \n",
       "1547  CTGTTTCCATACTGCTCAAGAAAATAAAGGTTCAACTTTGTGGGAT...   \n",
       "\n",
       "                          ALT SVTYPE   SVLEN  \n",
       "0           [chr22:20272153[C    BND     NaN  \n",
       "1     ACAGTGC]chr22:20302979]    BND     NaN  \n",
       "2                <DUP:TANDEM>    DUP   13285  \n",
       "3           C[chr12:96100887[    BND     NaN  \n",
       "4                       <DEL>    DEL  -19395  \n",
       "...                       ...    ...     ...  \n",
       "1543          T[chrX:6219006[    BND     NaN  \n",
       "1544          A]chrX:6219008]    BND     NaN  \n",
       "1545        A[chr9:129442980[    BND     NaN  \n",
       "1546                        G    DEL    -170  \n",
       "1547                        C    DEL    -341  \n",
       "\n",
       "[1548 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read input\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, '../scripts')\n",
    "\n",
    "import reading_utils\n",
    "reading_utils.var_set_size = 2000\n",
    "\n",
    "variants = reading_utils.read_input(in_file, 0)\n",
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "619b9c39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DEL    552\n",
       "DUP    294\n",
       "INS     80\n",
       "Name: SVTYPE, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at varinat types present\n",
    "variants.SVTYPE.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "442002db",
   "metadata": {},
   "source": [
    "# Score variants using tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "afd652d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run this command in the main directory:\n",
      "\n",
      "python scripts/score_var.py test/tumor.illumina.manta.EA_T_1.vcf --file tumor --dir test/tumor_output --augment --get_scores\n"
     ]
    }
   ],
   "source": [
    "# Get run command\n",
    "\n",
    "file = 'tumor'#'CTCF_del' # Output files prefix\n",
    "directory = 'test/tumor_output'# '../test/output' # Output directory\n",
    "\n",
    "print('Run this command in the main directory:\\n')\n",
    "print('python scripts/score_var.py', in_file,\n",
    "      '--file', file, # File name prefix for outputs\n",
    "      '--dir', directory, # Path to save output in \n",
    "      '--augment', # Get the average augmented scores\n",
    "      '--get_scores') # Get disruption scores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d3033225",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get path to output files\n",
    "out_file = os.path.join(directory, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0275dcd9",
   "metadata": {},
   "source": [
    "# Prioritize most disruptive variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ac876712",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var_index</th>\n",
       "      <th>mse_mean</th>\n",
       "      <th>corr_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.024681</td>\n",
       "      <td>0.955341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.015164</td>\n",
       "      <td>0.973157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>0.009095</td>\n",
       "      <td>0.964207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>0.012784</td>\n",
       "      <td>0.964684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   var_index  mse_mean  corr_mean\n",
       "0          0       NaN        NaN\n",
       "1          1       NaN        NaN\n",
       "2          2  0.024681   0.955341\n",
       "3          3       NaN        NaN\n",
       "4          4  0.015164   0.973157\n",
       "5          5       NaN        NaN\n",
       "6          6       NaN        NaN\n",
       "7          7  0.009095   0.964207\n",
       "8          8  0.012784   0.964684\n",
       "9          9       NaN        NaN"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read scores output\n",
    "\n",
    "scores = pd.read_csv(f'../{out_file}_scores', sep = '\\t')\n",
    "scores.iloc[:10,:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f990d942",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>END</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>SVLEN</th>\n",
       "      <th>var_index</th>\n",
       "      <th>mse_mean</th>\n",
       "      <th>corr_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1117831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>[chr22:20272153[C</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1119512</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>ACAGTGC]chr22:20302979]</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3721048</td>\n",
       "      <td>3734333</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>13285</td>\n",
       "      <td>2</td>\n",
       "      <td>0.024681</td>\n",
       "      <td>0.955341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6742482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C[chr12:96100887[</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9357666</td>\n",
       "      <td>9377061</td>\n",
       "      <td>G</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-19395</td>\n",
       "      <td>4</td>\n",
       "      <td>0.015164</td>\n",
       "      <td>0.973157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9495567</td>\n",
       "      <td>36665592</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>27170025</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13349505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A]chr8:101508951]</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13931876</td>\n",
       "      <td>13932299</td>\n",
       "      <td>TGCCCAGGCTGGAGTGCAGTGGCACGATCTTGGCTCACTGCAACCT...</td>\n",
       "      <td>T</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-423</td>\n",
       "      <td>7</td>\n",
       "      <td>0.009095</td>\n",
       "      <td>0.964207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17393975</td>\n",
       "      <td>17394067</td>\n",
       "      <td>GATGAGATGGCCTTCTGCTACACCCAGGCTCCCCACAAGACAACGT...</td>\n",
       "      <td>G</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-92</td>\n",
       "      <td>8</td>\n",
       "      <td>0.012784</td>\n",
       "      <td>0.964684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18359587</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A]chr6:163603712]</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM       POS       END  \\\n",
       "0  chr1   1117831       NaN   \n",
       "1  chr1   1119512       NaN   \n",
       "2  chr1   3721048   3734333   \n",
       "3  chr1   6742482       NaN   \n",
       "4  chr1   9357666   9377061   \n",
       "5  chr1   9495567  36665592   \n",
       "6  chr1  13349505       NaN   \n",
       "7  chr1  13931876  13932299   \n",
       "8  chr1  17393975  17394067   \n",
       "9  chr1  18359587       NaN   \n",
       "\n",
       "                                                 REF                      ALT  \\\n",
       "0                                                  C        [chr22:20272153[C   \n",
       "1                                                  A  ACAGTGC]chr22:20302979]   \n",
       "2                                                  T             <DUP:TANDEM>   \n",
       "3                                                  C        C[chr12:96100887[   \n",
       "4                                                  G                    <DEL>   \n",
       "5                                                  T             <DUP:TANDEM>   \n",
       "6                                                  A        A]chr8:101508951]   \n",
       "7  TGCCCAGGCTGGAGTGCAGTGGCACGATCTTGGCTCACTGCAACCT...                        T   \n",
       "8  GATGAGATGGCCTTCTGCTACACCCAGGCTCCCCACAAGACAACGT...                        G   \n",
       "9                                                  A        A]chr6:163603712]   \n",
       "\n",
       "  SVTYPE     SVLEN  var_index  mse_mean  corr_mean  \n",
       "0    BND       NaN          0       NaN        NaN  \n",
       "1    BND       NaN          1       NaN        NaN  \n",
       "2    DUP     13285          2  0.024681   0.955341  \n",
       "3    BND       NaN          3       NaN        NaN  \n",
       "4    DEL    -19395          4  0.015164   0.973157  \n",
       "5    DUP  27170025          5       NaN        NaN  \n",
       "6    BND       NaN          6       NaN        NaN  \n",
       "7    DEL      -423          7  0.009095   0.964207  \n",
       "8    DEL       -92          8  0.012784   0.964684  \n",
       "9    BND       NaN          9       NaN        NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add scores dataframe to input dataframe to match them with information about the perturbations\n",
    "\n",
    "scores = pd.concat([variants, scores], axis = 1)\n",
    "scores.iloc[:10,:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2b898669",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[829, 64, 549, 25, 894, 1355, 439, 421, 1172, 397]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get top 3 most disruptive variants \n",
    "\n",
    "top_variants = []\n",
    "\n",
    "for method in ['mse_mean', 'corr_mean']:\n",
    "    \n",
    "    for SVTYPE in [x for x in scores.SVTYPE.unique() if x is not np.nan]:\n",
    "\n",
    "        top_var = list(scores[(scores.SVTYPE == SVTYPE) & (~np.isnan(scores[method]))]\n",
    "                            .sort_values(method, ascending = False)\n",
    "                            .head(3)\n",
    "                            .index\n",
    "                            .values)\n",
    "\n",
    "        for var in top_var:\n",
    "            top_variants.append(var)\n",
    "        \n",
    "top_variants[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ee45ee",
   "metadata": {},
   "source": [
    "# Get maps for top scoring variants using tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2bdb3919",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>END</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>SVLEN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>58064887</td>\n",
       "      <td>58560456</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-495569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149799777</td>\n",
       "      <td>150031617</td>\n",
       "      <td>G</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>231840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1233843</td>\n",
       "      <td>1233907</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1233870</td>\n",
       "      <td>1233870</td>\n",
       "      <td>A</td>\n",
       "      <td>AGAGGAGGCTGTCGGGGGACCGGGAGGACCTCGCCTGCTGTCCATG...</td>\n",
       "      <td>INS</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>chr6</td>\n",
       "      <td>4259990</td>\n",
       "      <td>4260069</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>chr6</td>\n",
       "      <td>17874942</td>\n",
       "      <td>17874942</td>\n",
       "      <td>T</td>\n",
       "      <td>TAATTCATTCCTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTGAG...</td>\n",
       "      <td>INS</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>chr6</td>\n",
       "      <td>23978115</td>\n",
       "      <td>23978115</td>\n",
       "      <td>G</td>\n",
       "      <td>GAAAAATGACAACCAGGCCGGGTGCGGTGGCTCACGCCTGTAATCC...</td>\n",
       "      <td>INS</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>chr6</td>\n",
       "      <td>72060264</td>\n",
       "      <td>72637247</td>\n",
       "      <td>C</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>576983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>chr6</td>\n",
       "      <td>168695904</td>\n",
       "      <td>168696064</td>\n",
       "      <td>GCCATTCAGATCATATTTCATGGAGGCCAGGTGTGCTGTGGAATGT...</td>\n",
       "      <td>G</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>829</th>\n",
       "      <td>chr10</td>\n",
       "      <td>4834359</td>\n",
       "      <td>5133665</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;DUP:TANDEM&gt;</td>\n",
       "      <td>DUP</td>\n",
       "      <td>299306</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS        END  \\\n",
       "25    chr1   58064887   58560456   \n",
       "64    chr1  149799777  150031617   \n",
       "397   chr6    1233843    1233907   \n",
       "398   chr6    1233870    1233870   \n",
       "403   chr6    4259990    4260069   \n",
       "421   chr6   17874942   17874942   \n",
       "439   chr6   23978115   23978115   \n",
       "549   chr6   72060264   72637247   \n",
       "607   chr6  168695904  168696064   \n",
       "829  chr10    4834359    5133665   \n",
       "\n",
       "                                                   REF  \\\n",
       "25                                                   T   \n",
       "64                                                   G   \n",
       "397                                                  A   \n",
       "398                                                  A   \n",
       "403                                                  A   \n",
       "421                                                  T   \n",
       "439                                                  G   \n",
       "549                                                  C   \n",
       "607  GCCATTCAGATCATATTTCATGGAGGCCAGGTGTGCTGTGGAATGT...   \n",
       "829                                                  A   \n",
       "\n",
       "                                                   ALT SVTYPE    SVLEN  \n",
       "25                                               <DEL>    DEL  -495569  \n",
       "64                                        <DUP:TANDEM>    DUP   231840  \n",
       "397                                       <DUP:TANDEM>    DUP       64  \n",
       "398  AGAGGAGGCTGTCGGGGGACCGGGAGGACCTCGCCTGCTGTCCATG...    INS       64  \n",
       "403                                       <DUP:TANDEM>    DUP       79  \n",
       "421  TAATTCATTCCTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTGAG...    INS      323  \n",
       "439  GAAAAATGACAACCAGGCCGGGTGCGGTGGCTCACGCCTGTAATCC...    INS      323  \n",
       "549                                       <DUP:TANDEM>    DUP   576983  \n",
       "607                                                  G    DEL     -160  \n",
       "829                                       <DUP:TANDEM>    DUP   299306  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants[[x in top_variants for x in variants.index]][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "73940a54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run this command in the main directory:\n",
      "\n",
      "python scripts/score_var.py test/tumor.illumina.manta.EA_T_1.vcf --file tumor_top --dir test/tumor_output --augment --get_scores get_maps get_tracks\n"
     ]
    }
   ],
   "source": [
    "# Get run command\n",
    "\n",
    "file_top = 'tumor_top'# Output files prefix\n",
    "directory_top = 'test/tumor_output'# Output directory\n",
    "\n",
    "print('Run this command in the main directory:\\n')\n",
    "print('python scripts/score_var.py', in_file,\n",
    "      '--file', file, # File name prefix for outputs\n",
    "      '--dir', directory, # Path to save output in \n",
    "      '--augment', # Get the average augmented scores\n",
    "      '--get_scores', # Get disruption scores\n",
    "      'get_maps', # Get predicted contact frequency maps\n",
    "      'get_tracks') # Get disruption tracks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0dd1354b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get path to output files\n",
    "out_file_top = os.path.join(directory_top, file_top)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea0c101d",
   "metadata": {},
   "source": [
    "# Plot maps for top scoring variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d6f160c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e4cb3f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4277c35a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9612610",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
