{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e5a7bfa",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# Packages and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "620378a9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"functional_3\"\n",
      "__________________________________________________________________________________________________\n",
      "Layer (type)                    Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      "sequence (InputLayer)           [(None, 1048576, 4)] 0                                            \n",
      "__________________________________________________________________________________________________\n",
      "stochastic_reverse_complement ( ((None, 1048576, 4), 0           sequence[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "stochastic_shift (StochasticShi (None, 1048576, 4)   0           stochastic_reverse_complement[0][\n",
      "__________________________________________________________________________________________________\n",
      "re_lu (ReLU)                    (None, 1048576, 4)   0           stochastic_shift[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d (Conv1D)                 (None, 1048576, 96)  4224        re_lu[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization (BatchNorma (None, 1048576, 96)  384         conv1d[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d (MaxPooling1D)    (None, 524288, 96)   0           batch_normalization[0][0]        \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_1 (ReLU)                  (None, 524288, 96)   0           max_pooling1d[0][0]              \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_1 (Conv1D)               (None, 524288, 96)   46080       re_lu_1[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_1 (BatchNor (None, 524288, 96)   384         conv1d_1[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_1 (MaxPooling1D)  (None, 262144, 96)   0           batch_normalization_1[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_2 (ReLU)                  (None, 262144, 96)   0           max_pooling1d_1[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_2 (Conv1D)               (None, 262144, 96)   46080       re_lu_2[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_2 (BatchNor (None, 262144, 96)   384         conv1d_2[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_2 (MaxPooling1D)  (None, 131072, 96)   0           batch_normalization_2[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_3 (ReLU)                  (None, 131072, 96)   0           max_pooling1d_2[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_3 (Conv1D)               (None, 131072, 96)   46080       re_lu_3[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_3 (BatchNor (None, 131072, 96)   384         conv1d_3[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_3 (MaxPooling1D)  (None, 65536, 96)    0           batch_normalization_3[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_4 (ReLU)                  (None, 65536, 96)    0           max_pooling1d_3[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_4 (Conv1D)               (None, 65536, 96)    46080       re_lu_4[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_4 (BatchNor (None, 65536, 96)    384         conv1d_4[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_4 (MaxPooling1D)  (None, 32768, 96)    0           batch_normalization_4[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_5 (ReLU)                  (None, 32768, 96)    0           max_pooling1d_4[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_5 (Conv1D)               (None, 32768, 96)    46080       re_lu_5[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_5 (BatchNor (None, 32768, 96)    384         conv1d_5[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_5 (MaxPooling1D)  (None, 16384, 96)    0           batch_normalization_5[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_6 (ReLU)                  (None, 16384, 96)    0           max_pooling1d_5[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_6 (Conv1D)               (None, 16384, 96)    46080       re_lu_6[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_6 (BatchNor (None, 16384, 96)    384         conv1d_6[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_6 (MaxPooling1D)  (None, 8192, 96)     0           batch_normalization_6[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_7 (ReLU)                  (None, 8192, 96)     0           max_pooling1d_6[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_7 (Conv1D)               (None, 8192, 96)     46080       re_lu_7[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_7 (BatchNor (None, 8192, 96)     384         conv1d_7[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_7 (MaxPooling1D)  (None, 4096, 96)     0           batch_normalization_7[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_8 (ReLU)                  (None, 4096, 96)     0           max_pooling1d_7[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_8 (Conv1D)               (None, 4096, 96)     46080       re_lu_8[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_8 (BatchNor (None, 4096, 96)     384         conv1d_8[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_8 (MaxPooling1D)  (None, 2048, 96)     0           batch_normalization_8[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_9 (ReLU)                  (None, 2048, 96)     0           max_pooling1d_8[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_9 (Conv1D)               (None, 2048, 96)     46080       re_lu_9[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_9 (BatchNor (None, 2048, 96)     384         conv1d_9[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_9 (MaxPooling1D)  (None, 1024, 96)     0           batch_normalization_9[0][0]      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_10 (ReLU)                 (None, 1024, 96)     0           max_pooling1d_9[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_10 (Conv1D)              (None, 1024, 96)     46080       re_lu_10[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_10 (BatchNo (None, 1024, 96)     384         conv1d_10[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_10 (MaxPooling1D) (None, 512, 96)      0           batch_normalization_10[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_11 (ReLU)                 (None, 512, 96)      0           max_pooling1d_10[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_11 (Conv1D)              (None, 512, 48)      13824       re_lu_11[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_11 (BatchNo (None, 512, 48)      192         conv1d_11[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_12 (ReLU)                 (None, 512, 48)      0           batch_normalization_11[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_12 (Conv1D)              (None, 512, 96)      4608        re_lu_12[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_12 (BatchNo (None, 512, 96)      384         conv1d_12[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout (Dropout)               (None, 512, 96)      0           batch_normalization_12[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add (Add)                       (None, 512, 96)      0           max_pooling1d_10[0][0]           \n",
      "                                                                 dropout[0][0]                    \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_13 (ReLU)                 (None, 512, 96)      0           add[0][0]                        \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_13 (Conv1D)              (None, 512, 48)      13824       re_lu_13[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_13 (BatchNo (None, 512, 48)      192         conv1d_13[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_14 (ReLU)                 (None, 512, 48)      0           batch_normalization_13[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_14 (Conv1D)              (None, 512, 96)      4608        re_lu_14[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_14 (BatchNo (None, 512, 96)      384         conv1d_14[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_1 (Dropout)             (None, 512, 96)      0           batch_normalization_14[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_1 (Add)                     (None, 512, 96)      0           add[0][0]                        \n",
      "                                                                 dropout_1[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_15 (ReLU)                 (None, 512, 96)      0           add_1[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_15 (Conv1D)              (None, 512, 48)      13824       re_lu_15[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_15 (BatchNo (None, 512, 48)      192         conv1d_15[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_16 (ReLU)                 (None, 512, 48)      0           batch_normalization_15[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_16 (Conv1D)              (None, 512, 96)      4608        re_lu_16[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_16 (BatchNo (None, 512, 96)      384         conv1d_16[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_2 (Dropout)             (None, 512, 96)      0           batch_normalization_16[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_2 (Add)                     (None, 512, 96)      0           add_1[0][0]                      \n",
      "                                                                 dropout_2[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_17 (ReLU)                 (None, 512, 96)      0           add_2[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_17 (Conv1D)              (None, 512, 48)      13824       re_lu_17[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_17 (BatchNo (None, 512, 48)      192         conv1d_17[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_18 (ReLU)                 (None, 512, 48)      0           batch_normalization_17[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_18 (Conv1D)              (None, 512, 96)      4608        re_lu_18[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_18 (BatchNo (None, 512, 96)      384         conv1d_18[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_3 (Dropout)             (None, 512, 96)      0           batch_normalization_18[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_3 (Add)                     (None, 512, 96)      0           add_2[0][0]                      \n",
      "                                                                 dropout_3[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_19 (ReLU)                 (None, 512, 96)      0           add_3[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_19 (Conv1D)              (None, 512, 48)      13824       re_lu_19[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_19 (BatchNo (None, 512, 48)      192         conv1d_19[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_20 (ReLU)                 (None, 512, 48)      0           batch_normalization_19[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_20 (Conv1D)              (None, 512, 96)      4608        re_lu_20[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_20 (BatchNo (None, 512, 96)      384         conv1d_20[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_4 (Dropout)             (None, 512, 96)      0           batch_normalization_20[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_4 (Add)                     (None, 512, 96)      0           add_3[0][0]                      \n",
      "                                                                 dropout_4[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_21 (ReLU)                 (None, 512, 96)      0           add_4[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_21 (Conv1D)              (None, 512, 48)      13824       re_lu_21[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_21 (BatchNo (None, 512, 48)      192         conv1d_21[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_22 (ReLU)                 (None, 512, 48)      0           batch_normalization_21[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_22 (Conv1D)              (None, 512, 96)      4608        re_lu_22[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_22 (BatchNo (None, 512, 96)      384         conv1d_22[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_5 (Dropout)             (None, 512, 96)      0           batch_normalization_22[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_5 (Add)                     (None, 512, 96)      0           add_4[0][0]                      \n",
      "                                                                 dropout_5[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_23 (ReLU)                 (None, 512, 96)      0           add_5[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_23 (Conv1D)              (None, 512, 48)      13824       re_lu_23[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_23 (BatchNo (None, 512, 48)      192         conv1d_23[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_24 (ReLU)                 (None, 512, 48)      0           batch_normalization_23[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_24 (Conv1D)              (None, 512, 96)      4608        re_lu_24[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_24 (BatchNo (None, 512, 96)      384         conv1d_24[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_6 (Dropout)             (None, 512, 96)      0           batch_normalization_24[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_6 (Add)                     (None, 512, 96)      0           add_5[0][0]                      \n",
      "                                                                 dropout_6[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_25 (ReLU)                 (None, 512, 96)      0           add_6[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_25 (Conv1D)              (None, 512, 48)      13824       re_lu_25[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_25 (BatchNo (None, 512, 48)      192         conv1d_25[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_26 (ReLU)                 (None, 512, 48)      0           batch_normalization_25[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_26 (Conv1D)              (None, 512, 96)      4608        re_lu_26[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_26 (BatchNo (None, 512, 96)      384         conv1d_26[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_7 (Dropout)             (None, 512, 96)      0           batch_normalization_26[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_7 (Add)                     (None, 512, 96)      0           add_6[0][0]                      \n",
      "                                                                 dropout_7[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_27 (ReLU)                 (None, 512, 96)      0           add_7[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_27 (Conv1D)              (None, 512, 64)      30720       re_lu_27[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_27 (BatchNo (None, 512, 64)      256         conv1d_27[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_28 (ReLU)                 (None, 512, 64)      0           batch_normalization_27[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "one_to_two (OneToTwo)           (None, 512, 512, 64) 0           re_lu_28[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "concat_dist2d (ConcatDist2D)    (None, 512, 512, 65) 0           one_to_two[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_29 (ReLU)                 (None, 512, 512, 65) 0           concat_dist2d[0][0]              \n",
      "__________________________________________________________________________________________________\n",
      "conv2d (Conv2D)                 (None, 512, 512, 48) 28080       re_lu_29[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_28 (BatchNo (None, 512, 512, 48) 192         conv2d[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d (Symmetrize2D)     (None, 512, 512, 48) 0           batch_normalization_28[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_30 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d[0][0]               \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_1 (Conv2D)               (None, 512, 512, 24) 10368       re_lu_30[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_29 (BatchNo (None, 512, 512, 24) 96          conv2d_1[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_31 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_29[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_2 (Conv2D)               (None, 512, 512, 48) 1152        re_lu_31[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_30 (BatchNo (None, 512, 512, 48) 192         conv2d_2[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "dropout_8 (Dropout)             (None, 512, 512, 48) 0           batch_normalization_30[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_8 (Add)                     (None, 512, 512, 48) 0           symmetrize2d[0][0]               \n",
      "                                                                 dropout_8[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_1 (Symmetrize2D)   (None, 512, 512, 48) 0           add_8[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_32 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_1[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_3 (Conv2D)               (None, 512, 512, 24) 10368       re_lu_32[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_31 (BatchNo (None, 512, 512, 24) 96          conv2d_3[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_33 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_31[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_4 (Conv2D)               (None, 512, 512, 48) 1152        re_lu_33[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_32 (BatchNo (None, 512, 512, 48) 192         conv2d_4[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "dropout_9 (Dropout)             (None, 512, 512, 48) 0           batch_normalization_32[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_9 (Add)                     (None, 512, 512, 48) 0           symmetrize2d_1[0][0]             \n",
      "                                                                 dropout_9[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_2 (Symmetrize2D)   (None, 512, 512, 48) 0           add_9[0][0]                      \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_34 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_2[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_5 (Conv2D)               (None, 512, 512, 24) 10368       re_lu_34[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_33 (BatchNo (None, 512, 512, 24) 96          conv2d_5[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_35 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_33[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_6 (Conv2D)               (None, 512, 512, 48) 1152        re_lu_35[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_34 (BatchNo (None, 512, 512, 48) 192         conv2d_6[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "dropout_10 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_34[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_10 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_2[0][0]             \n",
      "                                                                 dropout_10[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_3 (Symmetrize2D)   (None, 512, 512, 48) 0           add_10[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_36 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_3[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_7 (Conv2D)               (None, 512, 512, 24) 10368       re_lu_36[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_35 (BatchNo (None, 512, 512, 24) 96          conv2d_7[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_37 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_35[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_8 (Conv2D)               (None, 512, 512, 48) 1152        re_lu_37[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_36 (BatchNo (None, 512, 512, 48) 192         conv2d_8[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "dropout_11 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_36[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_11 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_3[0][0]             \n",
      "                                                                 dropout_11[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_4 (Symmetrize2D)   (None, 512, 512, 48) 0           add_11[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_38 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_4[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_9 (Conv2D)               (None, 512, 512, 24) 10368       re_lu_38[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_37 (BatchNo (None, 512, 512, 24) 96          conv2d_9[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_39 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_37[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_10 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_39[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_38 (BatchNo (None, 512, 512, 48) 192         conv2d_10[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_12 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_38[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_12 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_4[0][0]             \n",
      "                                                                 dropout_12[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_5 (Symmetrize2D)   (None, 512, 512, 48) 0           add_12[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_40 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_5[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_11 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_40[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_39 (BatchNo (None, 512, 512, 24) 96          conv2d_11[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_41 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_39[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_12 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_41[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_40 (BatchNo (None, 512, 512, 48) 192         conv2d_12[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_13 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_40[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_13 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_5[0][0]             \n",
      "                                                                 dropout_13[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_6 (Symmetrize2D)   (None, 512, 512, 48) 0           add_13[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "cropping2d (Cropping2D)         (None, 448, 448, 48) 0           symmetrize2d_6[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "upper_tri (UpperTri)            (None, 99681, 48)    0           cropping2d[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "dense (Dense)                   (None, 99681, 5)     245         upper_tri[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "switch_reverse_triu (SwitchReve (None, 99681, 5)     0           dense[0][0]                      \n",
      "                                                                 stochastic_reverse_complement[0][\n",
      "==================================================================================================\n",
      "Total params: 751,653\n",
      "Trainable params: 746,149\n",
      "Non-trainable params: 5,504\n",
      "__________________________________________________________________________________________________\n",
      "None\n",
      "model_strides [2048]\n",
      "target_lengths [99681]\n",
      "target_crops [-49585]\n",
      "successfully loaded\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"functional_7\"\n",
      "__________________________________________________________________________________________________\n",
      "Layer (type)                    Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      "sequence (InputLayer)           [(None, 1048576, 4)] 0                                            \n",
      "__________________________________________________________________________________________________\n",
      "stochastic_reverse_complement_1 ((None, 1048576, 4), 0           sequence[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "stochastic_shift_1 (StochasticS (None, 1048576, 4)   0           stochastic_reverse_complement_1[0\n",
      "__________________________________________________________________________________________________\n",
      "re_lu_42 (ReLU)                 (None, 1048576, 4)   0           stochastic_shift_1[0][0]         \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_28 (Conv1D)              (None, 1048576, 96)  4224        re_lu_42[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_41 (BatchNo (None, 1048576, 96)  384         conv1d_28[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_11 (MaxPooling1D) (None, 524288, 96)   0           batch_normalization_41[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_43 (ReLU)                 (None, 524288, 96)   0           max_pooling1d_11[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_29 (Conv1D)              (None, 524288, 96)   46080       re_lu_43[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_42 (BatchNo (None, 524288, 96)   384         conv1d_29[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_12 (MaxPooling1D) (None, 262144, 96)   0           batch_normalization_42[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_44 (ReLU)                 (None, 262144, 96)   0           max_pooling1d_12[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_30 (Conv1D)              (None, 262144, 96)   46080       re_lu_44[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_43 (BatchNo (None, 262144, 96)   384         conv1d_30[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_13 (MaxPooling1D) (None, 131072, 96)   0           batch_normalization_43[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_45 (ReLU)                 (None, 131072, 96)   0           max_pooling1d_13[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_31 (Conv1D)              (None, 131072, 96)   46080       re_lu_45[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_44 (BatchNo (None, 131072, 96)   384         conv1d_31[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_14 (MaxPooling1D) (None, 65536, 96)    0           batch_normalization_44[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_46 (ReLU)                 (None, 65536, 96)    0           max_pooling1d_14[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_32 (Conv1D)              (None, 65536, 96)    46080       re_lu_46[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_45 (BatchNo (None, 65536, 96)    384         conv1d_32[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_15 (MaxPooling1D) (None, 32768, 96)    0           batch_normalization_45[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_47 (ReLU)                 (None, 32768, 96)    0           max_pooling1d_15[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_33 (Conv1D)              (None, 32768, 96)    46080       re_lu_47[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_46 (BatchNo (None, 32768, 96)    384         conv1d_33[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_16 (MaxPooling1D) (None, 16384, 96)    0           batch_normalization_46[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_48 (ReLU)                 (None, 16384, 96)    0           max_pooling1d_16[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_34 (Conv1D)              (None, 16384, 96)    46080       re_lu_48[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_47 (BatchNo (None, 16384, 96)    384         conv1d_34[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_17 (MaxPooling1D) (None, 8192, 96)     0           batch_normalization_47[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_49 (ReLU)                 (None, 8192, 96)     0           max_pooling1d_17[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_35 (Conv1D)              (None, 8192, 96)     46080       re_lu_49[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_48 (BatchNo (None, 8192, 96)     384         conv1d_35[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_18 (MaxPooling1D) (None, 4096, 96)     0           batch_normalization_48[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_50 (ReLU)                 (None, 4096, 96)     0           max_pooling1d_18[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_36 (Conv1D)              (None, 4096, 96)     46080       re_lu_50[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_49 (BatchNo (None, 4096, 96)     384         conv1d_36[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_19 (MaxPooling1D) (None, 2048, 96)     0           batch_normalization_49[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_51 (ReLU)                 (None, 2048, 96)     0           max_pooling1d_19[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_37 (Conv1D)              (None, 2048, 96)     46080       re_lu_51[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_50 (BatchNo (None, 2048, 96)     384         conv1d_37[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_20 (MaxPooling1D) (None, 1024, 96)     0           batch_normalization_50[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_52 (ReLU)                 (None, 1024, 96)     0           max_pooling1d_20[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_38 (Conv1D)              (None, 1024, 96)     46080       re_lu_52[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_51 (BatchNo (None, 1024, 96)     384         conv1d_38[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_21 (MaxPooling1D) (None, 512, 96)      0           batch_normalization_51[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_53 (ReLU)                 (None, 512, 96)      0           max_pooling1d_21[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_39 (Conv1D)              (None, 512, 48)      13824       re_lu_53[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_52 (BatchNo (None, 512, 48)      192         conv1d_39[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_54 (ReLU)                 (None, 512, 48)      0           batch_normalization_52[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_40 (Conv1D)              (None, 512, 96)      4608        re_lu_54[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_53 (BatchNo (None, 512, 96)      384         conv1d_40[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_14 (Dropout)            (None, 512, 96)      0           batch_normalization_53[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_14 (Add)                    (None, 512, 96)      0           max_pooling1d_21[0][0]           \n",
      "                                                                 dropout_14[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_55 (ReLU)                 (None, 512, 96)      0           add_14[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_41 (Conv1D)              (None, 512, 48)      13824       re_lu_55[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_54 (BatchNo (None, 512, 48)      192         conv1d_41[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_56 (ReLU)                 (None, 512, 48)      0           batch_normalization_54[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_42 (Conv1D)              (None, 512, 96)      4608        re_lu_56[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_55 (BatchNo (None, 512, 96)      384         conv1d_42[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_15 (Dropout)            (None, 512, 96)      0           batch_normalization_55[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_15 (Add)                    (None, 512, 96)      0           add_14[0][0]                     \n",
      "                                                                 dropout_15[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_57 (ReLU)                 (None, 512, 96)      0           add_15[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_43 (Conv1D)              (None, 512, 48)      13824       re_lu_57[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_56 (BatchNo (None, 512, 48)      192         conv1d_43[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_58 (ReLU)                 (None, 512, 48)      0           batch_normalization_56[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_44 (Conv1D)              (None, 512, 96)      4608        re_lu_58[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_57 (BatchNo (None, 512, 96)      384         conv1d_44[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_16 (Dropout)            (None, 512, 96)      0           batch_normalization_57[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_16 (Add)                    (None, 512, 96)      0           add_15[0][0]                     \n",
      "                                                                 dropout_16[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_59 (ReLU)                 (None, 512, 96)      0           add_16[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_45 (Conv1D)              (None, 512, 48)      13824       re_lu_59[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_58 (BatchNo (None, 512, 48)      192         conv1d_45[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_60 (ReLU)                 (None, 512, 48)      0           batch_normalization_58[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_46 (Conv1D)              (None, 512, 96)      4608        re_lu_60[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_59 (BatchNo (None, 512, 96)      384         conv1d_46[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_17 (Dropout)            (None, 512, 96)      0           batch_normalization_59[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_17 (Add)                    (None, 512, 96)      0           add_16[0][0]                     \n",
      "                                                                 dropout_17[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_61 (ReLU)                 (None, 512, 96)      0           add_17[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_47 (Conv1D)              (None, 512, 48)      13824       re_lu_61[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_60 (BatchNo (None, 512, 48)      192         conv1d_47[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_62 (ReLU)                 (None, 512, 48)      0           batch_normalization_60[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_48 (Conv1D)              (None, 512, 96)      4608        re_lu_62[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_61 (BatchNo (None, 512, 96)      384         conv1d_48[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_18 (Dropout)            (None, 512, 96)      0           batch_normalization_61[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_18 (Add)                    (None, 512, 96)      0           add_17[0][0]                     \n",
      "                                                                 dropout_18[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_63 (ReLU)                 (None, 512, 96)      0           add_18[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_49 (Conv1D)              (None, 512, 48)      13824       re_lu_63[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_62 (BatchNo (None, 512, 48)      192         conv1d_49[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_64 (ReLU)                 (None, 512, 48)      0           batch_normalization_62[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_50 (Conv1D)              (None, 512, 96)      4608        re_lu_64[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_63 (BatchNo (None, 512, 96)      384         conv1d_50[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_19 (Dropout)            (None, 512, 96)      0           batch_normalization_63[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_19 (Add)                    (None, 512, 96)      0           add_18[0][0]                     \n",
      "                                                                 dropout_19[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_65 (ReLU)                 (None, 512, 96)      0           add_19[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_51 (Conv1D)              (None, 512, 48)      13824       re_lu_65[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_64 (BatchNo (None, 512, 48)      192         conv1d_51[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_66 (ReLU)                 (None, 512, 48)      0           batch_normalization_64[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_52 (Conv1D)              (None, 512, 96)      4608        re_lu_66[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_65 (BatchNo (None, 512, 96)      384         conv1d_52[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_20 (Dropout)            (None, 512, 96)      0           batch_normalization_65[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_20 (Add)                    (None, 512, 96)      0           add_19[0][0]                     \n",
      "                                                                 dropout_20[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_67 (ReLU)                 (None, 512, 96)      0           add_20[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_53 (Conv1D)              (None, 512, 48)      13824       re_lu_67[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_66 (BatchNo (None, 512, 48)      192         conv1d_53[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_68 (ReLU)                 (None, 512, 48)      0           batch_normalization_66[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_54 (Conv1D)              (None, 512, 96)      4608        re_lu_68[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_67 (BatchNo (None, 512, 96)      384         conv1d_54[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_21 (Dropout)            (None, 512, 96)      0           batch_normalization_67[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_21 (Add)                    (None, 512, 96)      0           add_20[0][0]                     \n",
      "                                                                 dropout_21[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_69 (ReLU)                 (None, 512, 96)      0           add_21[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_55 (Conv1D)              (None, 512, 64)      30720       re_lu_69[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_68 (BatchNo (None, 512, 64)      256         conv1d_55[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_70 (ReLU)                 (None, 512, 64)      0           batch_normalization_68[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "one_to_two_1 (OneToTwo)         (None, 512, 512, 64) 0           re_lu_70[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "concat_dist2d_1 (ConcatDist2D)  (None, 512, 512, 65) 0           one_to_two_1[0][0]               \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_71 (ReLU)                 (None, 512, 512, 65) 0           concat_dist2d_1[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_13 (Conv2D)              (None, 512, 512, 48) 28080       re_lu_71[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_69 (BatchNo (None, 512, 512, 48) 192         conv2d_13[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_7 (Symmetrize2D)   (None, 512, 512, 48) 0           batch_normalization_69[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_72 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_7[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_14 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_72[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_70 (BatchNo (None, 512, 512, 24) 96          conv2d_14[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_73 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_70[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_15 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_73[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_71 (BatchNo (None, 512, 512, 48) 192         conv2d_15[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_22 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_71[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_22 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_7[0][0]             \n",
      "                                                                 dropout_22[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_8 (Symmetrize2D)   (None, 512, 512, 48) 0           add_22[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_74 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_8[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_16 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_74[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_72 (BatchNo (None, 512, 512, 24) 96          conv2d_16[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_75 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_72[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_17 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_75[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_73 (BatchNo (None, 512, 512, 48) 192         conv2d_17[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_23 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_73[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_23 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_8[0][0]             \n",
      "                                                                 dropout_23[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_9 (Symmetrize2D)   (None, 512, 512, 48) 0           add_23[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_76 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_9[0][0]             \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_18 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_76[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_74 (BatchNo (None, 512, 512, 24) 96          conv2d_18[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_77 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_74[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_19 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_77[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_75 (BatchNo (None, 512, 512, 48) 192         conv2d_19[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_24 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_75[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_24 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_9[0][0]             \n",
      "                                                                 dropout_24[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_10 (Symmetrize2D)  (None, 512, 512, 48) 0           add_24[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_78 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_10[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_20 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_78[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_76 (BatchNo (None, 512, 512, 24) 96          conv2d_20[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_79 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_76[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_21 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_79[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_77 (BatchNo (None, 512, 512, 48) 192         conv2d_21[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_25 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_77[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_25 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_10[0][0]            \n",
      "                                                                 dropout_25[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_11 (Symmetrize2D)  (None, 512, 512, 48) 0           add_25[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_80 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_11[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_22 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_80[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_78 (BatchNo (None, 512, 512, 24) 96          conv2d_22[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_81 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_78[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_23 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_81[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_79 (BatchNo (None, 512, 512, 48) 192         conv2d_23[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_26 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_79[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_26 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_11[0][0]            \n",
      "                                                                 dropout_26[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_12 (Symmetrize2D)  (None, 512, 512, 48) 0           add_26[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_82 (ReLU)                 (None, 512, 512, 48) 0           symmetrize2d_12[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_24 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_82[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_80 (BatchNo (None, 512, 512, 24) 96          conv2d_24[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_83 (ReLU)                 (None, 512, 512, 24) 0           batch_normalization_80[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_25 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_83[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_81 (BatchNo (None, 512, 512, 48) 192         conv2d_25[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_27 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_81[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_27 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_12[0][0]            \n",
      "                                                                 dropout_27[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_13 (Symmetrize2D)  (None, 512, 512, 48) 0           add_27[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "cropping2d_1 (Cropping2D)       (None, 448, 448, 48) 0           symmetrize2d_13[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "upper_tri_1 (UpperTri)          (None, 99681, 48)    0           cropping2d_1[0][0]               \n",
      "__________________________________________________________________________________________________\n",
      "dense_1 (Dense)                 (None, 99681, 5)     245         upper_tri_1[0][0]                \n",
      "__________________________________________________________________________________________________\n",
      "switch_reverse_triu_1 (SwitchRe (None, 99681, 5)     0           dense_1[0][0]                    \n",
      "                                                                 stochastic_reverse_complement_1[0\n",
      "==================================================================================================\n",
      "Total params: 751,653\n",
      "Trainable params: 746,149\n",
      "Non-trainable params: 5,504\n",
      "__________________________________________________________________________________________________\n",
      "None\n",
      "model_strides [2048]\n",
      "target_lengths [99681]\n",
      "target_crops [-49585]\n",
      "Akita successfully loaded\n"
     ]
    }
   ],
   "source": [
    "data_dir = '/pollard/data/projects/kgjoni/'\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, f\"{data_dir}Akita/akita_imaging/\")\n",
    "import pred_utils_1mb_KG as pred\n",
    "\n",
    "sys.path.insert(0, f'{data_dir}Akita_variant_scoring/scripts/')\n",
    "import OLD_utils as utils\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import os\n",
    "\n",
    "half_patch_size = 2**19\n",
    "MB = 2*half_patch_size\n",
    "pixel_size = 2048\n",
    "bins = 448\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a57ac4aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import io\n",
    "\n",
    "def read_vcf(path):\n",
    "    \n",
    "    '''\n",
    "    Read  vcf files into dataframe.\n",
    "    Source: https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "        \n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def read_vcf_gz(path):\n",
    "    \n",
    "    '''\n",
    "    Read  gzipped vcf files into dataframe.\n",
    "    Adapted from: https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    with io.TextIOWrapper(gzip.open(path,'r')) as f:\n",
    "        lines =[l for l in f if not l.startswith('##')]\n",
    "\n",
    "    return pd.read_csv(\n",
    "            io.StringIO(''.join(lines)),\n",
    "            dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                   'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "            sep='\\t'\n",
    "        ).rename(columns={'#CHROM':'CHROM'})\n",
    "\n",
    "\n",
    "\n",
    "def read_input(in_file):\n",
    "\n",
    "    '''\n",
    "    Read and reformat variant dataset. Accepted formats are .vcf .vcf.gz from 4.1 version, \n",
    "    .bed file with the following columns: [CHROM, POS, REF, ALT, END, SVTYPE, SVLEN], \n",
    "    and .tsv from ANNOVAR annotSV.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    \n",
    "    if 'vcf' in in_file:\n",
    "        \n",
    "        # For gzipped files\n",
    "        if in_file.endswith('.gz'):\n",
    "            variants = read_vcf_gz(in_file)\n",
    "        else:\n",
    "            variants = read_vcf(in_file)\n",
    "            \n",
    "        # Read SVs\n",
    "        if any(['SVTYPE' in x for x in variants.INFO]):\n",
    "            \n",
    "            variants['END'] = variants.INFO.str.split('END=').str[1].str.split(';').str[0] # this SVLEN (END-POS) would be 0 for SNPs\n",
    "            variants.loc[~pd.isnull(variants.END), 'END'] = variants.loc[~pd.isnull(variants.END), 'END'].astype('int')\n",
    "            variants['SVTYPE'] = variants.INFO.str.split('SVTYPE=').str[1].str.split(';').str[0]\n",
    "            variants['SVLEN'] = variants.INFO.str.split('SVLEN=').str[1].str.split(';').str[0]\n",
    "            variants = variants[['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']]\n",
    "\n",
    "        # Read simple variants \n",
    "        else:\n",
    "            \n",
    "            variants = variants[['CHROM', 'POS', 'REF', 'ALT']]       \n",
    "            \n",
    " \n",
    "        \n",
    "    elif 'bed' in in_file:\n",
    "        \n",
    "        colnames = ['CHROM', 'POS', 'REF', 'ALT', 'END', 'SVTYPE', 'SVLEN']\n",
    "        ncols = len(pd.read_csv(in_file, sep = '\\t', nrows = 0, low_memory=False).columns)\n",
    "\n",
    "        variants = pd.read_csv(in_file, sep = '\\t', names = colnames[:ncols], low_memory=False)\n",
    "        \n",
    "        \n",
    "    elif 'tsv' in in_file:\n",
    "        \n",
    "        with (gzip.open if in_file.endswith(\".gz\") else open)(in_file, \"rt\", encoding=\"utf-8\") as variants:\n",
    "            variants = (pd.read_csv(in_file, sep = '\\t', low_memory=False)\n",
    "                        .rename(columns = {'SV_chrom':'CHROM', \n",
    "                                           'SV_start':'POS',\n",
    "                                           'SV_end':'END', \n",
    "                                           'SV_type':'SVTYPE',\n",
    "                                           'SV_length':'SVLEN'})\n",
    "                       [['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']])\n",
    "            variants['CHROM'] = ['chr' + str(x) for x in variants['CHROM']]\n",
    "            variants.loc[~pd.isnull(variants.END), 'END'] = variants.loc[~pd.isnull(variants.END), 'END'].astype('int')\n",
    "\n",
    "            \n",
    "    else:\n",
    "        raise ValueError('Input file type not accepted. Make sure it has the right extension.')\n",
    "        \n",
    "        \n",
    "    variants.reset_index(inplace = True, drop = True)\n",
    "    \n",
    "    \n",
    "    return variants\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5c215d3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"functional_11\"\n",
      "__________________________________________________________________________________________________\n",
      "Layer (type)                    Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      "sequence (InputLayer)           [(None, 1048576, 4)] 0                                            \n",
      "__________________________________________________________________________________________________\n",
      "stochastic_reverse_complement_2 ((None, 1048576, 4), 0           sequence[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "stochastic_shift_2 (StochasticS (None, 1048576, 4)   0           stochastic_reverse_complement_2[0\n",
      "__________________________________________________________________________________________________\n",
      "re_lu_84 (ReLU)                 (None, 1048576, 4)   0           stochastic_shift_2[0][0]         \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_56 (Conv1D)              (None, 1048576, 96)  4224        re_lu_84[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_82 (BatchNo (None, 1048576, 96)  384         conv1d_56[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_22 (MaxPooling1D) (None, 524288, 96)   0           batch_normalization_82[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_85 (ReLU)                 (None, 524288, 96)   0           max_pooling1d_22[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_57 (Conv1D)              (None, 524288, 96)   46080       re_lu_85[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_83 (BatchNo (None, 524288, 96)   384         conv1d_57[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_23 (MaxPooling1D) (None, 262144, 96)   0           batch_normalization_83[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_86 (ReLU)                 (None, 262144, 96)   0           max_pooling1d_23[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_58 (Conv1D)              (None, 262144, 96)   46080       re_lu_86[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_84 (BatchNo (None, 262144, 96)   384         conv1d_58[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_24 (MaxPooling1D) (None, 131072, 96)   0           batch_normalization_84[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_87 (ReLU)                 (None, 131072, 96)   0           max_pooling1d_24[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_59 (Conv1D)              (None, 131072, 96)   46080       re_lu_87[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_85 (BatchNo (None, 131072, 96)   384         conv1d_59[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_25 (MaxPooling1D) (None, 65536, 96)    0           batch_normalization_85[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_88 (ReLU)                 (None, 65536, 96)    0           max_pooling1d_25[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_60 (Conv1D)              (None, 65536, 96)    46080       re_lu_88[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_86 (BatchNo (None, 65536, 96)    384         conv1d_60[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_26 (MaxPooling1D) (None, 32768, 96)    0           batch_normalization_86[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_89 (ReLU)                 (None, 32768, 96)    0           max_pooling1d_26[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_61 (Conv1D)              (None, 32768, 96)    46080       re_lu_89[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_87 (BatchNo (None, 32768, 96)    384         conv1d_61[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_27 (MaxPooling1D) (None, 16384, 96)    0           batch_normalization_87[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_90 (ReLU)                 (None, 16384, 96)    0           max_pooling1d_27[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_62 (Conv1D)              (None, 16384, 96)    46080       re_lu_90[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_88 (BatchNo (None, 16384, 96)    384         conv1d_62[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_28 (MaxPooling1D) (None, 8192, 96)     0           batch_normalization_88[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_91 (ReLU)                 (None, 8192, 96)     0           max_pooling1d_28[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_63 (Conv1D)              (None, 8192, 96)     46080       re_lu_91[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_89 (BatchNo (None, 8192, 96)     384         conv1d_63[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_29 (MaxPooling1D) (None, 4096, 96)     0           batch_normalization_89[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_92 (ReLU)                 (None, 4096, 96)     0           max_pooling1d_29[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_64 (Conv1D)              (None, 4096, 96)     46080       re_lu_92[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_90 (BatchNo (None, 4096, 96)     384         conv1d_64[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_30 (MaxPooling1D) (None, 2048, 96)     0           batch_normalization_90[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_93 (ReLU)                 (None, 2048, 96)     0           max_pooling1d_30[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_65 (Conv1D)              (None, 2048, 96)     46080       re_lu_93[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_91 (BatchNo (None, 2048, 96)     384         conv1d_65[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_31 (MaxPooling1D) (None, 1024, 96)     0           batch_normalization_91[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_94 (ReLU)                 (None, 1024, 96)     0           max_pooling1d_31[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_66 (Conv1D)              (None, 1024, 96)     46080       re_lu_94[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_92 (BatchNo (None, 1024, 96)     384         conv1d_66[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "max_pooling1d_32 (MaxPooling1D) (None, 512, 96)      0           batch_normalization_92[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_95 (ReLU)                 (None, 512, 96)      0           max_pooling1d_32[0][0]           \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_67 (Conv1D)              (None, 512, 48)      13824       re_lu_95[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_93 (BatchNo (None, 512, 48)      192         conv1d_67[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_96 (ReLU)                 (None, 512, 48)      0           batch_normalization_93[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_68 (Conv1D)              (None, 512, 96)      4608        re_lu_96[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_94 (BatchNo (None, 512, 96)      384         conv1d_68[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_28 (Dropout)            (None, 512, 96)      0           batch_normalization_94[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_28 (Add)                    (None, 512, 96)      0           max_pooling1d_32[0][0]           \n",
      "                                                                 dropout_28[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_97 (ReLU)                 (None, 512, 96)      0           add_28[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_69 (Conv1D)              (None, 512, 48)      13824       re_lu_97[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_95 (BatchNo (None, 512, 48)      192         conv1d_69[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_98 (ReLU)                 (None, 512, 48)      0           batch_normalization_95[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_70 (Conv1D)              (None, 512, 96)      4608        re_lu_98[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_96 (BatchNo (None, 512, 96)      384         conv1d_70[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_29 (Dropout)            (None, 512, 96)      0           batch_normalization_96[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_29 (Add)                    (None, 512, 96)      0           add_28[0][0]                     \n",
      "                                                                 dropout_29[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_99 (ReLU)                 (None, 512, 96)      0           add_29[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_71 (Conv1D)              (None, 512, 48)      13824       re_lu_99[0][0]                   \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_97 (BatchNo (None, 512, 48)      192         conv1d_71[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_100 (ReLU)                (None, 512, 48)      0           batch_normalization_97[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_72 (Conv1D)              (None, 512, 96)      4608        re_lu_100[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_98 (BatchNo (None, 512, 96)      384         conv1d_72[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_30 (Dropout)            (None, 512, 96)      0           batch_normalization_98[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "add_30 (Add)                    (None, 512, 96)      0           add_29[0][0]                     \n",
      "                                                                 dropout_30[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_101 (ReLU)                (None, 512, 96)      0           add_30[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_73 (Conv1D)              (None, 512, 48)      13824       re_lu_101[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_99 (BatchNo (None, 512, 48)      192         conv1d_73[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_102 (ReLU)                (None, 512, 48)      0           batch_normalization_99[0][0]     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_74 (Conv1D)              (None, 512, 96)      4608        re_lu_102[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_100 (BatchN (None, 512, 96)      384         conv1d_74[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_31 (Dropout)            (None, 512, 96)      0           batch_normalization_100[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_31 (Add)                    (None, 512, 96)      0           add_30[0][0]                     \n",
      "                                                                 dropout_31[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_103 (ReLU)                (None, 512, 96)      0           add_31[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_75 (Conv1D)              (None, 512, 48)      13824       re_lu_103[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_101 (BatchN (None, 512, 48)      192         conv1d_75[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_104 (ReLU)                (None, 512, 48)      0           batch_normalization_101[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_76 (Conv1D)              (None, 512, 96)      4608        re_lu_104[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_102 (BatchN (None, 512, 96)      384         conv1d_76[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_32 (Dropout)            (None, 512, 96)      0           batch_normalization_102[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_32 (Add)                    (None, 512, 96)      0           add_31[0][0]                     \n",
      "                                                                 dropout_32[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_105 (ReLU)                (None, 512, 96)      0           add_32[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_77 (Conv1D)              (None, 512, 48)      13824       re_lu_105[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_103 (BatchN (None, 512, 48)      192         conv1d_77[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_106 (ReLU)                (None, 512, 48)      0           batch_normalization_103[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_78 (Conv1D)              (None, 512, 96)      4608        re_lu_106[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_104 (BatchN (None, 512, 96)      384         conv1d_78[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_33 (Dropout)            (None, 512, 96)      0           batch_normalization_104[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_33 (Add)                    (None, 512, 96)      0           add_32[0][0]                     \n",
      "                                                                 dropout_33[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_107 (ReLU)                (None, 512, 96)      0           add_33[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_79 (Conv1D)              (None, 512, 48)      13824       re_lu_107[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_105 (BatchN (None, 512, 48)      192         conv1d_79[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_108 (ReLU)                (None, 512, 48)      0           batch_normalization_105[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_80 (Conv1D)              (None, 512, 96)      4608        re_lu_108[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_106 (BatchN (None, 512, 96)      384         conv1d_80[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_34 (Dropout)            (None, 512, 96)      0           batch_normalization_106[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_34 (Add)                    (None, 512, 96)      0           add_33[0][0]                     \n",
      "                                                                 dropout_34[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_109 (ReLU)                (None, 512, 96)      0           add_34[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_81 (Conv1D)              (None, 512, 48)      13824       re_lu_109[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_107 (BatchN (None, 512, 48)      192         conv1d_81[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_110 (ReLU)                (None, 512, 48)      0           batch_normalization_107[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_82 (Conv1D)              (None, 512, 96)      4608        re_lu_110[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_108 (BatchN (None, 512, 96)      384         conv1d_82[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_35 (Dropout)            (None, 512, 96)      0           batch_normalization_108[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_35 (Add)                    (None, 512, 96)      0           add_34[0][0]                     \n",
      "                                                                 dropout_35[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_111 (ReLU)                (None, 512, 96)      0           add_35[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "conv1d_83 (Conv1D)              (None, 512, 64)      30720       re_lu_111[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_109 (BatchN (None, 512, 64)      256         conv1d_83[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_112 (ReLU)                (None, 512, 64)      0           batch_normalization_109[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "one_to_two_2 (OneToTwo)         (None, 512, 512, 64) 0           re_lu_112[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "concat_dist2d_2 (ConcatDist2D)  (None, 512, 512, 65) 0           one_to_two_2[0][0]               \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_113 (ReLU)                (None, 512, 512, 65) 0           concat_dist2d_2[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_26 (Conv2D)              (None, 512, 512, 48) 28080       re_lu_113[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_110 (BatchN (None, 512, 512, 48) 192         conv2d_26[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_14 (Symmetrize2D)  (None, 512, 512, 48) 0           batch_normalization_110[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_114 (ReLU)                (None, 512, 512, 48) 0           symmetrize2d_14[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_27 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_114[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_111 (BatchN (None, 512, 512, 24) 96          conv2d_27[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_115 (ReLU)                (None, 512, 512, 24) 0           batch_normalization_111[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_28 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_115[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_112 (BatchN (None, 512, 512, 48) 192         conv2d_28[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_36 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_112[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_36 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_14[0][0]            \n",
      "                                                                 dropout_36[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_15 (Symmetrize2D)  (None, 512, 512, 48) 0           add_36[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_116 (ReLU)                (None, 512, 512, 48) 0           symmetrize2d_15[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_29 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_116[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_113 (BatchN (None, 512, 512, 24) 96          conv2d_29[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_117 (ReLU)                (None, 512, 512, 24) 0           batch_normalization_113[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_30 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_117[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_114 (BatchN (None, 512, 512, 48) 192         conv2d_30[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_37 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_114[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_37 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_15[0][0]            \n",
      "                                                                 dropout_37[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_16 (Symmetrize2D)  (None, 512, 512, 48) 0           add_37[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_118 (ReLU)                (None, 512, 512, 48) 0           symmetrize2d_16[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_31 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_118[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_115 (BatchN (None, 512, 512, 24) 96          conv2d_31[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_119 (ReLU)                (None, 512, 512, 24) 0           batch_normalization_115[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_32 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_119[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_116 (BatchN (None, 512, 512, 48) 192         conv2d_32[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_38 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_116[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_38 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_16[0][0]            \n",
      "                                                                 dropout_38[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_17 (Symmetrize2D)  (None, 512, 512, 48) 0           add_38[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_120 (ReLU)                (None, 512, 512, 48) 0           symmetrize2d_17[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_33 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_120[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_117 (BatchN (None, 512, 512, 24) 96          conv2d_33[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_121 (ReLU)                (None, 512, 512, 24) 0           batch_normalization_117[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_34 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_121[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_118 (BatchN (None, 512, 512, 48) 192         conv2d_34[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_39 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_118[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_39 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_17[0][0]            \n",
      "                                                                 dropout_39[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_18 (Symmetrize2D)  (None, 512, 512, 48) 0           add_39[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_122 (ReLU)                (None, 512, 512, 48) 0           symmetrize2d_18[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_35 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_122[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_119 (BatchN (None, 512, 512, 24) 96          conv2d_35[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_123 (ReLU)                (None, 512, 512, 24) 0           batch_normalization_119[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_36 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_123[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_120 (BatchN (None, 512, 512, 48) 192         conv2d_36[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_40 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_120[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_40 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_18[0][0]            \n",
      "                                                                 dropout_40[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_19 (Symmetrize2D)  (None, 512, 512, 48) 0           add_40[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_124 (ReLU)                (None, 512, 512, 48) 0           symmetrize2d_19[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_37 (Conv2D)              (None, 512, 512, 24) 10368       re_lu_124[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_121 (BatchN (None, 512, 512, 24) 96          conv2d_37[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "re_lu_125 (ReLU)                (None, 512, 512, 24) 0           batch_normalization_121[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "conv2d_38 (Conv2D)              (None, 512, 512, 48) 1152        re_lu_125[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "batch_normalization_122 (BatchN (None, 512, 512, 48) 192         conv2d_38[0][0]                  \n",
      "__________________________________________________________________________________________________\n",
      "dropout_41 (Dropout)            (None, 512, 512, 48) 0           batch_normalization_122[0][0]    \n",
      "__________________________________________________________________________________________________\n",
      "add_41 (Add)                    (None, 512, 512, 48) 0           symmetrize2d_19[0][0]            \n",
      "                                                                 dropout_41[0][0]                 \n",
      "__________________________________________________________________________________________________\n",
      "symmetrize2d_20 (Symmetrize2D)  (None, 512, 512, 48) 0           add_41[0][0]                     \n",
      "__________________________________________________________________________________________________\n",
      "cropping2d_2 (Cropping2D)       (None, 448, 448, 48) 0           symmetrize2d_20[0][0]            \n",
      "__________________________________________________________________________________________________\n",
      "upper_tri_2 (UpperTri)          (None, 99681, 48)    0           cropping2d_2[0][0]               \n",
      "__________________________________________________________________________________________________\n",
      "dense_2 (Dense)                 (None, 99681, 5)     245         upper_tri_2[0][0]                \n",
      "__________________________________________________________________________________________________\n",
      "switch_reverse_triu_2 (SwitchRe (None, 99681, 5)     0           dense_2[0][0]                    \n",
      "                                                                 stochastic_reverse_complement_2[0\n",
      "==================================================================================================\n",
      "Total params: 751,653\n",
      "Trainable params: 746,149\n",
      "Non-trainable params: 5,504\n",
      "__________________________________________________________________________________________________\n",
      "None\n",
      "model_strides [2048]\n",
      "target_lengths [99681]\n",
      "target_crops [-49585]\n",
      "Akita successfully loaded\n"
     ]
    }
   ],
   "source": [
    "# Get inputs from all arguments\n",
    "file_format = 'vcf'\n",
    "var_type = 'SV' \n",
    "fasta_path = '/pollard/data/vertebrate_genomes/human/hg38/hg38/hg38.fa'\n",
    "chrom_lengths_path = 'data/chrom_lengths_hg38'\n",
    "centromere_coords_path = 'data/centromere_coords_hg38'\n",
    "scores_to_use = ['mse', 'corr']\n",
    "revcomp = False\n",
    "no_revcomp = True\n",
    "svlen_limit = 700000\n",
    "seq_len = 1048576\n",
    "get_seq = True\n",
    "get_scores = True\n",
    "shift_by = [0]\n",
    "out_file = 'score_var_results'\n",
    "out_dir = 'temp'\n",
    "augment = False\n",
    "get_maps = False\n",
    "get_tracks = False\n",
    "var_set_size = 1000\n",
    "\n",
    "sys.path.insert(0, '../scripts')\n",
    "\n",
    "\n",
    "# Get example input file for each scenario\n",
    "input_file_types_dir = '/pollard/data/projects/kgjoni/CBTN_collab/input_file_types/'\n",
    "\n",
    "if file_format == 'df':\n",
    "    if var_type == 'simple':\n",
    "        in_file = f'{input_file_types_dir}d2acb167-24d7-4747-a7f8-de98205ad91a.consensus_somatic.norm.annot.public.maf'\n",
    "    elif var_type == 'SV':\n",
    "#         in_file = f'{input_file_types_dir}test.tsv'\n",
    "        in_file = f'{input_file_types_dir}909d602a-480c-4d70-ac31-05192f76eb08.manta.PASS.annotated.tsv.gz'\n",
    "elif file_format == 'vcf':\n",
    "    if var_type == 'simple':\n",
    "        in_file = f'{input_file_types_dir}e024fa3c-f459-44f5-aa0b-fc08eb5aea07.consensus_somatic.public.vcf'\n",
    "    elif var_type == 'SV':\n",
    "        in_file = f'{input_file_types_dir}30e05b01-b2be-477c-93eb-8e3625acb55a.somaticSV.vcf.gz'\n",
    "         \n",
    "# in_file = 'test/input/subset.somaticSV.vcf'        \n",
    "in_file = 'test/input/run_test_SV.bed'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # # # # # # # # # # # # # # # # # # # # # # # # # # # #\n",
    "# Adjust inputs from arguments\n",
    "\n",
    "\n",
    "# Handle argument dependencies\n",
    "\n",
    "if seq_len != 1048576:\n",
    "    get_scores = False\n",
    "    if svlen_limit > 0.66*seq_len:\n",
    "        svlen_limit = 0.66*seq_len\n",
    "        \n",
    "if not revcomp and not no_revcomp:\n",
    "    raise ValueError('Either revcomp and/or no_revcomp must be True.')\n",
    "if not get_seq and not get_scores:\n",
    "    raise ValueError('Either get_seq and/or get_scores must be True.')\n",
    "    \n",
    "\n",
    "# Adjust shift input: Remove shifts that are outside of allowed range\n",
    "max_shift = 0.4*seq_len\n",
    "shift_by = [x for x in shift_by if x > -max_shift and x < max_shift]\n",
    "\n",
    "\n",
    "# Adjust input for taking the reverse compliment\n",
    "revcomp_decision = []\n",
    "\n",
    "if no_revcomp:\n",
    "    revcomp_decision.append(False)\n",
    "if revcomp:\n",
    "    revcomp_decision.append(True)\n",
    "revcomp_decision_i = revcomp_decision\n",
    "\n",
    "    \n",
    "# Adjust input for taking the average score from augmented sequences\n",
    "if augment:\n",
    "    shift_by = [-1,0,1]\n",
    "    revcomp_decision = [True, False]\n",
    "#     scores_to_use = [x for x in scores_to_use if x in ['mse', 'corr']]\n",
    "\n",
    "\n",
    "# Create dictionaries to save sequences, maps, and disruption score tracks, if specified\n",
    "if get_seq:\n",
    "    sequences = {}\n",
    "    \n",
    "if get_maps:\n",
    "    variant_maps = {}\n",
    "    if get_scores == False:\n",
    "        get_scores = True\n",
    "        print('Must get scores to get maps: get_scores was set to True.')\n",
    "    \n",
    "if get_tracks:\n",
    "    variant_tracks = {}\n",
    "    if get_scores == False:\n",
    "        get_scores = True\n",
    "        print('Must get scores to get maps: get_scores was set to True.')\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # # # # # # # # # # # # # # # # # # # # # # # # # # # #\n",
    "# Get necessary files if they are not there\n",
    "\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "chrom_lengths_path = 'data/chrom_lengths_hg38'\n",
    "if not Path(chrom_lengths_path).is_file():\n",
    "    os.system('wget -P ./data/ https://raw.githubusercontent.com/ketringjoni/Akita_variant_scoring/main/data/chrom_lengths_hg38')\n",
    "    print('Chromosome lengths file downloaded as data/chrom_lengths_hg38.')\n",
    "\n",
    "centromere_coords_path = 'data/centromere_coords_hg38'\n",
    "if not Path(centromere_coords_path).is_file():\n",
    "    os.system('wget -P ./data/ https://raw.githubusercontent.com/ketringjoni/Akita_variant_scoring/main/data/centromere_coords_hg38')\n",
    "    print('Centromere coordinates file downloaded as data/centromere_coords_hg38.')\n",
    "\n",
    "if fasta_path == 'data/hg38.fa' and not Path(fasta_path).is_file():\n",
    "    os.system('wget -P ./data/ https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz')\n",
    "    os.system('gunzip data/hg38.fa.gz')\n",
    "    print('Fasta file downloaded as data/hg38.fa.')\n",
    "\n",
    "    \n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "out_file = os.path.join(out_dir, out_file)\n",
    " \n",
    "    \n",
    "    \n",
    "\n",
    "    \n",
    "# # # # # # # # # # # # # # # # # # # # # # # # # # # # #\n",
    "# Read in (and adjust) data \n",
    "    \n",
    "\n",
    "    \n",
    "\n",
    "import pandas as pd\n",
    "import pysam\n",
    "\n",
    "chrom_lengths = pd.read_table(chrom_lengths_path, header = None, names = ['CHROM', 'chrom_max'])\n",
    "centromere_coords = pd.read_table(centromere_coords_path, sep = '\\t')\n",
    "fasta_open = pysam.Fastafile(fasta_path)\n",
    "\n",
    "\n",
    "# Assign necessary values to variables across module\n",
    "\n",
    "# Module 1: reading utilities\n",
    "import reading_utils\n",
    "reading_utils.var_set_size = var_set_size\n",
    "\n",
    "\n",
    "# Module 2: get_seq utilities\n",
    "import get_seq_utils\n",
    "get_seq_utils.fasta_open = fasta_open\n",
    "get_seq_utils.chrom_lengths = chrom_lengths\n",
    "get_seq_utils.centromere_coords = centromere_coords\n",
    "\n",
    "get_seq_utils.svlen_limit = svlen_limit\n",
    "get_seq_utils.seq_length = seq_len\n",
    "get_seq_utils.half_patch_size = round(seq_len/2)\n",
    "\n",
    "\n",
    "# Module 2: get_scores utilities\n",
    "if get_scores:\n",
    "    import get_scores_utils\n",
    "    get_scores_utils.chrom_lengths = chrom_lengths\n",
    "    get_scores_utils.centromere_coords = centromere_coords\n",
    "\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Another way to do the filtering out:\n",
    "# filtered_out = [\n",
    "#     # Exclude mitochondrial variants\n",
    "#     list(variants[variants.CHROM == 'chrM'].var_index),\n",
    "#     # Exclude variants larger than limit\n",
    "#     [y for x,y in zip(variants.SVLEN, variants.var_index) if not pd.isnull(x) and abs(int(x)) > svlen_limit]]\n",
    "\n",
    "\n",
    "# add this to print in for loop for testing: str(i+1)+'/'+str(len(variants))\n",
    "\n",
    "\n",
    "\n",
    "# if yoou want to make gzip an argument:\n",
    "\n",
    "# parser.add_argument('--gz_input',\n",
    "#                     help = 'Use a gzipped input file', \n",
    "#                     action = 'store_const',\n",
    "#                     const = True, \n",
    "#                     default = 'No',\n",
    "#                     required = False)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Variants that should bring up each error:\n",
    "# centromeric variant: chr1:122845747-123875879\n",
    "# N composition > 5%: chr2:89274001-89283500\n",
    "# Variant larger than prediction window: chr1:30651501-35842000\n",
    "# Insertions or other types that are not supported\n",
    "\n",
    "# Other\n",
    "# Less than prediction window but too large to get score (nan): chr1:122845747-123875879\n",
    "# BND variants: 7: t[p[; 10: [p[t; 11: t]p]; 9: ]p]t\n",
    "# Different variant types:\n",
    "    # chrom_start chr2:400000-450000\n",
    "    # chrom_centro_left chr10:37800000-37900000\n",
    "    # chrom_centro_right  chr14:17600000-17700000\n",
    "    # chrom_end chr1:248746420-248766420\n",
    "    \n",
    "    \n",
    "# chr1\t8144964\t8145340\tCTTT\tC\tDEL\t-3\n",
    "# chr2\t54802149\tNaN\tC\t[chr19:3989901[C\tBND\tNaN\n",
    "# chr10\t132139208\t132139430\tG\t<DUP:TANDEM>\tDUP\t222\n",
    "# chr4\t186172375\t186177065\tC\t<DEL>\tDEL\t-4690\n",
    "# chr7\t66286935\t66286945\tT\t<INS>\tINS\tNaN\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cdfc148",
   "metadata": {},
   "source": [
    "# Scoring pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "08052e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 42\n",
    "shift = 0\n",
    "rev_comp = False\n",
    "\n",
    "variant = variants.iloc[i]\n",
    "\n",
    "CHR = variant.CHROM\n",
    "POS = variant.POS\n",
    "REF = variant.REF\n",
    "ALT = variant.ALT\n",
    "\n",
    "if 'END' in variants.columns:\n",
    "    END = variant.END\n",
    "    SVTYPE = variant.SVTYPE\n",
    "else:\n",
    "    END = np.nan\n",
    "    SVTYPE = np.nan\n",
    "    \n",
    "sequences_i = get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aeddd8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 547\n",
    "\n",
    "get_seq = False\n",
    "get_scores = True\n",
    "\n",
    "augment = True\n",
    "shift_by = [0]\n",
    "revcomp_decision = [False]\n",
    "scores_to_use = ['corr', 'mse']\n",
    "\n",
    "get_tracks = False\n",
    "get_maps = False\n",
    "\n",
    "sequences = {}\n",
    "variant_maps = {}\n",
    "variant_tracks = {}\n",
    "variant_scores = pd.DataFrame({'var_index':variants.index})\n",
    "\n",
    "\n",
    "for i in variants.index:\n",
    "\n",
    "    variant = variants.iloc[i]\n",
    "\n",
    "    var_index = variant.var_index\n",
    "    CHR = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    REF = variant.REF\n",
    "    ALT = variant.ALT\n",
    "\n",
    "\n",
    "    if 'SVTYPE' in variants.columns:\n",
    "        END = variant.END\n",
    "        SVTYPE = variant.SVTYPE\n",
    "        SVLEN = variant.SVLEN\n",
    "    else:\n",
    "        END = np.nan\n",
    "        SVTYPE = np.nan\n",
    "        SVLEN = 0\n",
    "\n",
    "    for shift in shift_by:\n",
    "\n",
    "        if augment: # if getting augmented score, take reverse complement only with 0 shift\n",
    "            if shift != 0 & True in revcomp_decision:\n",
    "                revcomp_decision_i = [False]\n",
    "            else:\n",
    "                revcomp_decision_i = revcomp_decision\n",
    "\n",
    "        for revcomp in revcomp_decision_i:\n",
    "\n",
    "            try:\n",
    "\n",
    "                if revcomp:\n",
    "                    revcomp_annot = '_revcomp'\n",
    "                else:\n",
    "                    revcomp_annot = ''\n",
    "\n",
    "                sequences_i = get_seq_utils.get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)\n",
    "\n",
    "\n",
    "                if get_seq:\n",
    "\n",
    "                    # Get relative position of variant in sequence\n",
    "                    var_rel_pos = str(sequences_i[-1]).replace(', ', '_')\n",
    "\n",
    "                    for ii in range(len(sequences_i[:-1][:3])): \n",
    "                        sequences[f'{var_index}_{shift}{revcomp_annot}_{ii}_{var_rel_pos}'] = sequences_i[:-1][ii]\n",
    "\n",
    "                if get_scores:\n",
    "\n",
    "                    scores = get_scores_utils.get_scores(POS, SVTYPE, SVLEN, \n",
    "                                                         sequences_i, scores_to_use, \n",
    "                                                         shift, revcomp, \n",
    "                                                         get_tracks, get_maps)\n",
    "\n",
    "                    if get_tracks:\n",
    "                        for track in [x for x in scores.keys() if 'track' in x]:\n",
    "                            variant_tracks[f'{var_index}_{track}_{shift}{revcomp_annot}'] = scores[track]\n",
    "                            del scores[track]\n",
    "\n",
    "                    if get_maps:\n",
    "                        variant_maps[f'{var_index}_{shift}{revcomp_annot}'] = scores['maps']\n",
    "                        del scores['maps']\n",
    "\n",
    "                    for score in scores:\n",
    "                        variant_scores.loc[variant_scores.var_index == var_index, \n",
    "                                           f'{score}_{shift}{revcomp_annot}'] = scores[score]\n",
    "\n",
    "\n",
    "                print(str(var_index) + ' (' + str(shift) + f' shift{revcomp_annot})')\n",
    "\n",
    "            except Exception as e: \n",
    "\n",
    "                print(str(var_index) + ' (' + str(shift) + f' shift{revcomp_annot})' + ': Error:', e)\n",
    "\n",
    "                pass\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "34aa4a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var_index</th>\n",
       "      <th>mse_0</th>\n",
       "      <th>corr_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.016132</td>\n",
       "      <td>0.973913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.007965</td>\n",
       "      <td>0.984667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.005242</td>\n",
       "      <td>0.995228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.009628</td>\n",
       "      <td>0.958351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>0.010467</td>\n",
       "      <td>0.976499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>0.979311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10</td>\n",
       "      <td>0.001901</td>\n",
       "      <td>0.950111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>11</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.968983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>12</td>\n",
       "      <td>0.003523</td>\n",
       "      <td>0.941520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>14</td>\n",
       "      <td>0.002650</td>\n",
       "      <td>0.972745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15</td>\n",
       "      <td>0.001685</td>\n",
       "      <td>0.928444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>16</td>\n",
       "      <td>0.001792</td>\n",
       "      <td>0.994262</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    var_index     mse_0    corr_0\n",
       "0           0  0.016132  0.973913\n",
       "1           1  0.007965  0.984667\n",
       "2           2  0.005242  0.995228\n",
       "3           3  0.009628  0.958351\n",
       "4           8  0.010467  0.976499\n",
       "5           9  0.000834  0.979311\n",
       "6          10  0.001901  0.950111\n",
       "7          11  0.000189  0.968983\n",
       "8          12  0.003523  0.941520\n",
       "9          13       NaN       NaN\n",
       "10         14  0.002650  0.972745\n",
       "11         15  0.001685  0.928444\n",
       "12         16  0.001792  0.994262"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check output\n",
    "\n",
    "out_file = 'temp/score_var_results'\n",
    "\n",
    "pd.read_csv(f'{out_file}_scores', sep = '\\t')\n",
    "\n",
    "# pd.read_csv(f'{out_file}_filtered_out', sep = '\\t')\n",
    "\n",
    "# np.load(f'{out_file}_tracks.npy', allow_pickle=\"TRUE\").item()\n",
    "\n",
    "# np.load(f'{out_file}_maps.npy', allow_pickle=\"TRUE\").item()\n",
    "\n",
    "# fasta_open2 = pysam.Fastafile(f'{out_file}_sequences.fa')\n",
    "# fasta_open2.references\n",
    "# fasta_open2.fetch(fasta_open2.references[0], 0, 100).upper()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c1cb7ae",
   "metadata": {},
   "source": [
    "# Tests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c317d39",
   "metadata": {},
   "source": [
    "## Edge cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "56532677",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "24d795db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate test set for simple variants\n",
    "\n",
    "# Features to test:\n",
    "SVTYPE2 = ['SNP', 'del', 'ins']\n",
    "var_pos_coord = {'chrom_mid':['chr1', 244397848],\n",
    "                  'chrom_start':['chr2', 400000],\n",
    "                  'chrom_centro_left':['chr10', 37900000], #37900000\n",
    "                  'chrom_centro_right':['chr3', 94400000], #14: 17600000\n",
    "                  'chrom_end':['chr1', 248766420],\n",
    "                  'centromere':['chr11', 52000000]}\n",
    "\n",
    "# Create variants dataframe\n",
    "run_test_simple = pd.DataFrame(tuple(product(var_pos_coord.items(), SVTYPE2)), \n",
    "                         columns = ['var_pos_coord', 'SVTYPE2'])\n",
    "\n",
    "# Split up var_pos_coord\n",
    "run_test_simple[['var_position', 'coord']] = pd.DataFrame(run_test_simple['var_pos_coord'].tolist(), \n",
    "                                                          index=run_test_simple.index)\n",
    "run_test_simple[['CHROM', 'POS']] = pd.DataFrame(run_test_simple['coord'].tolist(), index=run_test_simple.index)\n",
    "run_test_simple.drop(['var_pos_coord', 'coord'], axis = 1, inplace = True)\n",
    "\n",
    "\n",
    "# Get REF alleles and remaining ALT alleles\n",
    "nt = ['A', 'T', 'C', 'G']\n",
    "for i in run_test_simple.index:\n",
    "    \n",
    "    variant = run_test_simple.loc[i]\n",
    "\n",
    "    CHROM = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    SVTYPE2 = variant.SVTYPE2\n",
    "    \n",
    "    if SVTYPE2 == 'del':\n",
    "        REF = fasta_open.fetch(CHROM, POS - 1, POS - 1 + 50).upper()\n",
    "        run_test_simple.loc[i,'REF'] = REF\n",
    "        run_test_simple.loc[i,'ALT'] = REF[0]\n",
    "    elif SVTYPE2 == 'ins':\n",
    "        ALT = fasta_open.fetch(CHROM, POS - 1, POS - 1 + 50).upper()\n",
    "        run_test_simple.loc[i,'ALT'] = ALT\n",
    "        run_test_simple.loc[i,'REF'] = ALT[0]\n",
    "    elif SVTYPE2 == 'SNP':\n",
    "        REF = fasta_open.fetch(CHROM, POS - 1, POS).upper()\n",
    "        run_test_simple.loc[i,'REF'] = REF\n",
    "        run_test_simple.loc[i,'ALT'] = [x for x in nt if x != REF][0]\n",
    "        \n",
    "run_test_simple[['CHROM', 'POS', 'POS', 'REF', 'ALT']].to_csv('test_data/test_set_edge_cases/test_set_edge_simple.bed', \n",
    "                                                              sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "10c12dd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate test set for SVs\n",
    "\n",
    "# Features to test:\n",
    "SVTYPE2 = ['del', 'DEL', 'DUP', 'INV']\n",
    "SVLEN = [500,5000]\n",
    "var_pos_coord = {'chrom_mid':['chr1', 244397848],\n",
    "                  'chrom_start':['chr2', 400000],\n",
    "                  'chrom_centro_left':['chr10', 37600000],\n",
    "                  'chrom_centro_right':['chr3', 94400000],\n",
    "                  'chrom_end':['chr1', 248766420],\n",
    "                  'centromere':['chr11', 52000000]}\n",
    "\n",
    "# Create variants dataframe\n",
    "run_test_SV = pd.DataFrame(tuple(product(var_pos_coord.items(), SVLEN, SVTYPE2)), \n",
    "                         columns = ['var_pos_coord', 'SVLEN', 'SVTYPE2'])\n",
    "\n",
    "# Split up var_pos_coord\n",
    "run_test_SV[['var_position', 'coord']] = pd.DataFrame(run_test_SV['var_pos_coord'].tolist(), index=run_test_SV.index)\n",
    "run_test_SV[['CHROM', 'POS']] = pd.DataFrame(run_test_SV['coord'].tolist(), index=run_test_SV.index)\n",
    "run_test_SV.drop(['var_pos_coord', 'coord'], axis = 1, inplace = True)\n",
    "\n",
    "# Adust SVTYPE and ALT alleles\n",
    "run_test_SV[['SVTYPE']] = run_test_SV[['SVTYPE2']]\n",
    "run_test_SV.loc[run_test_SV.SVTYPE2 == 'del', 'SVTYPE'] = 'DEL'\n",
    "run_test_SV.loc[run_test_SV.SVTYPE2 == 'DUP', 'ALT'] = '<DUP:TANDEM>'\n",
    "run_test_SV.loc[run_test_SV.SVTYPE2 == 'INV', 'ALT'] = '<INV>'\n",
    "run_test_SV.loc[run_test_SV.SVTYPE2 == 'DEL', 'ALT'] = '<DEL>'\n",
    "\n",
    "# Remove del alleles that are 5000bp long\n",
    "run_test_SV['END'] = run_test_SV.POS+ run_test_SV.SVLEN\n",
    "run_test_SV = run_test_SV[~((run_test_SV.SVLEN == 5000) & (run_test_SV.SVTYPE2 == 'del'))]\n",
    "\n",
    "\n",
    "# Get REF alleles and remaining ALT alleles\n",
    "run_test_SV.reset_index(inplace = True, drop = True)\n",
    "\n",
    "for i in run_test_SV.index:\n",
    "    \n",
    "    variant = run_test_SV.loc[i]\n",
    "\n",
    "    CHROM = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    SVLEN = variant.SVLEN\n",
    "    SVTYPE2 = variant.SVTYPE2\n",
    "    \n",
    "    if SVTYPE2 == 'del':\n",
    "        REF = fasta_open.fetch(CHROM, POS - 1, POS - 1 + SVLEN).upper()\n",
    "        run_test_SV.loc[i,'REF'] = REF\n",
    "        run_test_SV.loc[i,'ALT'] = REF[0]\n",
    "    else:\n",
    "        run_test_SV.loc[i,'REF'] = fasta_open.fetch(CHROM, POS - 1, POS).upper()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "8984aa2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate test set for BNDs\n",
    "\n",
    "# Features to test:\n",
    "\n",
    "# Create all possible pairs of var_positions with coordinates\n",
    "var_pos_coord_1 = {'chrom_mid':['chr1', 244397848],\n",
    "                  'chrom_start':['chr2', 400000],\n",
    "                  'chrom_centro_left':['chr10', 37600000], \n",
    "                  'chrom_centro_right':['chr11', 56200000],\n",
    "                  'chrom_end':['chr1', 248766420],\n",
    "                  'centromere':['chr11', 52000000]}\n",
    "var_pos_coord_2 = {'chrom_mid':['chr6', 73541678],\n",
    "                  'chrom_start':['chr10', 400000],\n",
    "                  'chrom_centro_left':['chr5', 45700000],\n",
    "                  'chrom_centro_right':['chr19', 28400000],\n",
    "                  'chrom_end':['chr20', 64044167],\n",
    "                  'centromere':['chr15', 17500000]}\n",
    "var_pos_coord_pairs = list(product(var_pos_coord_1.items(), var_pos_coord_2.items()))\n",
    "\n",
    "ALT_type = ['t[p[', 't]p]', ']p]t', '[p[t']\n",
    "ALT_allele = ['SNP', 'ins']\n",
    "\n",
    "# Create variants dataframe\n",
    "run_test_BND = pd.DataFrame(tuple(product(var_pos_coord_pairs, ALT_type, ALT_allele)), \n",
    "                         columns = ['var_pos_coord', 'ALT_type', 'ALT_allele'])\n",
    "\n",
    "# Add remaining columns\n",
    "run_test_BND['END'] = np.nan\n",
    "run_test_BND['SVLEN'] = np.nan\n",
    "run_test_BND['SVTYPE'] = 'BND'\n",
    "\n",
    "# Split up var_pos_coord\n",
    "run_test_BND[['var_pos_coord1', 'var_pos_coord2']] = pd.DataFrame(run_test_BND['var_pos_coord'].tolist(), index=run_test_BND.index)\n",
    "run_test_BND[['var_position1', 'coord1']] = pd.DataFrame(run_test_BND['var_pos_coord1'].tolist(), index=run_test_BND.index)\n",
    "run_test_BND[['var_position2', 'coord2']] = pd.DataFrame(run_test_BND['var_pos_coord2'].tolist(), index=run_test_BND.index)\n",
    "run_test_BND[['CHROM', 'POS']] = pd.DataFrame(run_test_BND['coord1'].tolist(), index=run_test_BND.index)\n",
    "run_test_BND[['CHR2', 'POS2']] = pd.DataFrame(run_test_BND['coord2'].tolist(), index=run_test_BND.index)\n",
    "run_test_BND.drop(['var_pos_coord', 'var_pos_coord1', 'var_pos_coord2', 'coord1', 'coord2'], axis = 1, inplace = True)\n",
    "\n",
    "# Get REF alleles and remaining ALT alleles\n",
    "for i in run_test_BND.index:\n",
    "    \n",
    "    variant = run_test_BND.loc[i]\n",
    "\n",
    "    CHROM = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    CHR2 = variant.CHR2\n",
    "    POS2 = variant.POS2\n",
    "    ALT_type = variant.ALT_type\n",
    "    ALT_allele = variant.ALT_allele\n",
    "    \n",
    "    REF = fasta_open.fetch(CHROM, POS-1, POS).upper()\n",
    "    run_test_BND.loc[i,'REF'] = REF\n",
    "        \n",
    "    if ALT_allele == 'SNP':\n",
    "        ALT = REF\n",
    "    elif ALT_allele == 'ins':\n",
    "        ALT = fasta_open.fetch(CHROM, POS-1, POS-1 + 3).upper()\n",
    "        \n",
    "        \n",
    "    if ALT_type == 't[p[':\n",
    "        ALT = f'{ALT}[{CHR2}:{POS2}['      \n",
    "    elif ALT_type == 't]p]':\n",
    "        ALT = f'{ALT}]{CHR2}:{POS2}]'\n",
    "    elif ALT_type == ']p]t':\n",
    "        ALT = f']{CHR2}:{POS2}]{ALT}'\n",
    "    elif ALT_type == '[p[t':\n",
    "        ALT = f'[{CHR2}:{POS2}[{ALT}' \n",
    "            \n",
    "    run_test_BND.loc[i,'ALT'] = ALT\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "61a553db",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_test_SV = pd.concat([run_test_SV[['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']],\n",
    "                        run_test_BND[['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']]],\n",
    "                        axis = 0)\n",
    "run_test_SV.to_csv('test_data/test_set_edge_cases/test_set_edge_SV.bed', sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "767a5ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test this out - temp code\n",
    "\n",
    "variants = run_test_BND\n",
    "\n",
    "run_test_BND[run_test_BND.var_position1 == 'chrom_start']\n",
    "\n",
    "i = 48\n",
    "for i in run_test_BND[run_test_BND.var_position1 == 'chrom_start'].index.drop([58, 59, 62, 63, 64, 65, 68, 69, 74, 75, 78, 79, 80, 81, 84, 85]):\n",
    "    shift = 0\n",
    "    revcomp = False\n",
    "\n",
    "    variant = variants.loc[i]\n",
    "\n",
    "    # var_index = variant.var_index\n",
    "    CHR = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    REF = variant.REF\n",
    "    ALT = variant.ALT\n",
    "\n",
    "    if 'END' in variants.columns:\n",
    "        END = variant.END\n",
    "        SVTYPE = variant.SVTYPE\n",
    "    else:\n",
    "        END = np.nan\n",
    "        SVTYPE = np.nan\n",
    "\n",
    "    sequences_i = get_seq_utils.get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)\n",
    "\n",
    "    print(i, fasta_open.fetch(CHR, 0, MB).upper() == sequences_i[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3de01357",
   "metadata": {},
   "source": [
    "### Run test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4e008809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17, 329)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_path = 'test_data/test_set_edge_cases/test_set_edge_'\n",
    "\n",
    "run_test_simple = pd.read_csv(f'{test_path}simple.bed', sep ='\\t')\n",
    "run_test_SV = pd.read_csv(f'{test_path}SV.bed', sep ='\\t')\n",
    "\n",
    "len(run_test_simple), len(run_test_SV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fbef3a31",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = reading_utils.read_input(f'{test_path}SV.bed', 0)\n",
    "\n",
    "variants['var_index'] = variants.index\n",
    "variant_scores = pd.DataFrame({'var_index':variants.var_index})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e851f64e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Variant not centered; too close to chromosome arm ends.\n",
      "76 (-10000 shift)\n",
      "Warning: Variant not centered; too close to chromosome arm ends.\n",
      "76 (0 shift)\n",
      "Warning: Variant not centered; too close to chromosome arm ends.\n",
      "76 (10000 shift)\n"
     ]
    }
   ],
   "source": [
    "# Test all variants\n",
    "\n",
    "get_seq = False\n",
    "get_scores = True\n",
    "\n",
    "shift_by = [-10000,0,10000]\n",
    "revcomp_decision = [True, False]\n",
    "scores_to_use = ['corr', 'mse']\n",
    "\n",
    "for i in [76]:#variants.index:\n",
    "\n",
    "    variant = variants.iloc[i]\n",
    "\n",
    "    var_index = variant.var_index\n",
    "    CHR = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    REF = variant.REF\n",
    "    ALT = variant.ALT\n",
    "\n",
    "\n",
    "    if 'SVTYPE' in variants.columns:\n",
    "        END = variant.END\n",
    "        SVTYPE = variant.SVTYPE\n",
    "        SVLEN = variant.SVLEN\n",
    "    else:\n",
    "        END = np.nan\n",
    "        SVTYPE = np.nan\n",
    "        SVLEN = 0\n",
    "\n",
    "    for shift in shift_by:\n",
    "\n",
    "        if augment: # if getting augmented score, take reverse complement only with 0 shift\n",
    "            if shift != 0 & True in revcomp_decision:\n",
    "                revcomp_decision_i = [False]\n",
    "            else:\n",
    "                revcomp_decision_i = revcomp_decision\n",
    "\n",
    "        for revcomp in revcomp_decision_i:\n",
    "\n",
    "            try:\n",
    "\n",
    "                if revcomp:\n",
    "                    revcomp_annot = '_revcomp'\n",
    "                else:\n",
    "                    revcomp_annot = ''\n",
    "\n",
    "                sequences_i = get_seq_utils.get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)\n",
    "\n",
    "\n",
    "                if get_seq:\n",
    "\n",
    "                    # Get relative position of variant in sequence\n",
    "                    var_rel_pos = str(sequences_i[-1]).replace(', ', '_')\n",
    "\n",
    "                    for ii in range(len(sequences_i[:-1][:3])): \n",
    "                        sequences[f'{var_index}_{shift}{revcomp_annot}_{ii}_{var_rel_pos}'] = sequences_i[:-1][ii]\n",
    "\n",
    "                if get_scores:\n",
    "\n",
    "                    scores = get_scores_utils.get_scores(POS, SVTYPE, SVLEN, \n",
    "                                                         sequences_i, scores_to_use, \n",
    "                                                         shift, revcomp, \n",
    "                                                         get_tracks, get_maps)\n",
    "\n",
    "                    if get_tracks:\n",
    "                        for track in [x for x in scores.keys() if 'track' in x]:\n",
    "                            variant_tracks[f'{var_index}_{track}_{shift}{revcomp_annot}'] = scores[track]\n",
    "                            del scores[track]\n",
    "\n",
    "                    if get_maps:\n",
    "                        variant_maps[f'{var_index}_{shift}{revcomp_annot}'] = scores['maps']\n",
    "                        del scores['maps']\n",
    "\n",
    "                    for score in scores:\n",
    "                        variant_scores.loc[variant_scores.var_index == var_index, \n",
    "                                           f'{score}_{shift}{revcomp_annot}'] = scores[score]\n",
    "\n",
    "\n",
    "                print(str(var_index) + ' (' + str(shift) + f' shift{revcomp_annot})')\n",
    "\n",
    "            except Exception as e: \n",
    "\n",
    "                print(str(var_index) + ' (' + str(shift) + f' shift{revcomp_annot})' + ': Error:', e)\n",
    "\n",
    "                pass\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b0aa074",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33400b77",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7cc5554",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee16fd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dc240ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 76 seq length is off?, 100, 104-107, 110, 116, 117, 120-123, 126, 127"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7681a5a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Variant not centered; too close to chromosome arm ends\n",
      "Warning: Variant not centered; too close to chromosome arm ends\n",
      "Warning: Variant not centered; too close to chromosome arm ends\n",
      "Warning: Variant not centered; too close to chromosome arm ends\n",
      "Warning: Variant not centered; too close to chromosome arm ends\n",
      "Warning: Variant not centered; too close to chromosome arm ends\n"
     ]
    }
   ],
   "source": [
    "# Test one variant at a time\n",
    "\n",
    "i = 42\n",
    "\n",
    "get_seq = False\n",
    "get_scores = True\n",
    "\n",
    "shift_by = [-10000,0,10000]\n",
    "revcomp_decision = [True, False]\n",
    "scores_to_use = ['corr', 'mse']\n",
    "\n",
    "variant = variants.loc[i]\n",
    "\n",
    "var_index = variant.var_index\n",
    "CHR = variant.CHROM\n",
    "POS = variant.POS\n",
    "REF = variant.REF\n",
    "ALT = variant.ALT\n",
    "\n",
    "if 'END' in variants.columns:\n",
    "    END = variant.END\n",
    "    SVTYPE = variant.SVTYPE\n",
    "else:\n",
    "    END = np.nan\n",
    "    SVTYPE = np.nan\n",
    "\n",
    "for shift in shift_by:\n",
    "\n",
    "    for revcomp in revcomp_decision:\n",
    "\n",
    "\n",
    "        if revcomp:\n",
    "            revcomp_annot = '_revcomp'\n",
    "        else:\n",
    "            revcomp_annot = ''\n",
    "\n",
    "        sequences_i = utils.get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)\n",
    "\n",
    "        if get_seq:\n",
    "\n",
    "            for ii in range(len(sequences_i)): # note that for BNDs, the last entry in seq is BND_rel_pos\n",
    "                sequences[f'{var_index}_{shift}{revcomp_annot}_{ii}'] = sequences_i[ii]\n",
    "\n",
    "        if get_scores:\n",
    "\n",
    "            scores = utils.get_scores(CHR, POS, REF, ALT, sequences_i, SVTYPE, scores_to_use, shift, revcomp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34b5df94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run test sets\n",
    "\n",
    "# python scripts/score_var.py --in test/input/run_test_simple.bed --fa /pollard/data/vertebrate_genomes/human/hg38/hg38/hg38.fa \\\n",
    "# --shift_by -10000 0 10000 --revcomp --get_seq --dir test/output --file run_test_simple  \n",
    "\n",
    "# python score_var.py --in test/input/run_test_SV.bed --fa /pollard/data/vertebrate_genomes/human/hg38/hg38/hg38.fa \\\n",
    "# --shift_by -10000 0 10000 --revcomp --dir test/output --file run_test_SV  \n",
    "\n",
    "# Test sequence length\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "fasta_path_edited = 'test/output/run_test_simple_sequences.fa'\n",
    "fasta_open_edited = pysam.Fastafile(fasta_path_edited)\n",
    "\n",
    "len(fasta_open_edited.fetch('0_-10000_0', 0, MB+2).upper())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd83a05",
   "metadata": {},
   "source": [
    "## Sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90534dc2",
   "metadata": {},
   "source": [
    "Only the test sequences below were generated in this way, the rest were generated and annotated on SnapGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de5cc4f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test 4\n",
    "\n",
    "CHR = 'chr11'\n",
    "POS = 56200000\n",
    "\n",
    "\n",
    "# left side\n",
    "fasta_open.fetch(CHR, 55800000, POS - 1).upper()\n",
    "\n",
    "# deletion\n",
    "fasta_open.fetch(CHR, POS-1, POS-1 + 100000).upper()\n",
    "\n",
    "# right side\n",
    "fasta_open.fetch(CHR, POS-1 + 100000, 55800000+MB).upper()\n",
    "\n",
    "# ALT right side\n",
    "fasta_open.fetch(CHR, 55800000+MB, 55800000+MB+100000).upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd555dc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test 7: t[p[, chrom_start, chrom_centro_right\n",
    "\n",
    "# you are getting one extra bp on the left, yoou need the code to shift it right by 1\n",
    "CHR = 'chr2'\n",
    "POS = 400000\n",
    "\n",
    "CHR2 = 'chr19'\n",
    "POS2 = 28400000 # closer\n",
    "\n",
    "\n",
    "# right side\n",
    "fasta_open.fetch(CHR2, POS2 - 1, 28100000 + MB).upper()\n",
    "\n",
    "# right side REF (left end of it)\n",
    "fasta_open.fetch(CHR2, 28100000, POS2 - 1).upper()\n",
    "\n",
    "\n",
    "\n",
    "# left side\n",
    "left_side = POS2 - 1 - 28100000\n",
    "fasta_open.fetch(CHR, POS - left_side, POS).upper()\n",
    "\n",
    "# left side REF (right end of it)\n",
    "fasta_open.fetch(CHR, POS, POS + MB - left_side).upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3a9015d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test 8: ]p]t, chrom_centro_left, chrom_centro_left\n",
    "\n",
    "# you are getting one extra bp on the left, yoou need the code to shift it right by 1\n",
    "CHR = 'chr10'\n",
    "POS = 37900000 # closer\n",
    "\n",
    "CHR2 = 'chr5'\n",
    "POS2 = 45700000\n",
    "\n",
    "\n",
    "# right side\n",
    "fasta_open.fetch(CHR, POS - 1, 38000000).upper()\n",
    "# not including centromere start (0-based)\n",
    "\n",
    "# right side REF (left end of it)\n",
    "fasta_open.fetch(CHR, 38000000 - MB, POS - 1).upper()\n",
    "\n",
    "\n",
    "\n",
    "# left side\n",
    "fasta_open.fetch(CHR2, POS2 - (MB-distance), POS2).upper()\n",
    "\n",
    "\n",
    "# left side REF (right end of it)\n",
    "distance = 38000000 - (POS - 1)\n",
    "fasta_open.fetch(CHR2, POS2, POS2 + distance).upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2631306",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test 9: [p[t, chrom_centro_left, chrom_centro_left\n",
    "\n",
    "# you are getting one extra bp on the left, yoou need the code to shift it right by 1\n",
    "CHR = 'chr1'\n",
    "POS = 248766420 # closer\n",
    "\n",
    "CHR2 = 'chr10'\n",
    "POS2 = 400000\n",
    "\n",
    "\n",
    "# right side\n",
    "fasta_open.fetch(CHR, POS - 1, 248956422).upper()\n",
    "# not including centromere start (0-based)\n",
    "\n",
    "# right side REF (left end of it)\n",
    "fasta_open.fetch(CHR, 248956422 - MB, POS - 1).upper()\n",
    "\n",
    "\n",
    "# left side\n",
    "distance = 248956422 - (POS - 1)\n",
    "str(Seq(fasta_open.fetch(CHR2, POS2 - 1, POS2 - 1 + (MB - distance)).upper()).reverse_complement())\n",
    "\n",
    "# left side REF (right end of it)\n",
    "str(Seq(fasta_open.fetch(CHR2, POS2 - 1 - distance, POS2 - 1).upper()).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "26e39028",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_index</th>\n",
       "      <th>ID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>END</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>SVLEN</th>\n",
       "      <th>reason_for_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>test_0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8145000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAGGTCAGCACCATCCTGGTTAACAAGGTGAAGCCCCATCTCTACT...</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrom_mid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>test_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8145000</td>\n",
       "      <td>8145500.0</td>\n",
       "      <td>G</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-500.0</td>\n",
       "      <td>chrom_mid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11.0</td>\n",
       "      <td>test_2</td>\n",
       "      <td>chr2</td>\n",
       "      <td>400000</td>\n",
       "      <td>450000.0</td>\n",
       "      <td>C</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-50000.0</td>\n",
       "      <td>chrom_start</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>test_3</td>\n",
       "      <td>chr10</td>\n",
       "      <td>37800000</td>\n",
       "      <td>37900000.0</td>\n",
       "      <td>G</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-100000.0</td>\n",
       "      <td>chrom_centro_left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>test_4</td>\n",
       "      <td>chr11</td>\n",
       "      <td>56200000</td>\n",
       "      <td>56300000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-100000.0</td>\n",
       "      <td>chrom_centro_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>test_5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248746420</td>\n",
       "      <td>248766420.0</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>DEL</td>\n",
       "      <td>-20000.0</td>\n",
       "      <td>chrom_end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2.0</td>\n",
       "      <td>test_6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>244397848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A]chr6:73541678]</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrom_mid_chrom_mid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>72.0</td>\n",
       "      <td>test_7</td>\n",
       "      <td>chr2</td>\n",
       "      <td>400000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C[chr19:28400000[</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrom_start_chrom_centro_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>116.0</td>\n",
       "      <td>test_8</td>\n",
       "      <td>chr10</td>\n",
       "      <td>37900000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>]chr5:45700000]T</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrom_centro_left_chrom_centro_left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>206.0</td>\n",
       "      <td>test_9</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248766420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>[chr10:400000[A</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrom_end_chrom_start</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   test_index      ID  CHROM        POS          END  \\\n",
       "0         NaN  test_0   chr1    8145000          NaN   \n",
       "1         NaN  test_1   chr1    8145000    8145500.0   \n",
       "2        11.0  test_2   chr2     400000     450000.0   \n",
       "3         NaN  test_3  chr10   37800000   37900000.0   \n",
       "4         NaN  test_4  chr11   56200000   56300000.0   \n",
       "5         NaN  test_5   chr1  248746420  248766420.0   \n",
       "6         2.0  test_6   chr1  244397848          NaN   \n",
       "7        72.0  test_7   chr2     400000          NaN   \n",
       "8       116.0  test_8  chr10   37900000          NaN   \n",
       "9       206.0  test_9   chr1  248766420          NaN   \n",
       "\n",
       "                                                 REF                ALT  \\\n",
       "0  GAGGTCAGCACCATCCTGGTTAACAAGGTGAAGCCCCATCTCTACT...                  G   \n",
       "1                                                  G              <DEL>   \n",
       "2                                                  C              <DEL>   \n",
       "3                                                  G              <DEL>   \n",
       "4                                                  A              <DEL>   \n",
       "5                                                  T              <DEL>   \n",
       "6                                                  A   A]chr6:73541678]   \n",
       "7                                                  C  C[chr19:28400000[   \n",
       "8                                                  T   ]chr5:45700000]T   \n",
       "9                                                  A    [chr10:400000[A   \n",
       "\n",
       "  SVTYPE     SVLEN                      reason_for_test  \n",
       "0    NaN       NaN                            chrom_mid  \n",
       "1    DEL    -500.0                            chrom_mid  \n",
       "2    DEL  -50000.0                          chrom_start  \n",
       "3    DEL -100000.0                    chrom_centro_left  \n",
       "4    DEL -100000.0                   chrom_centro_right  \n",
       "5    DEL  -20000.0                            chrom_end  \n",
       "6    BND       NaN                  chrom_mid_chrom_mid  \n",
       "7    BND       NaN       chrom_start_chrom_centro_right  \n",
       "8    BND       NaN  chrom_centro_left_chrom_centro_left  \n",
       "9    BND       NaN                chrom_end_chrom_start  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = pd.read_csv('test_data/test_set_sequences/test_set_sequences.txt', sep ='\\t')\n",
    "variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2afa1fbe",
   "metadata": {},
   "source": [
    "### Run test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25d7cf39",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Variant not centered; too close to chromosome arm ends.\n",
      "7 BND\n",
      "True True\n",
      "TCCAAATATG TCCAAATATG\n",
      "True True\n",
      "TAGAACTGGC TAGAACTGGC\n",
      "True True\n",
      "TCCAAATATG TCCAAATATG\n"
     ]
    }
   ],
   "source": [
    "n = 10\n",
    "shift = 0\n",
    "revcomp = False\n",
    "\n",
    "for i in [7]:#range(0,9):\n",
    "\n",
    "    variant = variants.iloc[i]\n",
    "\n",
    "    var_index = variant.ID\n",
    "    CHR = variant.CHROM\n",
    "    POS = variant.POS\n",
    "    REF = variant.REF\n",
    "    ALT = variant.ALT\n",
    "\n",
    "    if 'END' in variants.columns:\n",
    "        END = variant.END\n",
    "        SVTYPE = variant.SVTYPE\n",
    "    else:\n",
    "        END = np.nan\n",
    "        SVTYPE = np.nan\n",
    "\n",
    "    sequences_i = get_seq_utils.get_sequences_SV(CHR, POS, REF, ALT, END, SVTYPE, shift, revcomp)\n",
    "\n",
    "    if var_index == 'test_0':\n",
    "        var_index = 'test_1'\n",
    "\n",
    "    print(i, SVTYPE)\n",
    "\n",
    "    if SVTYPE != 'BND':\n",
    "\n",
    "        # non_BND\n",
    "\n",
    "        custom_REF = pd.read_csv(f'test/sequences/test{var_index[-1]}_REF.txt', header = None).loc[0].values[0]\n",
    "        print(len(custom_REF) == MB, sequences_i[0]== custom_REF)\n",
    "\n",
    "        print(sequences_i[0][:n], custom_REF[:n])\n",
    "\n",
    "\n",
    "        custom_ALT = pd.read_csv(f'test/sequences/test{var_index[-1]}_ALT.txt', header = None).loc[0].values[0]\n",
    "        print(len(custom_ALT) == MB, sequences_i[1] == custom_ALT)\n",
    "\n",
    "        print(sequences_i[1][:n], custom_ALT[:n])\n",
    "\n",
    "    elif SVTYPE == 'BND':\n",
    "\n",
    "        # BND\n",
    "\n",
    "        custom_REF_L = pd.read_csv(f'test/sequences/test{var_index[-1]}_REF_L.txt', header = None).loc[0].values[0]\n",
    "        print(len(custom_REF_L) == MB, sequences_i[0]== custom_REF_L)\n",
    "\n",
    "        print(sequences_i[0][:n], custom_REF_L[:n])\n",
    "\n",
    "\n",
    "        custom_REF_R = pd.read_csv(f'test/sequences/test{var_index[-1]}_REF_R.txt', header = None).loc[0].values[0]\n",
    "        print(len(custom_REF_R) == MB, sequences_i[1]== custom_REF_R)\n",
    "\n",
    "        print(sequences_i[1][:n], custom_REF_R[:n])\n",
    "\n",
    "\n",
    "        custom_ALT = pd.read_csv(f'test/sequences/test{var_index[-1]}_ALT.txt', header = None).loc[0].values[0]\n",
    "        print(len(custom_ALT) == MB, sequences_i[2] == custom_ALT)\n",
    "\n",
    "        print(sequences_i[2][:n], custom_ALT[:n])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5e96769",
   "metadata": {},
   "source": [
    "# Time and memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae1af76d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To measure time and memory- paste the following in the corresponding lines in SuPreMo.py \n",
    "\n",
    "\n",
    "# Before everything\n",
    "\n",
    "import tracemalloc\n",
    "tracemalloc.start()\n",
    "\n",
    "\n",
    "# Before import sys\n",
    "\n",
    "import time\n",
    "start_time = time.time()\n",
    "first_size, first_peak = tracemalloc.get_traced_memory()\n",
    "\n",
    "\n",
    "# After closing log file\n",
    "\n",
    "end_time = time.time()\n",
    "print(end_time - start_time)\n",
    "second_size, second_peak = tracemalloc.get_traced_memory()\n",
    "\n",
    "\n",
    "# At the end of the script\n",
    "\n",
    "third_size, third_peak = tracemalloc.get_traced_memory()\n",
    "print(f\"first_size={first_size}, first_peak={first_peak}\")\n",
    "print(f\"second_size={second_size}, second_peak={second_peak}\")\n",
    "print(f\"third_size={third_size}, third_peak={third_peak}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a9dac17",
   "metadata": {},
   "source": [
    "# Extra code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a5ea9ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To combine tracks files from different sets (currently they accumulate in memory)\n",
    "\n",
    "tracks = np.load(f'{out_file}_tracks_0.npy', allow_pickle=\"TRUE\").item()\n",
    "\n",
    "for var_set in [1]:\n",
    "    tracks_n = np.load(f'{out_file}_tracks_{var_set}.npy', allow_pickle=\"TRUE\").item()\n",
    "\n",
    "    tracks.update(tracks_n)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66a28174",
   "metadata": {},
   "source": [
    "# Old version of some code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c2b61af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# before changing var_rel_pos\n",
    "\n",
    "\n",
    "def get_sequences(CHR, POS, REF, ALT, shift, revcomp: bool):\n",
    "  \n",
    "    '''\n",
    "    Get reference and alternate sequence for prediction from REF and ALT alleles by incorporating ALT into the reference genome.\n",
    "    Requires ALT allele to be a sequence and not a symbolic allele.\n",
    "    Use positive sign for a right shift and negative for a left shift.\n",
    "    revcomp: Take the reverse compliment of the resulting sequence.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    # Get reference sequence\n",
    "    \n",
    "    REF_len = len(REF)\n",
    "\n",
    "    REF_half_left = math.ceil((seq_length - REF_len)/2) - shift # if the REF allele is odd, shift right\n",
    "    REF_half_right = math.floor((seq_length - REF_len)/2) + shift\n",
    "\n",
    "    \n",
    "    # Annotate whether variant position with respect to chromosome arms ends\n",
    "    if len(REF) <= len(ALT):\n",
    "        var_position = get_variant_position(CHR, POS, REF_len, REF_half_left, REF_half_right)\n",
    "  \n",
    "    elif len(REF) > len(ALT):       \n",
    "        ALT_len = len(ALT)\n",
    "        ALT_half_left = math.ceil((seq_length - ALT_len)/2) - shift\n",
    "        ALT_half_right = math.floor((seq_length - ALT_len)/2) + shift   \n",
    "        var_position = get_variant_position(CHR, POS, ALT_len, ALT_half_left, ALT_half_right)\n",
    "    \n",
    "\n",
    "    # Get last coordinate of chromosome\n",
    "    chrom_max = int(chrom_lengths[chrom_lengths.CHROM == CHR[3:]]['chrom_max'])\n",
    "    \n",
    "    # Get centromere coordinates\n",
    "    centro_start = int(centromere_coords[centromere_coords.chrom == CHR]['start'])\n",
    "    centro_stop = int(centromere_coords[centromere_coords.chrom == CHR]['end'])\n",
    "    \n",
    "    \n",
    "    # Get start and end of reference sequence\n",
    "    if var_position == \"chrom_mid\":\n",
    "        REF_start = POS - REF_half_left\n",
    "        REF_stop = REF_start + seq_length \n",
    "    elif var_position == \"centromere\":\n",
    "        raise ValueError('Centromeric variant.')\n",
    "    else:\n",
    "        REF_start = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, 0, shift)\n",
    "        REF_stop = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, seq_length, shift)\n",
    "        print(\"Warning: Variant not centered; too close to chromosome arm ends.\")\n",
    "        \n",
    "        \n",
    "    # Get reference sequence\n",
    "    REF_seq = fasta_open.fetch(CHR, REF_start, REF_stop).upper()\n",
    "\n",
    "\n",
    "    # Error if N composition is more than 5% of sequence\n",
    "    if Counter(REF_seq)['N']/seq_length*100 > 5:\n",
    "        raise ValueError('N composition greater than 5%.')\n",
    "\n",
    "\n",
    "\n",
    "    # Error if reference sequence does not match given REF\n",
    "\n",
    "    if var_position == \"chrom_mid\":\n",
    "        var_rel_pos_REF = REF_half_left - 1 # subtract 1 bc POS is not included in variant\n",
    "\n",
    "    elif var_position == \"chrom_start\": \n",
    "        var_rel_pos_REF = POS - abs(shift) - 1\n",
    "\n",
    "    elif var_position == \"chrom_centro_right\": \n",
    "        var_rel_pos_REF = POS - centro_stop - abs(shift) - 1\n",
    "\n",
    "    elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "        var_rel_pos_REF = -(REF_stop - POS) - 1\n",
    "\n",
    "\n",
    "    if REF_seq[var_rel_pos_REF : var_rel_pos_REF + REF_len] != REF:\n",
    "        raise ValueError('Reference allele does not match hg38.')\n",
    "            \n",
    "            \n",
    "            \n",
    "    # Error if reference sequence is not the right length      \n",
    "    if len(REF_seq) != seq_length:\n",
    "        raise ValueError('Reference sequence generated is not the right length.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # For SNPs, MNPs, Insertions: \n",
    "    if len(REF) <= len(ALT):\n",
    "\n",
    "        # Create alternate sequence: change REF sequence at position from REF to ALT\n",
    "\n",
    "        ALT_seq = REF_seq\n",
    "\n",
    "        ALT_seq = ALT_seq[:var_rel_pos_REF] + ALT + ALT_seq[var_rel_pos_REF + REF_len:]\n",
    "\n",
    "\n",
    "        var_rel_pos_ALT = var_rel_pos_REF\n",
    "        \n",
    "        # Chop off ends of alternate sequence if it's longer \n",
    "        if len(ALT_seq) > len(REF_seq):\n",
    "            to_remove = (len(ALT_seq) - len(REF_seq))/2\n",
    "\n",
    "            if to_remove == 0.5:\n",
    "                ALT_seq = ALT_seq[1:]\n",
    "                var_rel_pos_ALT = var_rel_pos_REF - 1\n",
    "                \n",
    "            else:\n",
    "                ALT_seq = ALT_seq[math.ceil(to_remove) : -math.floor(to_remove)]\n",
    "                var_rel_pos_ALT = var_rel_pos_REF - math.ceil(to_remove)\n",
    "                \n",
    "            \n",
    "\n",
    "\n",
    "    # For Deletions\n",
    "    elif len(REF) > len(ALT):\n",
    "\n",
    "\n",
    "        del_len = len(REF) - len(ALT)\n",
    "        \n",
    "        to_add_left = math.ceil(del_len/2)\n",
    "        to_add_right = math.floor(del_len/2) \n",
    "\n",
    "        # Get start and end of reference sequence\n",
    "        if var_position == \"chrom_mid\":\n",
    "            ALT_start = REF_start - to_add_left\n",
    "            ALT_stop = REF_stop + to_add_right\n",
    "\n",
    "        elif var_position in [\"chrom_start\", \"chrom_centro_right\"]: \n",
    "            ALT_start = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, 0, shift)\n",
    "            ALT_stop = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, seq_length + del_len, shift)\n",
    "            \n",
    "        elif var_position in [\"chrom_centro_left\", \"chrom_end\"]: \n",
    "            ALT_start = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, 0 - del_len, shift)\n",
    "            ALT_stop = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, seq_length, shift)\n",
    "        \n",
    "        \n",
    "        \n",
    "        # Get alternate sequence\n",
    "        ALT_seq = fasta_open.fetch(CHR, ALT_start, ALT_stop).upper()\n",
    "        \n",
    "        \n",
    "        \n",
    "        # Error if alternate sequence does not match REF at POS\n",
    "\n",
    "        if var_position == \"chrom_mid\":\n",
    "            var_rel_pos_ALT = REF_half_left + to_add_left - 1\n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            var_rel_pos_ALT = POS - abs(shift) - 1\n",
    "            \n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            var_rel_pos_ALT = POS - centro_stop - abs(shift) - 1\n",
    "            \n",
    "        elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "            var_rel_pos_ALT = -(REF_stop - POS) - 1\n",
    "                \n",
    "                \n",
    "        if ALT_seq[var_rel_pos_ALT : var_rel_pos_ALT + REF_len] != REF:\n",
    "            raise ValueError('Sequence for the alternate allele does not match hg38 at REF position.')\n",
    "\n",
    "\n",
    "    \n",
    "        # Change alternate sequence to match ALT at POS\n",
    "        ALT_seq = ALT_seq[:var_rel_pos_ALT] + ALT + ALT_seq[var_rel_pos_ALT + REF_len:]\n",
    "\n",
    "            \n",
    "    if len(ALT_seq) != seq_length:\n",
    "        raise ValueError('Alternate sequence generated is not the right length.')\n",
    "         \n",
    "            \n",
    "    # Take reverse compliment of sequence\n",
    "    if revcomp:\n",
    "        REF_seq, ALT_seq = [str(Seq(x).reverse_complement()) for x in [REF_seq, ALT_seq]]\n",
    "\n",
    "        \n",
    "    return REF_seq, ALT_seq, [var_rel_pos_REF, var_rel_pos_ALT]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244e7fea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Changed the following to keep the middle of the ALT maps and to have nan there instead of 0s in REF map\n",
    "\n",
    "def assemple_BND_maps(vector_repr_L, vector_repr_R, BND_rel_pos_map, matrix_len = target_length_cropped, num_diags = 2):\n",
    "    \n",
    "    '''\n",
    "    This applies to BND predcitions.\n",
    "    \n",
    "    Get predicted matrix from Akita predictions. \n",
    "    Output is a 448x448 array with the contact frequency at each 2048 bp bin corresponding to a 917,504 bp sequence (32 bins are cropped on each end from the prediction).\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    z = np.zeros((matrix_len,matrix_len))\n",
    "\n",
    "    # make df out of matrix indices and filter to get top left and bottom right parts\n",
    "    indices = np.triu_indices(matrix_len, num_diags)\n",
    "    indices = pd.DataFrame(np.column_stack(indices), columns = ['rows', 'cols'])\n",
    "\n",
    "    indices_L = tuple(indices.query('cols < @BND_rel_pos_map').T.apply(np.array, axis=1))\n",
    "    indices_R = tuple(indices.query('rows >= @BND_rel_pos_map').T.apply(np.array, axis=1))\n",
    "    \n",
    "    z[indices_L] = vector_repr_L\n",
    "    z[indices_R] = vector_repr_R\n",
    "    \n",
    "    for i in range(-num_diags+1,num_diags):\n",
    "        set_diag(z, np.nan, i)\n",
    "        \n",
    "    return z + z.T\n",
    "\n",
    "def get_masked_BND_maps(matrices, rel_pos_map):\n",
    "\n",
    "    # Get REF and ALT vectors, excluding diagonal \n",
    "    indexes_left = np.triu_indices(rel_pos_map, 2)\n",
    "    indexes_right = np.triu_indices(target_length_cropped - rel_pos_map, 2)\n",
    "\n",
    "    REF_L = get_left_BND_map(matrices[0], rel_pos_map)[indexes_left]\n",
    "    REF_R = get_right_BND_map(matrices[1], rel_pos_map)[indexes_right]\n",
    "    ALT_L = get_left_BND_map(matrices[2], rel_pos_map)[indexes_left]\n",
    "    ALT_R = get_right_BND_map(matrices[2], rel_pos_map)[indexes_right]\n",
    "\n",
    "    return (assemple_BND_maps(REF_L, REF_R, rel_pos_map),\n",
    "            assemple_BND_maps(ALT_L, ALT_R, rel_pos_map))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe5dbd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code did not work well - the subplots would not always align based on the size of the gene subplot\n",
    "\n",
    "\n",
    "# import itertools\n",
    "\n",
    "bin_size = 2048\n",
    "target_length_cropped = 448\n",
    "\n",
    "map_length = bin_size * target_length_cropped\n",
    "\n",
    "\n",
    "def pcolormesh_45deg(ax, mat, lines, linewidth, *args, **kwargs):\n",
    "    #https://stackoverflow.com/questions/12848581/is-there-a-way-to-rotate-a-matplotlib-plot-by-45-degrees\n",
    "    n = mat.shape[0]\n",
    "    # create rotation/scaling matrix\n",
    "    t = np.array([[1,0.5],[-1,0.5]])\n",
    "    # create coordinate matrix and transform it\n",
    "    A = np.dot(np.array([(i[1],i[0]) for i in itertools.product(range(n,-1,-1),range(0,n+1,1))]),t)\n",
    "    # plot\n",
    "    im = ax.pcolormesh(A[:,1].reshape(n+1,n+1),A[:,0].reshape(n+1,n+1),np.flipud(mat),*args, **kwargs)\n",
    "    im.set_rasterized(True)\n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            ax.plot([line,line + (448-line)/2], [0,448-line], color = 'gray', linewidth=linewidth, linestyle = 'dashed')\n",
    "            ax.plot([line,line/2], [0,line], color = 'gray', linewidth=linewidth, linestyle = 'dashed')\n",
    "    ax.set_ylim(0,n)\n",
    "    ax.set_ylim(0,target_length_cropped)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    _ = ax.set_xticks([])\n",
    "    _ = ax.set_yticks([])\n",
    "    ax.plot([0, n/2], [0, n], 'k-',linewidth=1)\n",
    "    ax.plot([n/2, n], [n, 0], 'k-',linewidth=1)\n",
    "    \n",
    "    ax.set_aspect(.5)\n",
    "    return im\n",
    "\n",
    "\n",
    "def plot_maps(mat1, mat2, lines, disruption_track, genes_in_map, size):\n",
    "    \n",
    "    gene_track = list(genes_in_map[['Start', 'width']].to_records(index = False))\n",
    "    linestyle = 'dashed'\n",
    "    linewidth = size/10*1.5\n",
    "    color = 'gray'\n",
    "    \n",
    "    nrows = 8.5\n",
    "    fig, (ax1,ax2,ax3,ax4,ax5) = plt.subplots(nrows=5,ncols=1,figsize=(size/2,size), \n",
    "                                               sharex = True,\n",
    "                                               gridspec_kw={'height_ratios': [2/nrows, 2/nrows, 2/nrows, \n",
    "                                                                              1/nrows, 1.5/nrows]})\n",
    "\n",
    "    pcolormesh_45deg(ax1, mat1, lines, linewidth, cmap= 'RdBu_r', vmax=2, vmin=-2)\n",
    "    pcolormesh_45deg(ax2, mat2, lines, linewidth, cmap= 'RdBu_r', vmax=2, vmin=-2)\n",
    "    pcolormesh_45deg(ax3, mat1 - mat2, lines, linewidth, cmap= 'PRGn', vmax=1, vmin=-1)\n",
    "    \n",
    "    ax4.plot(list(range(target_length_cropped)), disruption_track, color = 'black', linewidth = linewidth)\n",
    "#     plt.ylabel(scoring_method, rotation = 90)\n",
    "    plt.xlim([0,target_length_cropped])\n",
    "    ax4.set_xticks([])\n",
    "    ax4.set_yticks([])\n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            ax4.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "\n",
    "    \n",
    "    ax5.broken_barh(gene_track[::3], (14.5, 0.5), facecolors='tab:blue')\n",
    "    ax5.broken_barh(gene_track[1:][::3], (9.5, 0.5), facecolors='tab:blue')\n",
    "    ax5.broken_barh(gene_track[2:][::3], (4.5, 0.5), facecolors='tab:blue')\n",
    "    \n",
    "    bar_locations = ([14.5, 9.5, 4.5]*math.ceil(len(genes_in_map)/3))[:len(genes_in_map)]\n",
    "    for i in range(len(genes_in_map)):\n",
    "        \n",
    "        bar_location = bar_locations[i]\n",
    "        \n",
    "        gene = genes_in_map.loc[i,'Gene']\n",
    "        location = genes_in_map.loc[i,'Start']\n",
    "        \n",
    "        ax5.annotate(gene, (location,bar_location), # annotate gene at the start\n",
    "                     rotation = 45, \n",
    "                     ha = \"right\", va = \"top\", # set horizontal and vertical alignment\n",
    "                     annotation_clip = False, # keep annotation if outside of window\n",
    "                     fontsize = size/10*14) \n",
    "            \n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.ylabel('Genes',rotation=90)\n",
    "    plt.ylim([0,15])\n",
    "    plt.xlim([0,448])\n",
    "    plt.axis('off')\n",
    "    \n",
    "    fig.tight_layout()\n",
    "    plt.show()\n",
    "    \n",
    "\n",
    "plot_maps(REF_pred, ALT_pred, lines, disruption_track, genes_in_map, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a99b692",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_maps_genes_tracks(REF_pred, ALT_pred, genes_in_map, lines, disruption_track, scoring_method):\n",
    "    \n",
    "    '''\n",
    "    Plot the reference and alternate predicted contact frequency maps with lines at the beginning and end of the variant. For duplications, there will be 3 lines marking the two regions that are duplicates. \n",
    "    \n",
    "    Plot genes that match the regions in the reference map.\n",
    "    \n",
    "    Plot disruption score tracks.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    gene_track = list(genes_in_map[['Start', 'width']].to_records(index = False))\n",
    "    \n",
    "    vmin = -2\n",
    "    vmax = 2\n",
    "    linestyle = 'dashed'\n",
    "    linewidth = 0.3\n",
    "    color = 'black'\n",
    "\n",
    "    plt.rcParams['font.size']= 10\n",
    "    plot_width = 3\n",
    "    plot_width1D = 1\n",
    "    fig, gs = gridspec_inches([plot_width], [plot_width, .15, plot_width, .15, \n",
    "                                             plot_width1D, .15, plot_width1D*2])\n",
    "\n",
    "\n",
    "    plt.subplot(gs[0,0])\n",
    "    plt.matshow(REF_pred, fignum=False, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    plt.ylabel('Reference matrix',rotation=90)\n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "            plt.axhline(y=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.gca().yaxis.tick_right()\n",
    "    plt.xticks([])\n",
    "\n",
    "    plt.subplot(gs[2,0])\n",
    "    plt.matshow(ALT_pred, fignum=False, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    plt.ylabel('Alternate matrix',rotation=90)\n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "            plt.axhline(y=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.gca().yaxis.tick_right()\n",
    "    plt.xticks([])\n",
    "    \n",
    "    plt.subplot(gs[4,0])\n",
    "    plt.plot(list(range(target_length_cropped)), disruption_track)\n",
    "    plt.ylabel(scoring_method, rotation = 90)\n",
    "    plt.xlim([0,target_length_cropped])\n",
    "    plt.gca().yaxis.tick_right()\n",
    "    plt.xticks([])\n",
    "    \n",
    "\n",
    "    ax1 = fig.add_subplot(gs[6,0])\n",
    "            \n",
    "    ax1.broken_barh(gene_track[::3], (14.5, 0.5), facecolors='tab:blue')\n",
    "    ax1.broken_barh(gene_track[1:][::3], (9.5, 0.5), facecolors='tab:blue')\n",
    "    ax1.broken_barh(gene_track[2:][::3], (4.5, 0.5), facecolors='tab:blue')\n",
    "    \n",
    "    bar_locations = ([14.5, 9.5, 4.5]*math.ceil(len(genes_in_map)/3))[:len(genes_in_map)]\n",
    "    for i in range(len(genes_in_map)):\n",
    "        \n",
    "        bar_location = bar_locations[i]\n",
    "        \n",
    "        gene = genes_in_map.loc[i,'Gene']\n",
    "        location = genes_in_map.loc[i,'Start']\n",
    "        \n",
    "        ax1.annotate(gene, (location,bar_location), # annotate gene at the start\n",
    "                     rotation = 45, \n",
    "                     ha = \"right\", va = \"top\", # set horizontal and vertical alignment\n",
    "                     annotation_clip = False, # keep annotation if outside of window\n",
    "                     fontsize = 8) \n",
    "            \n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.ylabel('Genes',rotation=90)\n",
    "    plt.ylim([0,15])\n",
    "    plt.xlim([0,448])\n",
    "    plt.axis('off')\n",
    "\n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def plot_maps_genes(REF_pred, ALT_pred, genes_in_map, lines):\n",
    "\n",
    "        \n",
    "    '''\n",
    "    Plot the reference and alternate predicted contact frequency maps with lines at the beginning and end of the variant. For duplications, there will be 3 lines marking the two regions that are duplicates. \n",
    "    \n",
    "    Plot genes that match the regions in the reference map.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    \n",
    "    gene_track = list(genes_in_map[['Start', 'width']].to_records(index = False))\n",
    "    \n",
    "    vmin = -2\n",
    "    vmax = 2\n",
    "    linestyle = 'dashed'\n",
    "    linewidth = 0.3\n",
    "    color = 'black'\n",
    "\n",
    "    plt.rcParams['font.size']= 10\n",
    "    plot_width = 3\n",
    "    plot_width1D = 1\n",
    "    fig, gs = gridspec_inches([plot_width], [plot_width, .15, plot_width, .15, plot_width1D])\n",
    "\n",
    "\n",
    "    plt.subplot(gs[0,0])\n",
    "    plt.matshow(REF_pred, fignum=False, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    plt.ylabel('Reference matrix',rotation=90)\n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "            plt.axhline(y=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.gca().yaxis.tick_right()\n",
    "    plt.xticks([])\n",
    "\n",
    "    plt.subplot(gs[2,0])\n",
    "    plt.matshow(ALT_pred, fignum=False, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    plt.ylabel('Alternate matrix',rotation=90)\n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "            plt.axhline(y=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.gca().yaxis.tick_right()\n",
    "    plt.xticks([])\n",
    "\n",
    "    ax1 = fig.add_subplot(gs[4,0])\n",
    "            \n",
    "    ax1.broken_barh(gene_track[::3], (14.5, 0.5), facecolors='tab:blue')\n",
    "    ax1.broken_barh(gene_track[1:][::3], (9.5, 0.5), facecolors='tab:blue')\n",
    "    ax1.broken_barh(gene_track[2:][::3], (4.5, 0.5), facecolors='tab:blue')\n",
    "    \n",
    "    bar_locations = ([14.5, 9.5, 4.5]*math.ceil(len(genes_in_map)/3))[:len(genes_in_map)]\n",
    "    for i in range(len(genes_in_map)):\n",
    "        \n",
    "        bar_location = bar_locations[i]\n",
    "        \n",
    "        gene = genes_in_map.loc[i,'Gene']\n",
    "        location = genes_in_map.loc[i,'Start']\n",
    "        \n",
    "        ax1.annotate(gene, (location,bar_location), # annotate gene at the start\n",
    "                     rotation = 45, \n",
    "                     ha = \"right\", va = \"top\", # set horizontal and vertical alignment\n",
    "                     annotation_clip = False, # keep annotation if outside of window\n",
    "                     fontsize = 8) \n",
    "            \n",
    "    if lines is not None:\n",
    "        for line in lines:\n",
    "            plt.axvline(x=line, color=color, linestyle=linestyle, linewidth = linewidth)\n",
    "    plt.ylabel('Genes',rotation=90)\n",
    "    plt.ylim([0,15])\n",
    "    plt.xlim([0,448])\n",
    "    plt.axis('off')\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de55b3ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removed the following from get_scores_utils.py after changing masking function\n",
    "\n",
    "get_variant_position(CHR, POS, var_len, half_left, half_right) # same as in get_seq\n",
    "\n",
    "\n",
    "\n",
    "def get_variant_type(REF, ALT):\n",
    "    \n",
    "    '''\n",
    "    Get the type of variant from a REF and ALT allele of sequences.\n",
    "    One of 4 variant types:\n",
    "    1. Deletion\n",
    "    2. Insertion\n",
    "    3. SNP: single nucleotide variant\n",
    "    4. MNP: multiple nucleotide variant\n",
    "    \n",
    "    '''\n",
    "\n",
    "    # Annotate variant as one of the 6 categories below based on REF and ALT allele\n",
    "    \n",
    "    if len(REF) > len(ALT):\n",
    "        variant_type = \"Deletion\"\n",
    "    elif len(REF) < len(ALT):\n",
    "        variant_type = \"Insertion\"\n",
    "        \n",
    "    elif len(REF) == 1 and len(ALT) ==1:\n",
    "        variant_type = \"SNP\"\n",
    "    elif len(REF) == len(ALT) and len(REF) != 1:\n",
    "        variant_type = \"MNP\"\n",
    "    \n",
    "    return variant_type\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "731097db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting function for when there might be more than 2 maps\n",
    "\n",
    "def plot_maps(maps, genes_in_map):\n",
    "\n",
    "    vmin = -2\n",
    "    vmax = 2\n",
    "\n",
    "    plt.rcParams['font.size']= 10\n",
    "    plot_width = 3\n",
    "    plot_width1D = 1\n",
    "    if len(maps) == 2:\n",
    "        fig, gs = gridspec_inches([plot_width], [plot_width, .15, plot_width, .15, plot_width1D])\n",
    "#     if len(maps) == 3:\n",
    "#         fig, gs = gridspec_inches([plot_width], [plot_width, .15, plot_width, .15, plot_width, .15, plot_width1D])\n",
    "\n",
    "\n",
    "    for i in range(len(maps)):\n",
    "        \n",
    "        plt.subplot(gs[i*2,0])\n",
    "        plt.matshow(maps[i], fignum=False, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "        plt.gca().yaxis.tick_right()\n",
    "        plt.xticks([])\n",
    "\n",
    "   \n",
    "    gene_track = list(genes_in_map[['Start', 'width']].to_records(index = False))\n",
    "\n",
    "    ax1 = fig.add_subplot(gs[len(maps)*2+2,0])\n",
    "\n",
    "    ax1.broken_barh(gene_track[::2], (9.5, 0.5), facecolors='tab:blue')\n",
    "    ax1.broken_barh(gene_track[1:][::2], (4.5, 0.5), facecolors='tab:blue')\n",
    "\n",
    "    for i in range(len(genes_in_map)):\n",
    "\n",
    "        if (i % 2) == 0:\n",
    "            bar_location = 9.5\n",
    "        else:\n",
    "            bar_location = 4.5\n",
    "\n",
    "        gene = genes_in_map.loc[i,'Gene']\n",
    "        location = genes_in_map.loc[i,'Start']\n",
    "\n",
    "        ax1.annotate(gene, (location,bar_location), # annotate gene at the start\n",
    "                     rotation = 45, \n",
    "                     ha = \"right\", va = \"top\", # set horizontal and vertical alignment\n",
    "                     annotation_clip = False, # keep annotation if outside of window\n",
    "                     fontsize = 6) \n",
    "\n",
    "    plt.ylabel('Genes',rotation=90)\n",
    "    plt.ylim([0,10])\n",
    "    plt.xlim([0,target_length_cropped])\n",
    "    plt.axis('off')\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "5adb8ebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mask_matrices(CHR, POS, REF, ALT, REF_pred, ALT_pred, shift):\n",
    "\n",
    "    '''\n",
    "    This applied to non-BND predicted matrices.\n",
    "    \n",
    "    Mask reference and alternate predicted matrices based on the type of variant.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    variant_type = get_variant_type(REF, ALT)\n",
    "    \n",
    "    # Get last coordinate of chromosome\n",
    "    chrom_max = int(chrom_lengths[chrom_lengths.CHROM == CHR[3:]]['chrom_max']) \n",
    "\n",
    "    # Get centromere coordinate\n",
    "    centro_start = int(centromere_coords[centromere_coords.chrom == CHR]['start'])\n",
    "    centro_stop = int(centromere_coords[centromere_coords.chrom == CHR]['end'])\n",
    "\n",
    "    \n",
    "    # Insertions: Mask REF, add nans if necessary, and mirror nans to ALT\n",
    "    if variant_type in [\"Insertion\", \"Deletion\"]:\n",
    "        \n",
    "        if variant_type == \"Deletion\":\n",
    "            # this works the same exact way but the commands are swapped\n",
    "            REF, ALT = ALT, REF\n",
    "            REF_pred, ALT_pred = ALT_pred, REF_pred\n",
    "\n",
    "        # Get REF allele sections\n",
    "        REF_len = len(REF)\n",
    "\n",
    "        REF_half_left = math.ceil((seq_length - REF_len)/2) - shift # if the REF allele is odd, shift right\n",
    "        REF_half_right = math.floor((seq_length - REF_len)/2) + shift\n",
    "\n",
    "        \n",
    "        # Get ALT allele sections\n",
    "        ALT_len = len(ALT)\n",
    "        \n",
    "        ALT_half_left = math.ceil((seq_length - ALT_len)/2) - shift\n",
    "        ALT_half_right = math.floor((seq_length - ALT_len)/2) + shift\n",
    "        \n",
    "        \n",
    "        # Annotate whether variant is close to beginning or end of chromosome\n",
    "        var_position = get_variant_position(CHR, POS, REF_len, REF_half_left, REF_half_right)\n",
    "\n",
    "\n",
    "        # Get start and end bins of REF and ALT alleles\n",
    "        if var_position == \"chrom_mid\":\n",
    "            \n",
    "            var_start = get_bin(REF_half_left - 1)\n",
    "            var_end = get_bin(REF_half_left - 1 + REF_len)\n",
    "            \n",
    "            var_start_ALT = get_bin(ALT_half_left - 1)\n",
    "            var_end_ALT = get_bin(ALT_half_left - 1 + ALT_len)\n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            \n",
    "            var_start = get_bin(POS - 1 - abs(shift))\n",
    "            var_end = get_bin(POS - 1 + REF_len - abs(shift))\n",
    "            \n",
    "            var_start_ALT = var_start\n",
    "            var_end_ALT = get_bin(POS - 1 + ALT_len - abs(shift))\n",
    "\n",
    "        elif var_position == \"chrom_centro_left\": \n",
    "            \n",
    "            var_start = get_bin(POS - (centro_start - seq_length - abs(shift)) - 1)\n",
    "            var_end = get_bin(POS - (centro_start - seq_length - abs(shift)) - 1 + REF_len)\n",
    "            \n",
    "            var_start_ALT = get_bin(POS - (centro_start - seq_length - abs(shift)) - 1 - ALT_len)\n",
    "            var_end_ALT = var_end\n",
    "\n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            \n",
    "            var_start = get_bin(POS - centro_stop - 1 - abs(shift))\n",
    "            var_end = get_bin(POS - centro_stop - 1 + REF_len - abs(shift))\n",
    "            \n",
    "            var_start_ALT = var_start\n",
    "            var_end_ALT = get_bin(POS - centro_stop - 1 + ALT_len - abs(shift))\n",
    "\n",
    "        elif var_position == \"chrom_end\": \n",
    "            \n",
    "            var_start = get_bin(POS - (chrom_max - seq_length - abs(shift)) - 1)\n",
    "            var_end = get_bin(POS - (chrom_max - seq_length - abs(shift)) - 1 + REF_len)\n",
    "            \n",
    "            var_start_ALT = get_bin(POS - (chrom_max - seq_length - abs(shift)) - 1 - ALT_len)\n",
    "            var_end_ALT = var_end\n",
    "\n",
    "\n",
    "        # Mask REF map: make variant bin(s) nan and add empty bins at the variant if applicable\n",
    "        \n",
    "        REF_pred_masked = REF_pred.copy()\n",
    "\n",
    "        REF_pred_masked[var_start:var_end + 1, :] = np.nan\n",
    "        REF_pred_masked[:, var_start:var_end + 1] = np.nan\n",
    "  \n",
    "        \n",
    "        # If the ALT allele falls on more bins than the REF allele, adjust ALT allele \n",
    "            # (add nan(s) to var and remove outside bin(s))\n",
    "            # Otherwise don't mask\n",
    "        \n",
    "        if var_end_ALT - var_start_ALT > var_end - var_start:\n",
    "            \n",
    "        \n",
    "            # Insert the rest of the nans corresponding to the ALT allele\n",
    "            to_add = (var_end_ALT - var_start_ALT) - (var_end - var_start)\n",
    "\n",
    "            for j in range(var_start, var_start + to_add): # range only includes the first variable \n",
    "                REF_pred_masked = np.insert(REF_pred_masked, j, np.nan, axis = 0)\n",
    "                REF_pred_masked = np.insert(REF_pred_masked, j, np.nan, axis = 1)\n",
    "\n",
    "            # Chop off the outside of the REF matrix \n",
    "            to_remove = len(REF_pred_masked) - 448\n",
    "\n",
    "            if var_position == \"chrom_mid\":\n",
    "                # remove less on the left bc that's where you put one less part of the variant with odd number of bp\n",
    "                REF_pred_masked = REF_pred_masked[math.floor(to_remove/2) : -math.ceil(to_remove/2), \n",
    "                                                  math.floor(to_remove/2) : -math.ceil(to_remove/2)]\n",
    "                \n",
    "            elif var_position in [\"chrom_start\", \"chrom_centro_right\"]: \n",
    "                # Remove all from the right\n",
    "                REF_pred_masked = REF_pred_masked[: -to_remove, \n",
    "                                                  : -to_remove]\n",
    "\n",
    "            elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "                # Remove all from the left\n",
    "                REF_pred_masked = REF_pred_masked[to_remove :, \n",
    "                                                  to_remove :]\n",
    "\n",
    "            assert len(REF_pred_masked) == 448, 'Masked reference matrix is not the right size.'\n",
    "            \n",
    "            \n",
    "        \n",
    "\n",
    "        # Mask ALT map: make all nans in REF_pred also nan in ALT_pred\n",
    "        \n",
    "        REF_pred_novalues = REF_pred_masked.copy()\n",
    "\n",
    "        REF_pred_novalues[np.invert(np.isnan(REF_pred_novalues))] = 0\n",
    "\n",
    "        ALT_pred_masked = ALT_pred + REF_pred_novalues\n",
    "\n",
    "        assert len(ALT_pred_masked) == 448, 'Masked alternate matrix is not the right size.'\n",
    "        \n",
    "        if variant_type == \"Deletion\":\n",
    "            # Swap back\n",
    "            REF_pred_masked, ALT_pred_masked = ALT_pred_masked, REF_pred_masked\n",
    "        \n",
    "    \n",
    "    # SNPs or MNPs: Mask REF and mirror nans to ALT\n",
    "    elif variant_type in ['SNP', 'MNP']:\n",
    "        \n",
    "        # Get REF allele sections\n",
    "        REF_len = len(REF)\n",
    "\n",
    "        REF_half_left = math.ceil((seq_length - REF_len)/2)  - shift # if the REF allele is odd, shift right\n",
    "        REF_half_right = math.floor((seq_length - REF_len)/2) + shift\n",
    "\n",
    "\n",
    "        # Annotate whether variant is close to beginning or end of chromosome\n",
    "        var_position = get_variant_position(CHR, POS, REF_len, REF_half_left, REF_half_right)\n",
    "\n",
    "\n",
    "        # Get start and end bins of REF and ALT alleles\n",
    "        if var_position == \"chrom_mid\":\n",
    "            \n",
    "            var_start = get_bin(REF_half_left - 1)\n",
    "            var_end = get_bin(REF_half_left - 1 + REF_len)\n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            \n",
    "            var_start = get_bin(POS - 1 + abs(shift))\n",
    "            var_end = get_bin(POS - 1 + REF_len + abs(shift))\n",
    "\n",
    "        elif var_position == \"chrom_centro_left\": \n",
    "            \n",
    "            var_start = get_bin(POS - (centro_start - seq_length - abs(shift)) - 1)\n",
    "            var_end = get_bin(POS - (centro_start - seq_length - abs(shift)) - 1 + REF_len)\n",
    "\n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            \n",
    "            var_start = get_bin(POS - centro_stop - 1 + abs(shift))\n",
    "            var_end = get_bin(POS - centro_stop - 1 + REF_len + abs(shift))\n",
    "\n",
    "        elif var_position == \"chrom_end\": \n",
    "            \n",
    "            var_start = get_bin(POS - (chrom_max - seq_length - abs(shift)) - 1)\n",
    "            var_end = get_bin(POS - (chrom_max - seq_length - abs(shift)) - 1 + REF_len)\n",
    "            \n",
    "            \n",
    "        # Mask REF map: make variant bin(s) nan \n",
    "        \n",
    "        REF_pred_masked = REF_pred.copy()\n",
    "\n",
    "        REF_pred_masked[var_start:var_end + 1, :] = np.nan\n",
    "        REF_pred_masked[:, var_start:var_end + 1] = np.nan\n",
    "\n",
    "        \n",
    "        # Mask ALT map: make all nans in REF_pred also nan in ALT_pred\n",
    "        \n",
    "        REF_pred_novalues = REF_pred_masked.copy()\n",
    "\n",
    "        REF_pred_novalues[np.invert(np.isnan(REF_pred_novalues))] = 0\n",
    "\n",
    "        ALT_pred_masked = ALT_pred + REF_pred_novalues\n",
    "\n",
    "        assert len(ALT_pred_masked) == 448, 'Masked alternate matrix is not the right size.'\n",
    "        \n",
    "    \n",
    "    \n",
    "    return REF_pred_masked, ALT_pred_masked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6606e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scores(CHR, POS, REF, ALT, sequences, SVTYPE, scores, shift, revcomp, get_tracks, get_maps): \n",
    "    \n",
    "    '''\n",
    "    Get disruption scores, disruption tracks, and/or predicted maps from variants and the sequences generated from them.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    \n",
    "    var_rel_pos = sequences[-1]\n",
    "    \n",
    "    # Error if variant position is too close to end of prediction window\n",
    "    if any([int(x) <= bin_size*32 or int(x) >= seq_length - bin_size*32 for x in var_rel_pos]):\n",
    "        raise ValueError('Variant outside prediction window after cropping')\n",
    "    \n",
    "    \n",
    "    # Make prediction\n",
    "    sequences = [x for x in sequences if type(x) == str]\n",
    "    matrices = [mat_from_vector(vector) for vector in [vector_from_seq(seq) for seq in sequences]]\n",
    "    \n",
    "    if revcomp:\n",
    "        matrices = [np.flipud(np.fliplr(x)) for x in matrices]\n",
    "    \n",
    "                \n",
    "    # Get processed matrices to score\n",
    "                     \n",
    "    if len(REF) > bin_size/2 or len(ALT) > bin_size/2:\n",
    "\n",
    "        # mask matrices\n",
    "        matrices = mask_matrices(CHR, POS, REF, ALT, matrices[0], matrices[1], shift)\n",
    "    \n",
    "    if SVTYPE == \"BND\":  \n",
    "        \n",
    "        BND_rel_pos_map = round(var_rel_pos[0]/bin_size - 32)\n",
    "        \n",
    "        # Get REF and ALT vectors, excluding diagonal \n",
    "        indexes_left = np.triu_indices(BND_rel_pos_map, 2)\n",
    "        indexes_right = np.triu_indices(target_length_cropped - BND_rel_pos_map, 2)\n",
    "\n",
    "        REF_L = get_left_BND_map(matrices[0], BND_rel_pos_map)[indexes_left]\n",
    "        REF_R = get_right_BND_map(matrices[1], BND_rel_pos_map)[indexes_right]\n",
    "        ALT_L = get_left_BND_map(matrices[2], BND_rel_pos_map)[indexes_left]\n",
    "        ALT_R = get_right_BND_map(matrices[2], BND_rel_pos_map)[indexes_right]\n",
    "\n",
    "        matrices = (assemple_BND_mats(REF_L, REF_R, BND_rel_pos_map),\n",
    "                    assemple_BND_mats(ALT_L, ALT_R, BND_rel_pos_map))\n",
    "\n",
    "\n",
    "    # Get disruption score and correlation for this variant\n",
    "    scores_results = {}\n",
    "    \n",
    "    if get_maps:\n",
    "        scores_results['maps'] = [matrices[0], matrices[1]]\n",
    "        \n",
    "        \n",
    "    for score in scores:\n",
    "        scores_results[score] = getattr(scoring_map_methods(matrices[0], matrices[1]), \n",
    "                                        score)()\n",
    "        \n",
    "        if get_tracks and score in ['corr', 'mse']:\n",
    "            scores_results[f'{score}_track'] = getattr(scoring_map_methods(matrices[0], matrices[1]), \n",
    "                                                       f'{score}_track')()\n",
    "            \n",
    "\n",
    "    return scores_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f333e714",
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_seq_ends_BND(CHR, position, adjust, shift):\n",
    "           \n",
    "    '''\n",
    "    Get start (adjust = 1) or end (adjust = MB) of sequence for prediction based on variant position \\\n",
    "    with respect to chromosome arm ends (defined in get_variant_position function).\n",
    "    Different from adjust_seq_ends because it does not require centro_start, centro_stop, and chrom_max as input.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    if position == 'chrom_start':\n",
    "        seq_pos = adjust + 1 + abs(shift) # 1 is added so the position is never 0. coordinates are 1-based\n",
    "\n",
    "    elif position == 'chrom_centro_right':\n",
    "        seq_pos = int(centromere_coords[centromere_coords.CHROM == CHR]['centro_stop']) + adjust + abs(shift) \n",
    "\n",
    "    elif position == 'chrom_end':\n",
    "        seq_pos = int(chrom_lengths[chrom_lengths.CHROM == CHR[3:]]['chrom_max']) - MB + adjust - abs(shift) \n",
    "\n",
    "    elif position == 'chrom_centro_left':\n",
    "        seq_pos = int(centromere_coords[centromere_coords.CHROM == CHR]['centro_start']) - MB + adjust - abs(shift) \n",
    "        \n",
    "    return seq_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "a76af838",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(CHR, POS, REF, ALT, shift, revcomp: bool):\n",
    "  \n",
    "    '''\n",
    "    Get reference and alternate sequence for prediction from REF and ALT alleles by incorporating ALT into the reference genome.\n",
    "    Requires ALT allele to be a sequence and not a symbolic allele.\n",
    "    Use positive sign for a right shift and negative for a left shift.\n",
    "    revcomp: Take the reverse compliment of the resulting sequence.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    # Get reference sequence\n",
    "    \n",
    "    REF_len = len(REF)\n",
    "\n",
    "    REF_half_left = math.ceil((MB - REF_len)/2) - shift # if the REF allele is odd, shift right\n",
    "    REF_half_right = math.floor((MB - REF_len)/2) + shift\n",
    "\n",
    "    \n",
    "    # Annotate whether variant position with respect to chromosome arms ends\n",
    "    if len(REF) <= len(ALT):\n",
    "        var_position = get_variant_position(CHR, POS, REF_len, REF_half_left, REF_half_right)\n",
    "  \n",
    "    elif len(REF) > len(ALT):       \n",
    "        ALT_len = len(ALT)\n",
    "        ALT_half_left = math.ceil((MB - ALT_len)/2) - shift\n",
    "        ALT_half_right = math.floor((MB - ALT_len)/2) + shift   \n",
    "        var_position = get_variant_position(CHR, POS, ALT_len, ALT_half_left, ALT_half_right)\n",
    "    \n",
    "\n",
    "    # Get last coordinate of chromosome\n",
    "    chrom_max = int(chrom_lengths[chrom_lengths.CHROM == CHR[3:]]['chrom_max'])\n",
    "    \n",
    "    # Get centromere coordinates\n",
    "    centro_start = int(centromere_coords[centromere_coords.CHROM == CHR]['centro_start']) + 1\n",
    "    centro_stop = int(centromere_coords[centromere_coords.CHROM == CHR]['centro_stop']) + 1\n",
    "    \n",
    "    \n",
    "    # Get start and end of reference sequence\n",
    "    if var_position == \"chrom_mid\":\n",
    "        REF_start = POS - REF_half_left\n",
    "        REF_stop = REF_start + MB \n",
    "    elif var_position == \"centromere\":\n",
    "        raise ValueError('Centromeric variant')\n",
    "    else:\n",
    "        REF_start = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, 0, shift)\n",
    "        REF_stop = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, MB, shift)\n",
    "        print(\"Warning: Variant not centered; too close to chromosome arm ends.\")\n",
    "        \n",
    "        \n",
    "    # Get reference sequence\n",
    "    REF_seq = fasta_open.fetch(CHR, REF_start - 1, REF_stop - 1).upper()\n",
    "\n",
    "\n",
    "    # Error if N composition is more than 5% of sequence\n",
    "    if Counter(REF_seq)['N']/MB*100 > 5:\n",
    "        raise ValueError('N composition greater than 5%')\n",
    "\n",
    "\n",
    "\n",
    "    # Error if reference sequence does not match given REF\n",
    "\n",
    "    if var_position == \"chrom_mid\":\n",
    "        var_rel_pos_REF = REF_half_left\n",
    "\n",
    "    elif var_position == \"chrom_start\": \n",
    "        var_rel_pos_REF = POS - abs(shift) - 1\n",
    "\n",
    "    elif var_position == \"chrom_centro_right\": \n",
    "        var_rel_pos_REF = POS - centro_stop - abs(shift)\n",
    "\n",
    "    elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "        var_rel_pos_REF = -(REF_stop - POS)\n",
    "\n",
    "\n",
    "    if REF_seq[var_rel_pos_REF : var_rel_pos_REF + REF_len] != REF:\n",
    "        raise ValueError('Reference allele does not match hg38.')\n",
    "            \n",
    "            \n",
    "            \n",
    "    # Error if reference sequence is not the right length      \n",
    "    if len(REF_seq) != MB:\n",
    "        raise ValueError('Reference sequence generated is not the right length.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # For SNPs, MNPs, Insertions: \n",
    "    if len(REF) <= len(ALT):\n",
    "\n",
    "        # Create alternate sequence: change REF sequence at position from REF to ALT\n",
    "\n",
    "        ALT_seq = REF_seq\n",
    "\n",
    "        ALT_seq = ALT_seq[:var_rel_pos_REF] + ALT + ALT_seq[var_rel_pos_REF + REF_len:]\n",
    "\n",
    "\n",
    "        var_rel_pos_ALT = var_rel_pos_REF\n",
    "        \n",
    "        # Chop off ends of alternate sequence if it's longer \n",
    "        if len(ALT_seq) > len(REF_seq):\n",
    "            to_remove = (len(ALT_seq) - len(REF_seq))/2\n",
    "\n",
    "            if to_remove == 0.5:\n",
    "                ALT_seq = ALT_seq[1:]\n",
    "                var_rel_pos_ALT = var_rel_pos_REF - 1\n",
    "                \n",
    "            else:\n",
    "                ALT_seq = ALT_seq[math.ceil(to_remove) : -math.floor(to_remove)]\n",
    "                var_rel_pos_ALT = var_rel_pos_REF - math.ceil(to_remove)\n",
    "                \n",
    "            \n",
    "\n",
    "\n",
    "    # For Deletions\n",
    "    elif len(REF) > len(ALT):\n",
    "\n",
    "\n",
    "        del_len = len(REF) - len(ALT)\n",
    "        \n",
    "        to_add_left = math.ceil(del_len/2)\n",
    "        to_add_right = math.floor(del_len/2) \n",
    "\n",
    "        # Get start and end of reference sequence\n",
    "        if var_position == \"chrom_mid\":\n",
    "            ALT_start = REF_start - to_add_left\n",
    "            ALT_stop = REF_stop + to_add_right\n",
    "\n",
    "        elif var_position in [\"chrom_start\", \"chrom_centro_right\"]: \n",
    "            ALT_start = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, 0, shift)\n",
    "            ALT_stop = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, MB + del_len, shift)\n",
    "            \n",
    "        elif var_position in [\"chrom_centro_left\", \"chrom_end\"]: \n",
    "            ALT_start = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, 0 - del_len, shift)\n",
    "            ALT_stop = adjust_seq_ends(centro_start, centro_stop, chrom_max, var_position, MB, shift)\n",
    "        \n",
    "        \n",
    "        \n",
    "        # Get alternate sequence\n",
    "        ALT_seq = fasta_open.fetch(CHR, ALT_start - 1, ALT_stop - 1).upper()\n",
    "        \n",
    "        \n",
    "        \n",
    "        # Error if alternate sequence does not match REF at POS\n",
    "\n",
    "        if var_position == \"chrom_mid\":\n",
    "            var_rel_pos_ALT = REF_half_left + to_add_left\n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            var_rel_pos_ALT = POS - abs(shift) - 1\n",
    "            \n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            var_rel_pos_ALT = POS - centro_stop - abs(shift)\n",
    "            \n",
    "        elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "            var_rel_pos_ALT = -(REF_stop - POS)\n",
    "                \n",
    "                \n",
    "        if ALT_seq[var_rel_pos_ALT : var_rel_pos_ALT + REF_len] != REF:\n",
    "            raise ValueError('Sequence for the alternate allele does not match hg38 at REF position.')\n",
    "\n",
    "\n",
    "    \n",
    "        # Change alternate sequence to match ALT at POS\n",
    "\n",
    "        if var_position == \"chrom_mid\":\n",
    "            ALT_seq = ALT_seq[:var_rel_pos_ALT - 1] + ALT + ALT_seq[var_rel_pos_ALT + REF_len - 1:] \n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            ALT_seq = ALT_seq[:var_rel_pos_ALT + 1] + ALT + ALT_seq[var_rel_pos_ALT + 1 + REF_len:]\n",
    "            \n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            ALT_seq = ALT_seq[:var_rel_pos_ALT] + ALT + ALT_seq[var_rel_pos_ALT + REF_len:]\n",
    "            \n",
    "        elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "            ALT_seq = ALT_seq[:var_rel_pos_ALT] + ALT + ALT_seq[var_rel_pos_ALT + REF_len:]\n",
    "\n",
    "            \n",
    "    if len(ALT_seq) != MB:\n",
    "        raise ValueError('Alternate sequence generated is not the right length.')\n",
    "         \n",
    "            \n",
    "    # Take reverse compliment of sequence\n",
    "    if revcomp:\n",
    "        REF_seq, ALT_seq = [str(Seq(x).reverse_complement()) for x in [REF_seq, ALT_seq]]\n",
    "\n",
    "        \n",
    "    return REF_seq, ALT_seq, [var_rel_pos_REF, var_rel_pos_ALT]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd04c929",
   "metadata": {},
   "outputs": [],
   "source": [
    "# removed this completely \n",
    "\n",
    "def from_upper_triu(vector_repr, matrix_len, num_diags):\n",
    "    \n",
    "    '''\n",
    "    Get a matrix from a vector representatin of the upper triangle.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    z = np.zeros((matrix_len,matrix_len))\n",
    "    triu_tup = np.triu_indices(matrix_len,num_diags)\n",
    "    \n",
    "    z[triu_tup] = vector_repr\n",
    "    \n",
    "    for i in range(-num_diags+1,num_diags):\n",
    "        set_diag(z, np.nan, i)\n",
    "        \n",
    "    return z + z.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96dc5d3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser.add_argument('--chrom',\n",
    "                    dest = 'chrom_lengths', \n",
    "                    help = 'File with lengths of chromosomes in hg38. Columns: chromosome (ex: 1), length; no header.', \n",
    "                    type = str,\n",
    "                    default = 'data/chrom_lengths_hg38',\n",
    "                    required = False)\n",
    "\n",
    "parser.add_argument('--centro',\n",
    "                    dest = 'centromere_coords', \n",
    "                    help = 'Centromere coordinates for hg38. Columns: chromosome (ex: chr1), start, end; no header.', \n",
    "                    type = str,\n",
    "                    default = 'data/centromere_coords_hg38',\n",
    "                    required = False)\n",
    "\n",
    "\n",
    "parser.add_argument('--revcomp',\n",
    "                    dest = 'revcomp', \n",
    "                    help = '''Make predictions with the reverse compliment of the sequence.\n",
    "''', \n",
    "                    action='store_true',\n",
    "                    required = False)\n",
    "\n",
    "parser.add_argument('--no_revcomp',\n",
    "                    dest = 'no_revcomp', \n",
    "                    help = '''Make predictions without taking the reverse compliment of the sequence. If specified, --revcomp must also be specified.\n",
    "''', \n",
    "                    action='store_false',\n",
    "                    required = False)\n",
    "\n",
    "if not revcomp and not no_revcomp:\n",
    "    raise ValueError('Either revcomp and/or no_revcomp must be True.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd69c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Warning: Variant not centered; too close to start of chromosome.\")\n",
    "print(\"Warning: Variant not centered; too close to right end of centromere.\")\n",
    "print(\"Warning: Variant not centered; too close to end of chromosome.\")\n",
    "print(\"Warning: Variant not centered; too close to left end of centromere.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c426e1a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_MSE_from_vector(vector1, vector2):\n",
    "\n",
    "    '''\n",
    "    Get MSE between two vectors.\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    # A vecotr of sum of squared differences for each row of matrix \n",
    "    sub_vec = [x - y for x,y in zip(vector1, vector2)]\n",
    "    \n",
    "    # Get number of bins that are not nan (not including ones that overlap deletion)\n",
    "    non_nan_values = np.count_nonzero(np.invert(np.isnan(sub_vec)))\n",
    "    \n",
    "    MSE = np.nansum([x**2 for x in sub_vec])/non_nan_values\n",
    "\n",
    "    return MSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2bf7a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from read_input\n",
    "\n",
    "elif 'maf' in in_file:\n",
    "\n",
    "    # These are formatted like SVs\n",
    "\n",
    "    variants = (pd.read_csv(in_file, skiprows = 1, sep = '\\t')\n",
    "                .rename(columns = {'Chromosome':'CHROM', \n",
    "                                   'Start_Position':'POS',\n",
    "                                   'End_Position':'END',\n",
    "                                   'Reference_Allele':'REF',\n",
    "                                   'Tumor_Seq_Allele2':'ALT'})\n",
    "               [['CHROM', 'POS', 'END', 'REF', 'ALT']])\n",
    "    variants['SVLEN'] = (variants.END - variants.POS).astype('int') # this SVLEN (END-POS) would be 0 for SNPs\n",
    "\n",
    "    # Might need to do this if there are homozygous variants where Tumor_Seq_Allele1 != Reference_Allele\n",
    "    # would need to only do this for those variants and concat with the rest\n",
    "    # variants = pd.melt(variants, \n",
    "    #                    id_vars = ['CHROM', 'POS', 'REF'], \n",
    "    #                    value_vars = ['ALT1', 'ALT2'], \n",
    "    #                    var_name = 'allele', \n",
    "    #                    value_name='ALT')\n",
    "\n",
    "    for i in variants[variants.ALT == '-'].index:\n",
    "        variants.loc[i,'SVTYPE'] = 'DEL'\n",
    "    for i in variants[variants.REF == '-'].index:\n",
    "        variants.loc[i,'SVTYPE'] = 'INS'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eabbc4c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if SVTYPE == \"DEL\":\n",
    "\n",
    "    # Get REF and ALT allele sequences first\n",
    "    REF = fasta_open.fetch(CHR, POS - 1, END).upper()\n",
    "    ALT = REF[0]\n",
    "\n",
    "\n",
    "elif SVTYPE == \"DUP\":\n",
    "\n",
    "    # Insert duplicated sequence before POS\n",
    "    ALT = fasta_open.fetch(CHR, POS - 1, END).upper()\n",
    "    REF = ALT[0]\n",
    "\n",
    "\n",
    "elif SVTYPE == \"INV\":\n",
    "\n",
    "    REF = fasta_open.fetch(CHR, POS - 1, END).upper()\n",
    "    ALT = REF[0] + str(Seq(REF[1:]).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6858095",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scores(REF_pred, ALT_pred, scores): # change this to score on maps instead of vectors\n",
    "\n",
    "    scores_results = {}\n",
    "    for score in scores:\n",
    "        \n",
    "        if score == 'corr':\n",
    "            correlation, corr_pval = spearmanr(REF_pred, ALT_pred, nan_policy='omit')\n",
    "            if corr_pval >= 0.05:\n",
    "                correlation = 1\n",
    "            scores_results[score] = correlation\n",
    "            \n",
    "        elif score == 'mse':\n",
    "            mse = get_MSE_from_vector(REF_pred, ALT_pred)\n",
    "            \n",
    "        scores_results[score] = mse\n",
    "        \n",
    "    return scores_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59039752",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(CHR, POS, REF, ALT, shift):\n",
    "  \n",
    "    # Get reference and alternate sequence from REF and ALT allele using reference genome \n",
    "    # use positive sign for a right shift and negative for a left shift\n",
    "\n",
    "    # Get reference sequence\n",
    "    \n",
    "    REF_len = len(REF)\n",
    "\n",
    "    REF_half_left = math.ceil((MB - REF_len)/2) - shift # if the REF allele is odd, shift right\n",
    "    REF_half_right = math.floor((MB - REF_len)/2) + shift\n",
    "\n",
    "    \n",
    "    # Annotate whether variant is near end of chromosome arms\n",
    "    if len(REF) <= len(ALT): # For SNPs, MNPs, Insertions\n",
    "        var_position = get_variant_position(CHR, POS, REF_len, REF_half_left, REF_half_right)\n",
    "  \n",
    "    elif len(REF) > len(ALT): # For Deletions        \n",
    "        ALT_len = len(ALT)\n",
    "        ALT_half_left = math.ceil((MB - ALT_len)/2) - shift\n",
    "        ALT_half_right = math.floor((MB - ALT_len)/2) + shift   \n",
    "        var_position = get_variant_position(CHR, POS, ALT_len, ALT_half_left, ALT_half_right)\n",
    "    \n",
    "\n",
    "    # Get last coordinate of chromosome\n",
    "    chrom_max = int(chrom_lengths[chrom_lengths.CHROM == CHR[3:]]['chrom_max']) \n",
    "    \n",
    "    # Get centromere coordinate\n",
    "    centro_start = int(centromere_coords[centromere_coords.CHROM == CHR]['centro_start'])\n",
    "    centro_stop = int(centromere_coords[centromere_coords.CHROM == CHR]['centro_stop'])\n",
    "    \n",
    "    \n",
    "    # Get start and end of reference sequence\n",
    "\n",
    "    if var_position == \"chrom_mid\":\n",
    "        REF_start = POS - REF_half_left\n",
    "        REF_stop = REF_start + MB \n",
    "\n",
    "    elif var_position == \"chrom_start\": \n",
    "        REF_start = 0 + abs(shift)\n",
    "        REF_stop = MB + abs(shift)\n",
    "        print(\"Warning: Variant not centered; too close to start of chromosome.\")\n",
    "\n",
    "    elif var_position == \"chrom_centro_left\": \n",
    "        REF_start = centro_start - MB - abs(shift)\n",
    "        REF_stop = centro_start - abs(shift)\n",
    "        print(\"Warning: Variant not centered; too close to left end of centromere.\")\n",
    "        \n",
    "    elif var_position == \"chrom_centro_right\": \n",
    "        REF_start = centro_stop + abs(shift)\n",
    "        REF_stop = centro_stop + MB + abs(shift)\n",
    "        print(\"Warning: Variant not centered; too close to right end of centromere.\")\n",
    "\n",
    "    elif var_position == \"chrom_end\": \n",
    "        REF_start = chrom_max - MB - abs(shift)\n",
    "        REF_stop = chrom_max - abs(shift)\n",
    "        print(\"Warning: Variant not centered; too close to end of chromosome.\")\n",
    "        \n",
    "    elif var_position == \"centromere\":\n",
    "        raise ValueError('Centromeric variant')\n",
    "\n",
    "        \n",
    "        \n",
    "    # Get reference sequence\n",
    "    REF_seq = fasta_open.fetch(CHR, REF_start, REF_stop).upper()\n",
    "\n",
    "\n",
    "    # Error if Ns are more than 5% of sequence\n",
    "    if Counter(REF_seq)['N']/MB*100 > 5:\n",
    "        raise ValueError('N composition greater than 5%')\n",
    "\n",
    "\n",
    "\n",
    "    # Make sure that reference sequence matches given REF\n",
    "\n",
    "    if var_position == \"chrom_mid\":\n",
    "        if REF_seq[(REF_half_left - 1) : (REF_half_left - 1 + REF_len)] != REF:\n",
    "            raise ValueError('Reference allele does not match hg38.')\n",
    "\n",
    "    elif var_position == \"chrom_start\": \n",
    "        if REF_seq[(POS - abs(shift) - 1) : (POS - abs(shift) - 1 + REF_len)] != REF:\n",
    "            raise ValueError('Reference allele does not match hg38.')\n",
    "\n",
    "    elif var_position == \"chrom_centro_right\": \n",
    "        POS_adj = POS - centro_stop - abs(shift)\n",
    "        if REF_seq[(POS_adj - 1) : (POS_adj - 1 + REF_len)] != REF:\n",
    "            raise ValueError('Reference allele does not match hg38.')\n",
    "\n",
    "    elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "        if REF_seq[-(REF_stop - POS + 1) : -(REF_stop - POS + 1 - REF_len)] != REF:\n",
    "            raise ValueError('Reference allele does not match hg38.')\n",
    "\n",
    "\n",
    "    if len(REF_seq) != MB:\n",
    "            raise ValueError('Reference sequence generated is not the right length.')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # For SNPs, MNPs, Insertions: \n",
    "    if len(REF) <= len(ALT):\n",
    "\n",
    "        # Create alternate sequence: change REF sequence at position from REF to ALT\n",
    "\n",
    "        ALT_seq = REF_seq\n",
    "\n",
    "        if var_position == \"chrom_mid\":\n",
    "            ALT_seq = ALT_seq[:(REF_half_left - 1)] + ALT + ALT_seq[(REF_half_left - 1 + REF_len):]\n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            ALT_seq = ALT_seq[:(POS - abs(shift) - 1)] + ALT + ALT_seq[(POS - abs(shift) - 1 + REF_len):]\n",
    "            \n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            POS_adj = POS - centro_stop - abs(shift)\n",
    "            ALT_seq = ALT_seq[:(POS_adj - 1)] + ALT + ALT_seq[(POS_adj - 1 + REF_len):]\n",
    "\n",
    "        elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "            ALT_seq = ALT_seq[:-(REF_stop - POS + 1)] + ALT + ALT_seq[-(REF_stop - POS + 1 - REF_len):]\n",
    "            \n",
    "            \n",
    "\n",
    "\n",
    "        # Chop off ends of alternate sequence if it's longer \n",
    "        if len(ALT_seq) > len(REF_seq):\n",
    "            to_remove = (len(ALT_seq) - len(REF_seq))/2\n",
    "\n",
    "            if to_remove == 0.5:\n",
    "                ALT_seq = ALT_seq[1:]\n",
    "            else:\n",
    "                ALT_seq = ALT_seq[math.ceil(to_remove) : -math.floor(to_remove)]\n",
    "\n",
    "\n",
    "    # For Deletions\n",
    "    elif len(REF) > len(ALT):\n",
    "\n",
    "\n",
    "        del_len = len(REF) - len(ALT)\n",
    "        \n",
    "        to_add_left = math.ceil(del_len/2)\n",
    "        to_add_right = math.floor(del_len/2) \n",
    "\n",
    "        # Get start and end of reference sequence\n",
    "        if var_position == \"chrom_mid\":\n",
    "            ALT_start = REF_start - to_add_left\n",
    "            ALT_stop = REF_stop + to_add_right\n",
    "\n",
    "        if var_position == \"chrom_start\": \n",
    "            ALT_start = 0 + abs(shift)\n",
    "            ALT_stop = MB + del_len + abs(shift)\n",
    "            \n",
    "        if var_position == \"chrom_centro_left\": \n",
    "            ALT_start = centro_start - MB - del_len - abs(shift)\n",
    "            ALT_stop = centro_start - abs(shift)\n",
    "\n",
    "        if var_position == \"chrom_centro_right\": \n",
    "            ALT_start = centro_stop + abs(shift)\n",
    "            ALT_stop = centro_stop + MB + del_len + abs(shift)\n",
    "\n",
    "        if var_position == \"chrom_end\": \n",
    "            ALT_start = chrom_max - MB - del_len - abs(shift)\n",
    "            ALT_stop = chrom_max - abs(shift)\n",
    "            \n",
    "            \n",
    "        # Get alternate sequence\n",
    "        ALT_seq = fasta_open.fetch(CHR, ALT_start, ALT_stop).upper()\n",
    "        \n",
    "        \n",
    "        \n",
    "        # Make sure that alternate sequence matches REF at POS\n",
    "\n",
    "        if var_position == \"chrom_mid\":\n",
    "            if ALT_seq[(REF_half_left - 1 + to_add_left) : (REF_half_left - 1 + to_add_left + REF_len)] != REF:\n",
    "                raise ValueError('Sequence for the alternate allele does not match hg38 at REF position.')\n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            if ALT_seq[(POS - abs(shift) - 1) : (POS - abs(shift) - 1 + REF_len)] != REF:\n",
    "                raise ValueError('Sequence for the alternate allele does not match hg38 at REF position.')\n",
    "            \n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            POS_adj = POS - centro_stop\n",
    "            if ALT_seq[(POS_adj - abs(shift) - 1) : (POS_adj - abs(shift) - 1 + REF_len)] != REF:\n",
    "                raise ValueError('Sequence for the alternate allele does not match hg38 at REF position.')\n",
    "            \n",
    "        elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "            if ALT_seq[-(REF_stop - POS + 1) : -(REF_stop - POS - REF_len + 1)] != REF:\n",
    "                raise ValueError('Sequence for the alternate allele does not match hg38 at REF position.')\n",
    "\n",
    "\n",
    "    \n",
    "        # Change alternate sequence to match ALT at POS\n",
    "\n",
    "        if var_position == \"chrom_mid\":\n",
    "            # [:N] does not include N but [N:] includes N\n",
    "            ALT_seq = ALT_seq[:(REF_half_left - 1 + to_add_left)] + ALT + ALT_seq[(REF_half_left - 1 + to_add_left + REF_len):] \n",
    "\n",
    "        elif var_position == \"chrom_start\": \n",
    "            ALT_seq = ALT_seq[:(POS - abs(shift) - 1)] + ALT + ALT_seq[(POS - abs(shift) - 1 + REF_len):]\n",
    "            \n",
    "        elif var_position == \"chrom_centro_right\": \n",
    "            POS_adj = POS - centro_stop\n",
    "            ALT_seq = ALT_seq[:(POS_adj - abs(shift) - 1)] + ALT + ALT_seq[(POS_adj - abs(shift) - 1 + REF_len):]\n",
    "            \n",
    "        elif var_position in [\"chrom_end\", \"chrom_centro_left\"]: \n",
    "            ALT_seq = ALT_seq[:-(REF_stop - POS + 1)] + ALT + ALT_seq[-(REF_stop - POS - REF_len + 1):]\n",
    "\n",
    "            \n",
    "    if len(ALT_seq) != MB:\n",
    "        raise ValueError('Alternate sequence generated is not the right length.')\n",
    "         \n",
    "        \n",
    "    return REF_seq, ALT_seq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12786cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scores_BND(REF_pred_L, REF_pred_R, ALT_pred):\n",
    "    \n",
    "    # Get REF and ALT vectors, excluding diagonal \n",
    "    indexes = np.triu_indices(bins/2, 2)\n",
    "    \n",
    "    REF_UL = upper_left(REF_pred_L)[indexes]\n",
    "    REF_LR = lower_right(REF_pred_R)[indexes]\n",
    "    ALT_UL = upper_left(ALT_pred)[indexes]\n",
    "    ALT_LR = lower_right(ALT_pred)[indexes]\n",
    "    \n",
    "    seq_REF = np.append(REF_UL, REF_LR)\n",
    "    seq_ALT = np.append(ALT_UL, ALT_LR)\n",
    "    \n",
    "    # Get disruption score \n",
    "    disruption_score = get_DS_from_vector(seq_REF, seq_ALT)\n",
    "    \n",
    "    # Get spearman correlation\n",
    "    correlation, corr_pval = spearmanr(seq_REF, seq_ALT)\n",
    "    \n",
    "    if corr_pval >= 0.05:\n",
    "        correlation = 0\n",
    "    \n",
    "    return disruption_score, correlation, corr_pval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "024e4ac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import itertools\n",
    "# position_pairs = np.array(list(itertools.product(variant_positions, variant_positions)))\n",
    "\n",
    "# position_pairs = [x+'_'+y for x,y in zip(np.array(position_pairs)[:,0], np.array(position_pairs)[:,1])]\n",
    "# dict(zip(position_pairs,len(position_pairs)*['test']))\n",
    "\n",
    "\n",
    "\n",
    "get_BND_ALT_sense_by_pos = {'chrom_mid_chrom_mid': [ignore, \n",
    "                                                    ignore, \n",
    "                                                    ignore], # M M\n",
    "                         'chrom_mid_chrom_start': [adjust_right, \n",
    "                                                   adjust_right_antisense_left, \n",
    "                                                   adjust_right_antisense_right], # M S\n",
    "                         'chrom_mid_chrom_centro_left': [adjust_right, \n",
    "                                                         adjust_right_antisense_left, \n",
    "                                                         adjust_right_antisense_right], # M S\n",
    "                         'chrom_mid_chrom_centro_right': [adjust_right, \n",
    "                                                          adjust_right_antisense_left, \n",
    "                                                          adjust_right_antisense_right], # M S\n",
    "                         'chrom_mid_chrom_end': [adjust_right, \n",
    "                                                 adjust_right_antisense_left, \n",
    "                                                 adjust_right_antisense_right], # M S\n",
    "                         'chrom_start_chrom_mid': [adjust_left, \n",
    "                                                   adjust_left_antisense_left, \n",
    "                                                   adjust_left_antisense_right], # S M\n",
    "                         'chrom_start_chrom_start': [adjust_both_start, \n",
    "                                                     raise_error, \n",
    "                                                     raise_error], # S S\n",
    "                         'chrom_start_chrom_centro_left': [raise_error,\n",
    "                                                           adjust_start_end_antisense_left,\n",
    "                                                           adjust_start_end_antisense_right], # S E\n",
    "                         'chrom_start_chrom_centro_right': [adjust_both_start, \n",
    "                                                            raise_error, \n",
    "                                                            raise_error], # S S\n",
    "                         'chrom_start_chrom_end': [raise_error,\n",
    "                                                   adjust_start_end_antisense_left,\n",
    "                                                   adjust_start_end_antisense_right], # S E\n",
    "                         'chrom_centro_left_chrom_mid': [adjust_left, \n",
    "                                                         adjust_left_antisense_left, \n",
    "                                                         adjust_left_antisense_right], # E M\n",
    "                         'chrom_centro_left_chrom_start': [raise_error,\n",
    "                                                           adjust_end_start_antisense_left,\n",
    "                                                           adjust_end_start_antisense_right], # E S\n",
    "                         'chrom_centro_left_chrom_centro_left': [adjust_both_end, \n",
    "                                                                 raise_error, \n",
    "                                                                 raise_error], # E E\n",
    "                         'chrom_centro_left_chrom_centro_right': [raise_error,\n",
    "                                                                  adjust_end_start_antisense_left,\n",
    "                                                                  adjust_end_start_antisense_right], # E S\n",
    "                         'chrom_centro_left_chrom_end': [adjust_both_end, \n",
    "                                                         raise_error, \n",
    "                                                         raise_error], # E E\n",
    "                         'chrom_centro_right_chrom_mid': [adjust_left, \n",
    "                                                          adjust_left_antisense_left, \n",
    "                                                          adjust_left_antisense_right], # S M\n",
    "                         'chrom_centro_right_chrom_start': [adjust_both_start, \n",
    "                                                            raise_error, \n",
    "                                                            raise_error], # S S\n",
    "                         'chrom_centro_right_chrom_centro_left': [raise_error,\n",
    "                                                                  adjust_start_end_antisense_left,\n",
    "                                                                  adjust_start_end_antisense_right], # S E\n",
    "                         'chrom_centro_right_chrom_centro_right': [adjust_both_start, \n",
    "                                                                   raise_error, \n",
    "                                                                   raise_error], # S S\n",
    "                         'chrom_centro_right_chrom_end': [raise_error,\n",
    "                                                          adjust_start_end_antisense_left,\n",
    "                                                          adjust_start_end_antisense_right], # S E\n",
    "                         'chrom_end_chrom_mid': [adjust_left, \n",
    "                                                 adjust_left_antisense_left, \n",
    "                                                 adjust_left_antisense_right], # E M\n",
    "                         'chrom_end_chrom_start': [raise_error, \n",
    "                                                   adjust_end_start_antisense_left,\n",
    "                                                   adjust_end_start_antisense_right], # E S\n",
    "                         'chrom_end_chrom_centro_left': [adjust_both_end, \n",
    "                                                         raise_error, \n",
    "                                                         raise_error], # E E\n",
    "                         'chrom_end_chrom_centro_right': [raise_error,\n",
    "                                                          adjust_end_start_antisense_left,\n",
    "                                                          adjust_end_start_antisense_right], # E S\n",
    "                         'chrom_end_chrom_end': [adjust_both_end, \n",
    "                                                 raise_error, \n",
    "                                                 raise_error]} # E E\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cac577e",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser.add_argument('--format',\n",
    "                    dest = 'file_format', \n",
    "                    help = 'Format for input file.', \n",
    "                    type = str,\n",
    "                    choices = ['vcf', 'df'],\n",
    "                    default = 'vcf',\n",
    "                    required = False)\n",
    "\n",
    "parser.add_argument('--type',\n",
    "                    dest = 'var_type', \n",
    "                    help = 'Variant type: simple or SV.', \n",
    "                    type = str,\n",
    "                    choices = ['simple', 'SV'],\n",
    "                    default = 'SV',\n",
    "                    required = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7883cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# it cannot take in boolean values\n",
    "\n",
    "parser.add_argument('--revcomp',\n",
    "                    dest = 'revcomp', \n",
    "                    nargs = '+', \n",
    "                    help = 'Make predictions for the reverse compliment sequence.', \n",
    "                    choices = [True, False],\n",
    "                    default = [False],\n",
    "                    required = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba7f238b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_input(in_file, file_format, var_type):\n",
    "\n",
    "    # Read and reformat variant dataset\n",
    "    \n",
    "    if file_format == 'df':\n",
    "        if var_type == 'simple':\n",
    "            variants = (pd.read_csv(in_file, skiprows = 1, sep = '\\t')\n",
    "                        .rename(columns = {'Chromosome':'CHROM', \n",
    "                                           'Start_Position':'POS',\n",
    "                                           'End_Position':'END',\n",
    "                                           'Reference_Allele':'REF',\n",
    "                                           'Tumor_Seq_Allele2':'ALT'})\n",
    "                       [['CHROM', 'POS', 'END', 'REF', 'ALT']])\n",
    "            variants['SVLEN'] = (variants.END - variants.POS).astype('int') # this SVLEN (END-POS) would be 0 for SNPs\n",
    "\n",
    "            # Might need to do this if there are homozygous variants where Tumor_Seq_Allele1 != Reference_Allele\n",
    "            # would need to only do this for those variants and concat with the rest\n",
    "            # variants = pd.melt(variants, \n",
    "            #                    id_vars = ['CHROM', 'POS', 'REF'], \n",
    "            #                    value_vars = ['ALT1', 'ALT2'], \n",
    "            #                    var_name = 'allele', \n",
    "            #                    value_name='ALT')\n",
    "\n",
    "            for i in variants[variants.ALT == '-'].index:\n",
    "                variants.loc[i,'SVTYPE'] = 'DEL'\n",
    "            for i in variants[variants.REF == '-'].index:\n",
    "                variants.loc[i,'SVTYPE'] = 'INS'\n",
    "\n",
    "            var_type = 'SV' # These are formatted like SVs\n",
    "\n",
    "        elif var_type == 'SV':\n",
    "            variants = (pd.read_csv(in_file, sep = '\\t', low_memory=False)\n",
    "                        .rename(columns = {'SV_chrom':'CHROM', \n",
    "                                           'SV_start':'POS',\n",
    "                                           'SV_end':'END', \n",
    "                                           'SV_type':'SVTYPE',\n",
    "                                           'SV_length':'SVLEN'})\n",
    "                       [['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']])\n",
    "            variants['CHROM'] = ['chr' + str(x) for x in variants['CHROM']]\n",
    "            variants.loc[~pd.isnull(variants.END), 'END'] = variants.loc[~pd.isnull(variants.END), 'END'].astype('int')\n",
    "\n",
    "\n",
    "    elif file_format == 'vcf':\n",
    "        if in_file.endswith('.gz'):\n",
    "            variants = read_vcf_gz(in_file)\n",
    "        else:\n",
    "            variants = read_vcf(in_file)\n",
    "            \n",
    "        if var_type == 'simple':\n",
    "            variants = variants[['CHROM', 'POS', 'REF', 'ALT']]\n",
    "\n",
    "        elif var_type == 'SV':\n",
    "            variants['END'] = variants.INFO.str.split('END=').str[1].str.split(';').str[0] # this SVLEN (END-POS) would be 0 for SNPs\n",
    "            variants.loc[~pd.isnull(variants.END), 'END'] = variants.loc[~pd.isnull(variants.END), 'END'].astype('int')\n",
    "            variants['SVTYPE'] = variants.INFO.str.split('SVTYPE=').str[1].str.split(';').str[0]\n",
    "            variants['SVLEN'] = variants.INFO.str.split('SVLEN=').str[1].str.split(';').str[0]\n",
    "            variants = variants[['CHROM', 'POS', 'END', 'REF', 'ALT', 'SVTYPE', 'SVLEN']]\n",
    "            \n",
    "    variants.reset_index(inplace = True, drop = True)\n",
    "    \n",
    "    return variants\n",
    "     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d28b24c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences_BND(CHR, POS, ALT, shift):\n",
    "\n",
    "    if '[' in ALT:\n",
    "\n",
    "        if ALT[0] in nt:\n",
    "\n",
    "            # t[p[\n",
    "\n",
    "            CHR2 = ALT.split(':')[0].split('[')[1]\n",
    "            POS2 = int(ALT.split('[')[1].split(':')[1])\n",
    "            ALT_t = ALT.split('[')[0]\n",
    "\n",
    "            ALT_left = fasta_open.fetch(CHR, POS - half_patch_size + shift, POS).upper() # don't inlcude POS\n",
    "\n",
    "            ALT_right = fasta_open.fetch(CHR2, POS2 + 1, POS2 + 1 + half_patch_size + shift).upper() \n",
    "\n",
    "            REF_for_left = fasta_open.fetch(CHR, POS - half_patch_size + shift, POS + half_patch_size + shift).upper()\n",
    "            REF_for_right = fasta_open.fetch(CHR2, POS2 - half_patch_size + shift, POS2 + half_patch_size + shift).upper() \n",
    "\n",
    "\n",
    "        elif ALT[0] not in nt:\n",
    "\n",
    "            #  [p[t\n",
    "\n",
    "            CHR2 = ALT.split(':')[0].split('[')[1]\n",
    "            POS2 = int(ALT.split('[')[1].split(':')[1])\n",
    "            ALT_t = ALT.split('[')[2]\n",
    "\n",
    "            ALT_left_revcomp = fasta_open.fetch(CHR2, POS2 + 1, POS2 + 1 + half_patch_size - shift).upper() # don't include POS2\n",
    "            ALT_left = str(Seq(ALT_left_revcomp).reverse_complement())\n",
    "\n",
    "            ALT_right = fasta_open.fetch(CHR, POS + 1, POS + 1 + half_patch_size + shift).upper()\n",
    "\n",
    "            REF_for_left_revcomp = fasta_open.fetch(CHR2, POS2 - half_patch_size - shift, POS2 + half_patch_size - shift).upper() \n",
    "            REF_for_left = str(Seq(REF_for_left_revcomp).reverse_complement())\n",
    "            REF_for_right = fasta_open.fetch(CHR, POS - half_patch_size + shift, POS + half_patch_size + shift).upper() \n",
    "\n",
    "\n",
    "    elif ']' in ALT:\n",
    "\n",
    "        if ALT[0] in nt:\n",
    "\n",
    "            # t]p]\n",
    "\n",
    "            CHR2 = ALT.split(':')[0].split(']')[1]\n",
    "            POS2 = int(ALT.split(']')[1].split(':')[1])\n",
    "            ALT_t = ALT.split(']')[0]\n",
    "\n",
    "            ALT_left = fasta_open.fetch(CHR, POS - half_patch_size + shift, POS).upper() # don't include POS\n",
    "\n",
    "            ALT_right_revcomp = fasta_open.fetch(CHR2, POS2 - half_patch_size - shift, POS2).upper() # don't include POS2\n",
    "            ALT_right = str(Seq(ALT_right_revcomp).reverse_complement())\n",
    "\n",
    "            REF_for_left = fasta_open.fetch(CHR, POS - half_patch_size + shift, POS + half_patch_size + shift).upper()\n",
    "            REF_for_right_revcomp = fasta_open.fetch(CHR2, POS2 - half_patch_size - shift, POS2 + half_patch_size - shift).upper()\n",
    "            REF_for_right = str(Seq(REF_for_right_revcomp).reverse_complement())\n",
    "\n",
    "\n",
    "\n",
    "        elif ALT[0] not in nt:\n",
    "\n",
    "            # ]p]t\n",
    "\n",
    "            CHR2 = ALT.split(':')[0].split(']')[1]\n",
    "            POS2 = int(ALT.split(']')[1].split(':')[1])\n",
    "            ALT_t = ALT.split(']')[2]\n",
    "\n",
    "            ALT_left = fasta_open.fetch(CHR2, POS2 - half_patch_size + shift, POS2).upper() # don't include POS2\n",
    "\n",
    "            ALT_right = fasta_open.fetch(CHR, POS + 1, POS + 1 + half_patch_size + shift).upper()\n",
    "\n",
    "            REF_for_left = fasta_open.fetch(CHR2, POS2 - half_patch_size + shift, POS2 + half_patch_size + shift).upper() \n",
    "            REF_for_right = fasta_open.fetch(CHR, POS - half_patch_size + shift, POS + half_patch_size + shift).upper() \n",
    "\n",
    "\n",
    "    ALT_seq = ALT_left + ALT_t + ALT_right\n",
    "\n",
    "    # chop off the sides if longer than ~1MB\n",
    "    if len(ALT_seq) > MB:\n",
    "        to_remove = (len(ALT_seq) - MB)/2\n",
    "\n",
    "        if to_remove == 0.5:\n",
    "            ALT_seq = ALT_seq[1:]\n",
    "        else:\n",
    "            ALT_seq = ALT_seq[math.ceil(to_remove) : -math.floor(to_remove)]\n",
    "\n",
    "        \n",
    "    return REF_for_left, REF_for_right, ALT_seq\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "227px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
